{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "consecutive-sword",
   "metadata": {},
   "source": [
    "# Reads pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "loaded-trading",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from starmap.sequencing import *\n",
    "from natsort import natsorted\n",
    "from scipy.io import loadmat, savemat\n",
    "from skimage.filters import threshold_otsu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "obvious-senior",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import wraps\n",
    "from time import time\n",
    "\n",
    "# Timer\n",
    "def timer(func):\n",
    "    @wraps(func)\n",
    "    def _time_it(*args, **kwargs):\n",
    "        start = int(round(time() * 1000))\n",
    "        try:\n",
    "            return func(*args, **kwargs)\n",
    "        finally:\n",
    "            end_ = int(round(time() * 1000)) - start\n",
    "            end_ = round(end_ / 1000, 4)\n",
    "            print(f\"Total execution time: {end_ if end_ > 0 else 0} s\")\n",
    "    return _time_it\n",
    "\n",
    "\n",
    "# Trim reads \n",
    "@timer\n",
    "def trim_reads(sample_dir, save_as=True):\n",
    "    \n",
    "    print(f\"Trimming reads...\")\n",
    "    current_coordinates = np.loadtxt(os.path.join(sample_dir, 'trim.txt'), dtype=int, delimiter=',')\n",
    "    current_dots = loadmat(os.path.join(sample_dir, 'merged_goodPoints_max3d.mat'))\n",
    "    # Get reads\n",
    "    bases = [str(i[0][0]) for i in current_dots[\"merged_reads\"]]\n",
    "    bases = np.array(bases)\n",
    "    # Get reads location\n",
    "    temp = current_dots[\"merged_points\"]\n",
    "    # Trim reads \n",
    "    to_remove = (temp[:, 0] < current_coordinates[0, 0]) | (temp[:, 0] > current_coordinates[1, 0]) | (temp[:, 1] < current_coordinates[0, 1]) | (temp[:, 1] > current_coordinates[1, 1])\n",
    "    temp = temp[~to_remove, :]\n",
    "    temp[:, 0] = temp[:, 0] - current_coordinates[0, 0]\n",
    "    temp[:, 1] = temp[:, 1] - current_coordinates[0, 1]\n",
    "    bases = bases[~to_remove]\n",
    "\n",
    "    if save_as:\n",
    "        output_dict = {'trimmed_reads': bases, 'trimmed_points': temp}\n",
    "        savemat(os.path.join(sample_dir, 'trimmed_goodPoints_max3d.mat'), output_dict)\n",
    "\n",
    "    # Convert to 0 indexed and switch axis for python\n",
    "    temp = temp[:, :2]\n",
    "    points = np.zeros(temp.shape)\n",
    "    points[:, 0] = np.round(temp[:, 1]-1)\n",
    "    points[:, 1] = np.round(temp[:, 0]-1)\n",
    "    print(f\"Number of reads: {len(bases)}\")\n",
    "    \n",
    "    return points, bases\n",
    "\n",
    "\n",
    "# Load genes.csv\n",
    "def load_genes(base_path):\n",
    "    genes2seqs = {}\n",
    "    seqs2genes = {}\n",
    "    with open(os.path.join(base_path, \"genes.csv\"), encoding='utf-8-sig') as f:\n",
    "        for l in f:\n",
    "            fields = l.rstrip().split(\",\")\n",
    "            # print(\"\".join([str(s+1) for s in encode_SOLID(fields[1][::-1])]))\n",
    "            curr_seg = \"\".join([str(s+1) for s in encode_SOLID(fields[1][::-1])])\n",
    "            curr_seg = curr_seg[5:] + curr_seg[:4]\n",
    "            # print(curr_seg)\n",
    "            genes2seqs[fields[0]] = curr_seg\n",
    "            seqs2genes[genes2seqs[fields[0]]] = fields[0]\n",
    "            \n",
    "    return genes2seqs, seqs2genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "unusual-somalia",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AD_mouse9494', 'AD_mouse9723', 'AD_mouse9735', 'AD_mouse9498']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parameter\n",
    "points_id = 'max3d'\n",
    "\n",
    "# IO path \n",
    "base_path = './'\n",
    "out_path = os.path.join(base_path, 'output', 'reads_pattern')\n",
    "if not os.path.exists(out_path):\n",
    "    os.mkdir(out_path)\n",
    "    \n",
    "sample_dirs = [d for d in os.listdir(base_path) if d.startswith(\"AD\")]\n",
    "sample_dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "hybrid-savings",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: AD_mouse9494\n",
      "Trimming reads...\n",
      "Number of reads: 5758057\n",
      "Total execution time: 51.054 s\n",
      "NRGN\n",
      "GPM6A\n",
      "CPLX1\n",
      "CPLX2\n",
      "Current sample: AD_mouse9723\n",
      "Trimming reads...\n",
      "Number of reads: 2995745\n",
      "Total execution time: 21.408 s\n",
      "NRGN\n",
      "GPM6A\n",
      "CPLX1\n",
      "CPLX2\n",
      "Current sample: AD_mouse9735\n",
      "Trimming reads...\n",
      "Number of reads: 4559723\n",
      "Total execution time: 33.213 s\n",
      "NRGN\n",
      "GPM6A\n",
      "CPLX1\n",
      "CPLX2\n",
      "Current sample: AD_mouse9498\n",
      "Trimming reads...\n",
      "Number of reads: 3658964\n",
      "Total execution time: 26.632 s\n",
      "NRGN\n",
      "GPM6A\n",
      "CPLX1\n",
      "CPLX2\n"
     ]
    }
   ],
   "source": [
    "genes = ['NRGN', 'GPM6A', 'CPLX1', 'CPLX2']\n",
    "\n",
    "\n",
    "for current_sample in sample_dirs:\n",
    "    \n",
    "    print(f\"Current sample: {current_sample}\")\n",
    "\n",
    "    # Load reads \n",
    "    points, bases = trim_reads(current_sample)\n",
    "\n",
    "    # Load genes\n",
    "    genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "    # Load dapi label\n",
    "    seg = load_label_image(os.path.join(current_sample, 'segmentation'), fname='labeled_cells.tif')\n",
    "\n",
    "    # Load 2D Nissl image \n",
    "    bk_img = load_nissl_image(os.path.join(current_sample, 'trimmed_images'), fname=\"dots_pi_max_overlay.tif\")\n",
    "\n",
    "    current_out_path = os.path.join(out_path, current_sample)\n",
    "    if not os.path.exists(current_out_path):\n",
    "        os.mkdir(current_out_path)\n",
    "        \n",
    "    points = points.astype(int)\n",
    "    reads_assignment = seg[points[:, 0], points[:, 1]]\n",
    "    \n",
    "    # Get assigned reads \n",
    "    assigned_index = np.argwhere(reads_assignment != 0).flatten()\n",
    "    assigned_bases = bases[assigned_index]\n",
    "    assigned_points = points[assigned_index, :]\n",
    "    \n",
    "    for i, gene in enumerate(genes):\n",
    "        print(gene)\n",
    "#         curr_index = np.argwhere(assigned_bases == genes2seqs[gene]).flatten()\n",
    "#         curr_points = assigned_points[curr_index, :]\n",
    "#         n_reads = curr_points.shape[0]\n",
    "        \n",
    "        curr_index = np.argwhere(bases == genes2seqs[gene]).flatten()\n",
    "        curr_points = points[curr_index, :]\n",
    "        n_reads = curr_points.shape[0]\n",
    "\n",
    "        # Plot\n",
    "        plt.figure(figsize=(np.floor(bk_img.shape[1] / 1000 * 5), np.floor(bk_img.shape[0] / 1000 * 5)))\n",
    "        plt.imshow(bk_img, cmap='gray')\n",
    "        plt.plot(curr_points[:, 1], curr_points[:, 0], '.', color='red', markersize=5)\n",
    "        plt.axis('off')\n",
    "        expr_figure_path = os.path.join(current_out_path, f\"{i+1}.{gene}_{n_reads}.png\")\n",
    "        plt.savefig(expr_figure_path)\n",
    "        plt.clf()\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "copyrighted-presence",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot\n",
    "plt.figure(figsize=(np.floor(bk_img.shape[1] / 1000 * 5), np.floor(bk_img.shape[0] / 1000 * 5)))\n",
    "plt.imshow(bk_img, cmap='gray')\n",
    "plt.plot(assigned_points[:, 1], assigned_points[:, 0], '.', color='red', markersize=1)\n",
    "plt.axis('off')\n",
    "expr_figure_path = os.path.join(current_out_path, \"assigned_points.png\")\n",
    "plt.savefig(expr_figure_path)\n",
    "plt.clf()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nuclear-library",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
