{
 "cells": [
  {
   "cell_type": "raw",
   "id": "cfd621a8",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"AD_Apr12\"\n",
    "author: \"Haowen Zhou\"\n",
    "date: \"4/12/2021\"\n",
    "output: html_document\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9f4640e",
   "metadata": {
    "name": "setup",
    "tags": [
     "remove_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(echo = TRUE)\n",
    "getwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "054e29a0",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "994a8dc2",
   "metadata": {
    "name": "load",
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "setwd(\"~/Desktop/At_Broad/AD\")\n",
    "#setwd(\"D:/share/AD\")\n",
    "#Convert Data\n",
    "#Convert(\"data/2022-03-31-Hu-AD-stardist-scaled.h5ad\", dest = \"h5seurat\", overwrite = TRUE)                                               \n",
    "#Convert(\"data/2022-02-13-Hu-AD-raw-temp.h5ad\", dest = \"h5seurat\", overwrite = TRUE)\n",
    "#Convert(\"2022-08-22-Hu-AD-stardist-scaled.h5ad\", dest = \"h5seurat\", overwrite = TRUE)\n",
    "\n",
    "#source(\"D:/share/AD/R_pub_code/func/AD_FunctionDefinition.R\")\n",
    "#source(\"D:/share/AD/R_pub_code/func/AD_DE_Function.R\")\n",
    "#source(\"D:/share/AD/R_pub_code/func/AD_SpatialFunction2.R\")\n",
    "#source(\"D:/share/AD/R_pub_code/func/AD_Pseudo-timeFunction.R\")\n",
    "\n",
    "source(\"R_pub_code/func/AD_FunctionDefinition.R\")\n",
    "source(\"R_pub_code/func/AD_DE_Function.R\")\n",
    "source(\"R_pub_code/func/AD_SpatialFunction2.R\")\n",
    "source(\"R_pub_code/func/AD_Pseudo-timeFunction.R\")\n",
    "#Load Scaled Data\n",
    "ad_ini()\n",
    "res_factor = 0.3 #resolution factor\n",
    "\n",
    "starmap_raw <- LoadH5Seurat(\"~/Desktop/At_Broad/AD/data/2022-08-28-Hu-AD-stardist-scaled.h5seurat\")\n",
    "starmap_64  <- LoadH5Seurat(\"~/Desktop/At_Broad/AD/data/2021-04-22-starmap-mAD-64-genes-raw.h5seurat\")\n",
    "\n",
    "#Merge Alveus Hippo\n",
    "#starmap_raw@misc$ADmouse_9723_2_morph[[\"region_bk\"]] <- starmap_raw@misc$ADmouse_9723_2_morph$region\n",
    "starmap_raw@misc$ADmouse_9723_2_morph[[\"region\"]][starmap_raw@misc$ADmouse_9723_2_morph[[\"region\"]] == 4] <- 3\n",
    "#starmap_raw@misc$ADmouse_9723_morph[[\"region_bk\"]] <- starmap_raw@misc$ADmouse_9723_morph$region\n",
    "starmap_raw@misc$ADmouse_9723_morph[[\"region\"]][starmap_raw@misc$ADmouse_9723_morph[[\"region\"]] == 4] <- 3\n",
    "#starmap_raw@misc$ADmouse_9494_morph[[\"region_bk\"]] <- starmap_raw@misc$ADmouse_9494_morph$region\n",
    "starmap_raw@misc$ADmouse_9494_morph[[\"region\"]][starmap_raw@misc$ADmouse_9494_morph[[\"region\"]] == 4] <- 3\n",
    "#starmap_raw@misc$ADmouse_11346_morph[[\"region_bk\"]] <- starmap_raw@misc$ADmouse_11346_morph$region\n",
    "starmap_raw@misc$ADmouse_11346_morph[[\"region\"]][starmap_raw@misc$ADmouse_11346_morph[[\"region\"]] == 4] <- 3\n",
    "#remove normal sample img\n",
    "starmap_raw@misc[[\"ADmouse_11351_morph\"]] <- NULL\n",
    "starmap_raw@misc[[\"ADmouse_9498_morph\"]] <- NULL\n",
    "starmap_raw@misc[[\"ADmouse_9735_morph\"]] <- NULL\n",
    "starmap_raw@misc[[\"ADmouse_9707_morph\"]] <- NULL\n",
    "gc()\n",
    "\n",
    "\n",
    "starmap_raw@meta.data$x <-starmap_raw@meta.data$x*res_factor\n",
    "starmap_raw@meta.data$y <-starmap_raw@meta.data$y*res_factor\n",
    "starmap_raw@meta.data$region_bk <- starmap_raw@meta.data$region#separated Alveus\n",
    "starmap_raw@meta.data$region <- starmap_raw@meta.data$region_merged\n",
    "starmap_raw@meta.data$cell_type_bk <- starmap_raw@meta.data$cell_type\n",
    "starmap_raw@meta.data$cell_type <- as.character(starmap_raw@meta.data$cell_type_bk)\n",
    "starmap_raw@meta.data$cell_type[starmap_raw@meta.data$top_level == \"Astro\"] <- as.character(starmap_raw@meta.data$cell_type_test[starmap_raw@meta.data$top_level == \"Astro\"])\n",
    "starmap_raw@meta.data$cell_type <- as.factor(starmap_raw@meta.data$cell_type)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "599f2a17",
   "metadata": {
    "name": "Prepare",
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#AD palette list\n",
    "top_level_palette_list <- as.character(unlist(starmap_raw@misc[[\"top_level_hex_dict\"]]))[!colnames(starmap_raw@misc[[\"top_level_hex_dict\"]]) %in% c(\"LHb\")]\n",
    "names(top_level_palette_list) <- colnames(starmap_raw@misc[[\"top_level_hex_dict\"]])[!colnames(starmap_raw@misc[[\"top_level_hex_dict\"]]) %in% c(\"LHb\")]\n",
    "names(top_level_palette_list)[5] <- \"CTX-Ex\"\n",
    "top_level_palette_list <- top_level_palette_list[c(2:9,1,10:12)]\n",
    "\n",
    "cell_type_list <- list(\"Ex\" = c(\"CTX-Ex1\", \"CTX-Ex2\", \"CTX-Ex3\", \"CTX-Ex4\"),\n",
    "                       \"Micro\" = c(\"Micro1\", \"Micro2\", \"Micro3\"),\n",
    "                       \"Oligo\" = c(\"Oligo1\", \"Oligo2\", \"Oligo3\",\"OPC\"),\n",
    "                       \"Astro\" = c(\"Astro1\",\"Astro2\",\"Astro3\"),\n",
    "                       \"Endo\" = c(\"Endo1\",\"Endo2\",\"Endo3\"),\n",
    "                       \"Inhi\" = c(\"Cnr1\",\"Lamp5\",\"Pvalb\",\"Pvalb_Nog\",\"Sst\",\"Vip\"),\n",
    "                       \"CADG\" = c(\"CA1-Ex\",\"CA2-Ex\",\"CA3-Ex\",\"DG\"))\n",
    "\n",
    "cell_type_palette_list <- list(\"Ex\" = c(\"CTX-Ex1\"=\"#70fffd\", \"CTX-Ex2\"=\"#ffb870\", \n",
    "                                        \"CTX-Ex3\"=\"#fbff00\", \"CTX-Ex4\"=\"#91ff00\"),\n",
    "                    \"Micro\" = c(\"Micro1\"=\"#2fe04d\", \"Micro2\"=\"#2c73e6\", \"Micro3\"=\"#fa733e\"),\n",
    "                    \"Oligo\" = c(\"Oligo1\"=\"#2fe04d\", \"Oligo2\"=\"#2c73e6\", \n",
    "                                \"Oligo3\"=\"#fa733e\", \"OPC\"=\"#8600d4\"),\n",
    "                    \"Astro\" = c(\"Astro1\"=\"#2fe04d\",\"Astro2\"=\"#2c73e6\",\"Astro3\"=\"#fa733e\"),\n",
    "                    #\"OPC\" = c(\"#0dd193\",\"#7a9deb\"),\n",
    "                    \"Inhi\" = c(\"Cnr1\"=\"#3774f8\",\"Lamp5\"=\"#2fccc2\",\"Pvalb\"=\"#43b500\",\n",
    "                               \"Pvalb_Nog\"=\"#bcaa03\",\"Sst\"=\"#f19cd3\",\"Vip\"=\"#bd3a3a\"),\n",
    "                    \"CADG\" = c(\"CA1-Ex\"=\"#70fffd\",\"CA2-Ex\"=\"#0084ff\",\"CA3-Ex\"=\"#9500ff\",\"DG\"=\"#910000\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "902441d9",
   "metadata": {},
   "source": [
    "## Calc Plaque Border-Dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e078899b",
   "metadata": {
    "name": "calc border-dist",
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_11346_morph\"]], 400, \"ADmouse_11346\", top_level_palette_list, \"all\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9723_2_morph\"]], 400, \"ADmouse_9723_2\", top_level_palette_list, \"all\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9723_morph\"]], 400, \"ADmouse_9723\", top_level_palette_list, \"all\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9494_morph\"]], 400, \"ADmouse_9494\", top_level_palette_list, \"all\")\n",
    "\n",
    "#Backup for WGCNA analysis\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9494_morph\"]], 400, \"ADmouse_9494\", top_level_palette_list, \"Cortex\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9494_morph\"]], 400, \"ADmouse_9494\", top_level_palette_list, \"Hippocampus\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9723_morph\"]], 400, \"ADmouse_9723\", top_level_palette_list, \"Cortex\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9723_morph\"]], 400, \"ADmouse_9723\", top_level_palette_list, \"Hippocampus\")\n",
    "\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_11346_morph\"]], 400, \"ADmouse_11346\", top_level_palette_list, \"Cortex\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_11346_morph\"]], 400, \"ADmouse_11346\", top_level_palette_list, \"Hippocampus\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9723_2_morph\"]], 400, \"ADmouse_9723_2\", top_level_palette_list, \"Cortex\")\n",
    "starmap_raw <- ad_calc_plaque_dist2(starmap_raw,starmap_raw@misc[[\"ADmouse_9723_2_morph\"]], 400, \"ADmouse_9723_2\", top_level_palette_list, \"Hippocampus\")\n",
    "\n",
    "tmp_obj <- starmap_raw\n",
    "tmp_obj@misc[[\"ADmouse_11346_morph\"]] <- NULL\n",
    "tmp_obj@misc[[\"ADmouse_11351_morph\"]] <- NULL\n",
    "tmp_obj@misc[[\"ADmouse_9494_morph\"]] <- NULL\n",
    "tmp_obj@misc[[\"ADmouse_9498_morph\"]] <- NULL\n",
    "tmp_obj@misc[[\"ADmouse_9723_morph\"]] <- NULL\n",
    "tmp_obj@misc[[\"ADmouse_9735_morph\"]] <- NULL\n",
    "tmp_obj@misc[[\"ADmouse_9707_morph\"]] <- NULL\n",
    "tmp_obj@misc[[\"ADmouse_9723_2_morph\"]] <- NULL\n",
    "saveRDS(tmp_obj, \"Aug04_starmap_w_all_spatial.RDS\")\n",
    "#load\n",
    "starmap_raw <- readRDS(\"Aug04_starmap_w_all_spatial.RDS\")\n",
    "\n",
    "starmap_raw <- ad_calc_plaque_dist(starmap_raw,starmap_raw@misc[[\"ADmouse_9494_morph\"]], 400, \"ADmouse_9494\", top_level_palette_list, \"Corpus Callosum\")\n",
    "starmap_raw <- ad_calc_plaque_dist(starmap_raw,starmap_raw@misc[[\"ADmouse_11346_morph\"]], 400, \"ADmouse_11346\", top_level_palette_list, \"Corpus Callosum\")\n",
    "starmap_raw <- ad_calc_plaque_dist(starmap_raw,starmap_raw@misc[[\"ADmouse_9723_morph\"]], 400, \"ADmouse_9723\", top_level_palette_list, \"Corpus Callosum\")\n",
    "starmap_raw <- ad_calc_plaque_dist(starmap_raw,starmap_raw@misc[[\"ADmouse_9723_2_morph\"]], 400, \"ADmouse_9723_2\", top_level_palette_list, \"Corpus Callosum\")\n",
    "\n",
    "#Export Cell Meta\n",
    "#top level\n",
    "wb = createWorkbook()\n",
    "for(i in c('ADmouse_9723','ADmouse_9494','ADmouse_9723_2','ADmouse_11346')){\n",
    "  sheet <- createSheet(wb, i)\n",
    "  addDataFrame(starmap_raw@misc[[paste0(i,\"_plaque_all\")]][[paste0(i,\"_cell_meta\")]], sheet=sheet, startColumn=1, startRow=1, row.names=T)\n",
    "}\n",
    "saveWorkbook(wb, \"Sep04_cell_meta.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "473c7c23",
   "metadata": {},
   "source": [
    "## Stacked Barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3181ce8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Cortex \n",
    "sample_list <- c('ADmouse_9723','ADmouse_9494','ADmouse_9723_2','ADmouse_11346')\n",
    "plot_list <- list()\n",
    "\n",
    "for(area_name in c(\"Cortex\",\"Hippocampus\",\"all\")){\n",
    "  for(i in 1:2){\n",
    "    \n",
    "    cell_meta <- rbind(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i],\"cell_meta\")]],\n",
    "                       starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i+2],\"cell_meta\")]])\n",
    "    if(area_name == \"Cortex\"){\n",
    "      palette_list <- top_level_palette_list[c(4,6:12)]\n",
    "      j = sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"morph\")]][[\"region\"]] == 1) + \n",
    "        sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"morph\")]][[\"region\"]] == 1)\n",
    "    }else if(area_name == \"Hippocampus\"){\n",
    "      palette_list <- top_level_palette_list[c(1:3,6:12)]\n",
    "      j = sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"morph\")]][[\"region\"]] == 3) + \n",
    "        sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"morph\")]][[\"region\"]] == 3)\n",
    "    }else{\n",
    "      palette_list <- top_level_palette_list[-5]\n",
    "      j = sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"morph\")]][[\"region\"]] != 0) + \n",
    "        sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"morph\")]][[\"region\"]] != 0)\n",
    "    }\n",
    "    plot_list[[paste(i,area_name,sep = \"_\")]] <- ad_plot_stacked_bar(cell_meta$interval,\n",
    "                                          cell_meta$top_level,\n",
    "                                          palette_list,\n",
    "                                          paste(\"Top Level\",sample_list[i],sample_list[i+2],area_name),\n",
    "                                          starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i],\"plaque_dilate_area\")]] + \n",
    "                                            starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i+2],\"plaque_dilate_area\")]], img_size = j)\n",
    "  }\n",
    "}\n",
    "\n",
    "for(j in c(\"Micro\",\"Astro\",\"Oligo\",\"Ex\",\"Inhi\",\"CADG\")){\n",
    "  for(area_name in c(\"Cortex\",\"Hippocampus\",\"all\")){\n",
    "    for(i in 1:2){\n",
    "      if(area_name == \"Cortex\"){\n",
    "      k = sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"morph\")]][[\"region\"]] == 1) + \n",
    "        sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"morph\")]][[\"region\"]] == 1)\n",
    "    }else if(area_name == \"Hippocampus\"){\n",
    "      k = sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"morph\")]][[\"region\"]] == 3) + \n",
    "        sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"morph\")]][[\"region\"]] == 3)\n",
    "    }else{\n",
    "      k = sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"morph\")]][[\"region\"]] != 0) + \n",
    "        sum(starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"morph\")]][[\"region\"]] != 0)\n",
    "    }\n",
    "      \n",
    "      cell_meta <- rbind(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i],\"cell_meta\")]],\n",
    "                       starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i+2],\"cell_meta\")]])\n",
    "      \n",
    "      plot_list[[paste(sample_list[i],sample_list[i+2],area_name,j,sep = \"_\")]] <- ad_plot_stacked_bar(cell_meta$interval,\n",
    "                                            cell_meta$cell_type,\n",
    "                                            cell_type_palette_list[[j]],\n",
    "                                            paste(sample_list[i],sample_list[i+2],j,area_name),\n",
    "                                            starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i],\"plaque_dilate_area\")]] + \n",
    "                                            starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i+2],\"plaque_dilate_area\")]], img_size = k)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "}\n",
    "\n",
    "pdf(file = \"Aug30_new_cell_type_density.pdf\", width = 24,height = 12)\n",
    "ggarrange(plotlist = plot_list,ncol = 3, nrow = 2) %>% print()\n",
    "dev.off()\n",
    "\n",
    "#Cell Type stacked bar plot \n",
    "for(i in 1:2){\n",
    "  cell_meta <- rbind(starmap_raw@misc[[paste(sep = \"_\",sample_list[i],\"plaque\",\"all\")]][[paste(sep = \"_\",sample_list[i],\"cell_meta\")]],\n",
    "                     starmap_raw@misc[[paste(sep = \"_\",sample_list[i+2],\"plaque\",\"all\")]][[paste(sep = \"_\",sample_list[i+2],\"cell_meta\")]])\n",
    "  cell_meta <- subset(cell_meta, interval %in% c(\"10um\",\"20um\",\"30um\",\"40um\") & \n",
    "                        cell_type %in% c(cell_type_list$Ex,cell_type_list$CADG,cell_type_list$Inhi,\n",
    "                                         \"Astro3\",\"Micro3\",\"Oligo2\",\"Oligo3\",\"OPC\"))\n",
    "  cell_meta$cell_type <- as.character(cell_meta$cell_type)\n",
    "  cell_meta$cell_type[cell_meta$cell_type %in% c(cell_type_list$Ex,cell_type_list$CADG,cell_type_list$Inhi)] <- \"Neuron\"\n",
    "  cell_meta$cell_type[cell_meta$cell_type %in% c(\"Oligo2\",\"Oligo3\")] <- \"Oligo2/3\"\n",
    "  \n",
    "  palette_list <- c(cell_type_palette_list$Micro[3],\n",
    "                    cell_type_palette_list$Astro[3],\n",
    "                    \"Oligo2/3\" = \"#BA4D9B\",\n",
    "                    cell_type_palette_list$Oligo[4],\n",
    "                    \"Neuron\" = \"#999999\")\n",
    "  cell_meta$cell_type <- factor(cell_meta$cell_type,levels = c(\"Astro3\",\"Micro3\",\"Oligo2/3\",\"OPC\",\"Neuron\"))\n",
    "  cell_meta <- cell_meta %>% group_by(interval,cell_type) %>% summarise(count = n())\n",
    "  for(j in 1:4){\n",
    "      k = (starmap_raw@misc[[paste0(sample_list[i+2],\"_plaque_all\")]][[paste0(sample_list[i+2],\"_plaque_dilate_area\")]][j] + \n",
    "    starmap_raw@misc[[paste0(sample_list[i],\"_plaque_all\")]][[paste0(sample_list[i],\"_plaque_dilate_area\")]][j])/2\n",
    "    cell_meta$count[cell_meta$interval == paste0(j,\"0um\")] <- cell_meta$count[cell_meta$interval == paste0(j,\"0um\")]/k*1000000\n",
    "  }\n",
    "  \n",
    "  plot_list[[i]] <- ggplot(data= cell_meta, aes(x = interval,y=count, fill = cell_type)) + \n",
    "    geom_bar(stat = \"identity\") +\n",
    "    facet_grid(cell_type~., scales = \"free_y\", switch = \"y\") +\n",
    "    theme(legend.position = \"none\", panel.spacing = unit(0, \"lines\"),\n",
    "          plot.title = element_text(hjust = 0.5),\n",
    "          strip.background = element_blank(),\n",
    "          strip.text = element_text(face = \"bold\"),\n",
    "          strip.text.y.left = element_text(angle = 0),\n",
    "          panel.border = element_blank(),\n",
    "          panel.grid.major = element_blank(),\n",
    "          panel.grid.minor = element_blank()) + \n",
    "    scale_fill_manual(values = palette_list) + \n",
    "    labs(title = paste(sample_list[i],sample_list[i+2]))\n",
    "}\n",
    "\n",
    "#Fig5 Oligo \n",
    "\n",
    "cell_meta <- subset(starmap_raw@meta.data, cell_type %in% c(\"Oligo1\",\"Oligo2\",\"Oligo3\",\"OPC\") & region == \"Alveus\")\n",
    "cell_meta$group <- paste(cell_meta$group,cell_meta$time,sep = \"_\")\n",
    "tmp_df <- cell_meta %>% group_by(group,cell_type) %>% summarise(count = n())\n",
    "\n",
    "tmp_vec <- c(\"control_13months\" = sum(starmap_raw@misc$ADmouse_9498_morph$region == 3) +\n",
    "               sum(starmap_raw@misc$ADmouse_11351_morph$region == 3),\n",
    "             \"control_8months\" = sum(starmap_raw@misc$ADmouse_9735_morph$region == 3) +\n",
    "               sum(starmap_raw@misc$ADmouse_9707_morph$region == 3),\n",
    "             \"disease_13months\" = sum(starmap_raw@misc$ADmouse_9494_morph$region == 3) +\n",
    "               sum(starmap_raw@misc$ADmouse_11346_morph$region == 3),\n",
    "             \"disease_8months\" = sum(starmap_raw@misc$ADmouse_9723_morph$region == 3) +\n",
    "               sum(starmap_raw@misc$ADmouse_9723_2_morph$region == 3))\n",
    "\n",
    "for(i in levels(as.factor(cell_meta$group))){\n",
    "  tmp_df$count[tmp_df$group == i] <- tmp_df$count[tmp_df$group == i]/tmp_vec[[i]]*1000000*3.3^2\n",
    "}\n",
    "\n",
    "tmp_df$group <- factor(tmp_df$group,levels = c(\"control_8months\", \"disease_8months\",\n",
    "                                               \"control_13months\", \"disease_13months\"))\n",
    "\n",
    "ggplot(data = tmp_df,aes(x= group, y=count, fill=cell_type))+\n",
    "  geom_bar(stat=\"identity\") + \n",
    "  labs(title = \"Oligo\") +\n",
    "  theme_classic() + \n",
    "  theme(axis.text.x = element_text(angle = 45,hjust = 1)) +\n",
    "  scale_fill_manual(values = cell_type_palette_list$Oligo)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fcbe3cb",
   "metadata": {
    "name": "Fig7 Ridge",
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "j = \"8mo\"\n",
    "if(j == \"8mo\"){\n",
    "  tmp_df  <- rbind(starmap_raw@misc$ADmouse_9723_plaque_all$ADmouse_9723_cell_meta,\n",
    "                 starmap_raw@misc$ADmouse_9723_2_plaque_all$ADmouse_9723_2_cell_meta)  \n",
    "}else if(j == \"13mo\"){\n",
    "  tmp_df  <- rbind(starmap_raw@misc$ADmouse_9494_plaque_all$ADmouse_9494_cell_meta,\n",
    "                 starmap_raw@misc$ADmouse_11346_plaque_all$ADmouse_11346_cell_meta)  \n",
    "}\n",
    "tmp_df <- rbind(tmp_df,tmp_df %>% mutate(interval = \"overall\"))\n",
    "tmp_df <- tmp_df %>%  group_by(cell_type,top_level,interval) %>%\n",
    "  summarise(time = n()) %>% subset((cell_type %in% c(\"Astro3\",\"Micro3\",\"Oligo2\",\"Oligo3\",\"OPC\") |\n",
    "                                      top_level %in% c(\"CTX-Ex\",\"CA1\",\"CA2\",\"CA3\",\"DG\",\"Inh\")) & \n",
    "                                      interval %in% c(\"10um\",\"20um\",\"30um\",\"40um\",\"overall\"))\n",
    "if(j == \"13mo\"){\n",
    "  tmp_vec <- c(starmap_raw@misc$ADmouse_9494_plaque_all$ADmouse_9494_plaque_dilate_area+\n",
    "                starmap_raw@misc$ADmouse_11346_plaque_all$ADmouse_11346_plaque_dilate_area,\n",
    "             \"overall\" = ad_area(dim(starmap_raw@misc$ADmouse_9494_plaque_all$ADmouse_9494_plaque_img))+\n",
    "               ad_area(dim(starmap_raw@misc$ADmouse_11346_plaque_all$ADmouse_11346_plaque_img))) \n",
    "}else if(j == \"8mo\"){\n",
    "  tmp_vec <- c(starmap_raw@misc$ADmouse_9723_plaque_all$ADmouse_9723_plaque_dilate_area+\n",
    "                starmap_raw@misc$ADmouse_9723_2_plaque_all$ADmouse_9723_2_plaque_dilate_area,\n",
    "             \"overall\" = ad_area(dim(starmap_raw@misc$ADmouse_9723_plaque_all$ADmouse_9723_plaque_img))+\n",
    "               ad_area(dim(starmap_raw@misc$ADmouse_9723_2_plaque_all$ADmouse_9723_2_plaque_img))) \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "tmp_df$top_level <- as.character(tmp_df$top_level)\n",
    "tmp_df$top_level[tmp_df$top_level%in% c(\"CTX-Ex\",\"CA1\",\"CA2\",\"CA3\",\"DG\",\"Inh\")] <- \"Neuron\"\n",
    "\n",
    "tmp_df <- tmp_df %>% group_by(top_level, interval) %>% summarise(dens = sum(time))\n",
    "\n",
    "for(i in 1:4){\n",
    "  tmp_df$dens[tmp_df$interval == paste0(i,\"0um\")] <- tmp_df$dens[tmp_df$interval == paste0(i,\"0um\")]/\n",
    "    tmp_vec[[paste0(i,\"0um\")]] * 3.3^2 * 1000000\n",
    "}\n",
    "tmp_df$dens[tmp_df$interval == \"overall\"] <- tmp_df$dens[tmp_df$interval == \"overall\"]/\n",
    "    tmp_vec[[\"overall\"]] * 3.3^2 * 1000000\n",
    "\n",
    "plot_list <- list()\n",
    "\n",
    "\n",
    "\n",
    "for(i in 1:5){\n",
    "  plot_list[[i]] <- ggplot(tmp_df[tmp_df$top_level == c(\"Micro\",\"Astro\",\"Oligo\",\"OPC\",\"Neuron\")[i],], \n",
    "                           aes(x = interval, y=dens)) + # the fill argument sets the color for the bars\n",
    " geom_col(show.legend = F,fill = c(\"#fcb900\",\"#29e322\",\"#8600d4\",\"#c800cc\",\"#aaaaaa\")[i]) +\n",
    "    labs(title = c(\"DAM\",\"DAA\",\"Oligo2/3\",\"OPC\",\"Neuron\")[i])+\n",
    "    ylim(0,c(1300,150,130,100,1600)[i])+\n",
    "    labs(title = paste(j,c(\"Micro\",\"Astro\",\"Oligo\",\"OPC\",\"Neuron\")[i])) + \n",
    " #scale_fill_manual(values = c(\"#fcb900\",\"#29e322\",\"#8600d4\",\"#c800cc\",\"#aaaaaa\")) +\n",
    " theme_classic() \n",
    "}\n",
    "\n",
    "pdf(file = paste0(\"Sep06_Fig7_cell_type_density_\",j,\".pdf\"), width = 20,height = 10)\n",
    "ggarrange(plotlist = plot_list,ncol = 3, nrow = 2) %>% print()\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3db61105",
   "metadata": {},
   "source": [
    "## Grid Analysis\n",
    "\n",
    "### Func Def"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd44f756",
   "metadata": {
    "name": "grid analysis function",
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "ad_area <- function(dims_array){dims_array[1]*dims_array[2]}\n",
    "\n",
    "ad_tau_grid_analysis <- function(tau_img,cell_meta,sample_name, \n",
    "                                 block_size, palette_list, level = \"top_level\", \n",
    "                                 plaque_img = NULL, region_img = NULL){\n",
    "  plot_list <- list()\n",
    "  tmp_img <- Image(tau_img!=0,colormode = Grayscale)\n",
    "  tmp_mat <- matrix(0,nrow = floor(dim(tmp_img)[1] / block_size),\n",
    "                    ncol = floor(dim(tmp_img)[2] / block_size))\n",
    "  for(i in 1:dim(tmp_mat)[1]){\n",
    "    for(j in 1:dim(tmp_mat)[2]){\n",
    "      tmp_mat[i,j] <- sum(tmp_img[((i-1)*block_size+1):(i*block_size),((j-1)*block_size+1):(j*block_size)])\n",
    "    }\n",
    "  }\n",
    "  require(ComplexHeatmap)\n",
    "  require(circlize)\n",
    "  plot_list[[\"tau_heatmap\"]] <- Heatmap(tmp_mat,cluster_rows = F,cluster_columns = F,\n",
    "                                        col =  colorRamp2(c(0, quantile(tmp_mat,probs = c(0.99))), \n",
    "                                                          c(\"white\", \"darkred\")))\n",
    "  #Cell Idty Stat\n",
    "  cell_stat <- data.frame(x = rep(c(1:dim(tmp_mat)[1]), dim(tmp_mat)[2]),\n",
    "                          y = rep(c(1:dim(tmp_mat)[2]), each = dim(tmp_mat)[1]))\n",
    "  cell_stat[[\"tau_level\"]] <- apply(cell_stat,1,FUN = function(x){tmp_mat[x[[\"x\"]], x[[\"y\"]] ]})\n",
    "  for(i in names(palette_list)){\n",
    "    cell_stat[[i]] <- apply(cell_stat,1,\n",
    "                            FUN = function(x){\n",
    "                              sum(cell_meta$x >= (x[[\"x\"]]-1)*block_size+1 &\n",
    "                                    cell_meta$x <= x[[\"x\"]]*block_size &\n",
    "                                    cell_meta$y >= (x[[\"y\"]]-1)*block_size+1 &\n",
    "                                    cell_meta$y <= x[[\"y\"]]*block_size &\n",
    "                                    cell_meta[[level]] == i)\n",
    "                              })\n",
    "    #print(i)\n",
    "  }\n",
    "  if(!is.null(region_img)){\n",
    "    cell_stat[[\"region\"]] <- apply(cell_stat,1,\n",
    "                            FUN = function(x){\n",
    "                              region_img[round((x[[\"x\"]]-1/2)*block_size), round((x[[\"y\"]]-1/2)*block_size)]\n",
    "                              })\n",
    "  }\n",
    "  require(reshape2)\n",
    "  if(!is.null(plaque_img)){\n",
    "    cell_stat[[\"plaque\"]] <- apply(cell_stat,1,\n",
    "                                   FUN = function(x){\n",
    "                                     return(sum(plaque_img[((x[[\"x\"]]-1)*block_size+1):(x[[\"x\"]]*block_size),\n",
    "                                                        ((x[[\"y\"]]-1)*block_size+1):(x[[\"y\"]]*block_size)]))\n",
    "                                   })\n",
    "    tmp_df <- data.frame(tau_interval = c(\"0%\",\"50%\",\"100%\",\"100%\",\"100%\"),\n",
    "                       min_tau = c(0,quantile(cell_stat$tau_level[cell_stat$tau_level != 0])[c(1,3,3,3)]),\n",
    "                       max_tau = c(quantile(cell_stat$tau_level[cell_stat$tau_level != 0])[c(1,3)],rep(block_size^2,3)),\n",
    "                       plaque = c(\"any\",\"any\",\"any\",\"with_plaque\",\"without_plaque\"))\n",
    "    for(i in 1:5){\n",
    "      for(j in names(palette_list)){\n",
    "        if(tmp_df$plaque[i] == \"any\"){\n",
    "          tmp_df[[j]][i] <- sum(cell_stat[[j]][cell_stat$tau_level >= tmp_df$min_tau[i] &\n",
    "                                               cell_stat$tau_level < tmp_df$max_tau[i]])/\n",
    "          sum(cell_stat$tau_level >= tmp_df$min_tau[i] & cell_stat$tau_level < tmp_df$max_tau[i])\n",
    "        }else if(tmp_df$plaque[i] == \"with_plaque\"){\n",
    "          tmp_df[[j]][i] <- sum(cell_stat[[j]][cell_stat$tau_level >= tmp_df$min_tau[i] &\n",
    "                                               cell_stat$tau_level < tmp_df$max_tau[i] & cell_stat$plaque > 0])/\n",
    "          sum(cell_stat$tau_level >= tmp_df$min_tau[i] & cell_stat$tau_level < tmp_df$max_tau[i] & cell_stat$plaque > 0)\n",
    "        }else if(tmp_df$plaque[i] == \"without_plaque\"){\n",
    "          tmp_df[[j]][i] <- sum(cell_stat[[j]][cell_stat$tau_level >= tmp_df$min_tau[i] &\n",
    "                                               cell_stat$tau_level < tmp_df$max_tau[i] & cell_stat$plaque == 0])/\n",
    "          sum(cell_stat$tau_level >= tmp_df$min_tau[i] & cell_stat$tau_level < tmp_df$max_tau[i] & cell_stat$plaque == 0)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    tmp_df_bk <- tmp_df \n",
    "    tmp_df <- tmp_df %>% melt(id.vars = c(\"tau_interval\",\"min_tau\",\"max_tau\",\"plaque\"))\n",
    "    tmp_df$tau_interval <- factor(tmp_df$tau_interval, levels = c(\"0%\",\"50%\",\"100%\"))\n",
    "    tmp_df$variable <- factor(tmp_df$variable, levels = names(palette_list))\n",
    "    tmp_df[[\"title\"]] <- factor(paste(tmp_df$tau_interval,tmp_df$plaque,sep = \"_\"),\n",
    "                                levels = c(\"0%_any\",\"50%_any\",\"100%_any\",\"100%_with_plaque\",\"100%_without_plaque\"))\n",
    "    \n",
    "    plot_list[[\"stacked_barplot\"]] <-\n",
    "    ggplot(data = tmp_df,aes(x = title, y = value, fill = variable, order = variable))+\n",
    "      geom_bar(stat=\"identity\") + \n",
    "      ggtitle(paste(sample_name,\"Cell Density in Tau Free blocks and 0-25%/.../75-100% quantile blocks with tau\")) +\n",
    "      scale_fill_manual(\"\",values = palette_list,\n",
    "                        breaks = names(palette_list),\n",
    "                        labels = names(palette_list)) +\n",
    "      theme_classic()\n",
    "  \n",
    "  }else{\n",
    "    #Ridge Barplot\n",
    "  tmp_df <- data.frame(tau_interval = c(\"0%\",\"50%\",\"100%\"),\n",
    "                       #min_tau = c(0,1,10,29,68),\n",
    "                       #max_tau = c(1,10,29,68,999),\n",
    "                       min_tau = c(0,quantile(cell_stat$tau_level[cell_stat$tau_level != 0])[c(1,3)]),\n",
    "                       max_tau = c(quantile(cell_stat$tau_level[cell_stat$tau_level != 0])[c(1,3)],block_size^2))\n",
    "  \n",
    "  for(i in 1:3){\n",
    "    for(j in names(palette_list)){\n",
    "      tmp_df[[j]][i] <- sum(cell_stat[[j]][cell_stat$tau_level >= tmp_df$min_tau[i] &\n",
    "                                             cell_stat$tau_level < tmp_df$max_tau[i]])/\n",
    "      sum(cell_stat$tau_level >= tmp_df$min_tau[i] & cell_stat$tau_level < tmp_df$max_tau[i])\n",
    "    }\n",
    "    \n",
    "  }\n",
    "  tmp_df_bk <- tmp_df \n",
    "  tmp_df <- tmp_df %>% melt(id.vars = c(\"tau_interval\",\"min_tau\",\"max_tau\"))\n",
    "  tmp_df$tau_interval <- factor(tmp_df$tau_interval, levels = c(\"0%\",\"50%\",\"100%\"))\n",
    "  tmp_df$variable <- factor(tmp_df$variable, levels = names(palette_list))\n",
    "  \n",
    "  plot_list[[\"stacked_barplot\"]] <-\n",
    "    ggplot(data = tmp_df,aes(x = tau_interval, y = value, fill = variable, order = variable))+\n",
    "    geom_bar(stat=\"identity\") + \n",
    "    ggtitle(paste(sample_name,\"Cell Density in Tau Free blocks and 0-25%/.../75-100% quantile blocks with tau\")) +\n",
    "    scale_fill_manual(\"\",values = palette_list,\n",
    "                      breaks = names(palette_list),\n",
    "                      labels = names(palette_list)) +\n",
    "    theme_classic()\n",
    "  }\n",
    "  tmp_mat2 <- matrix(2,nrow = nrow(tmp_mat), ncol = ncol(tmp_mat))\n",
    "  tmp_mat2[tmp_mat < quantile(cell_stat$tau_level[cell_stat$tau_level != 0], probs = c(0.75))] <- 1\n",
    "  tmp_mat2[tmp_mat < quantile(cell_stat$tau_level[cell_stat$tau_level != 0], probs = c(0.25))] <- 0\n",
    "  plot_list[[\"tau_labeled\"]] <- Heatmap(tmp_mat2,cluster_rows = F,cluster_columns = F,\n",
    "                                        col =  colorRamp2(c(0,1,2), c(\"#1f95c5\", \"#ffd86c\",\"#9e0018\")), rect_gp = gpar(col = \"gray80\", lwd = 0.5))\n",
    "  cell_stat[[\"tau_group\"]] <- apply(cell_stat,1,\n",
    "                                    FUN = function(x){\n",
    "                                      tau_level <- as.numeric(x[[\"tau_level\"]])\n",
    "                                      plaque_level <- as.numeric(x[[\"plaque\"]])\n",
    "                                      \n",
    "                                      tmp_vec <- quantile(cell_stat$tau_level[cell_stat$tau_level != 0], \n",
    "                                                          probs = c(0,0.5,1))\n",
    "                                      if(tau_level < tmp_vec[1]) \"Zero\"\n",
    "                                      else if(tau_level < tmp_vec[2]) \"Low\"\n",
    "                                      else if(tau_level >= tmp_vec[2] & plaque_level == 0) \"High w/o ABeta\"\n",
    "                                      else if(tau_level >= tmp_vec[2] & plaque_level > 0) \"High w ABeta\"\n",
    "                                      else \"NA\"\n",
    "                                    })\n",
    "  #Add cell meta\n",
    "  cell_meta[[\"tau_block\"]] <- apply(cell_meta,1,\n",
    "                                    FUN = function(x){\n",
    "                                      if((floor(as.numeric(x[[\"x\"]])/block_size)+1) <= max(cell_stat$x) && \n",
    "                                         (floor(as.numeric(x[[\"y\"]])/block_size)+1) <= max(cell_stat$y)){\n",
    "                                        cell_stat$tau_group[cell_stat$x == (floor(as.numeric(x[[\"x\"]])/block_size)+1) & \n",
    "                                                            cell_stat$y == (floor(as.numeric(x[[\"y\"]])/block_size)+1)] %>% as.character()\n",
    "                                      }else{\"NA\"}\n",
    "                                    })\n",
    "    \n",
    "  return(list(\"plot_list\" = plot_list,\n",
    "              \"cell_stat\" = cell_stat,\n",
    "              \"cell_meta\" = cell_meta,\n",
    "              \"interval_stat\" = tmp_df,\n",
    "              \"interval_stat_wide\" = tmp_df_bk))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08963d48",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bb40397",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#####Run tau grid analysis#####\n",
    "cell_meta <- starmap_raw@meta.data %>% subset(subset = sample == \"ADmouse_9494\")\n",
    "sample_name <- \"AD_mouse9494\"\n",
    "block_size = round(20*3.175) #20um\n",
    "tau_img <- starmap_raw@misc[[\"ADmouse_9494_morph\"]][[\"tau\"]]\n",
    "plaque_img <- starmap_raw@misc$ADmouse_9494_plaque_all$ADmouse_9494_plaque_img\n",
    "#region_img <- starmap@misc$AD_mouse9494_morph$region\n",
    "palette_list <- top_level_palette_list\n",
    "#1-Alveus\n",
    "require(tiff)\n",
    "region_img <- readTIFF(source = \"data/2021-06-22-mAD-molecular-labels/2766\\ genes/AD_mouse9494-molecular-labels.tif\",\n",
    "                       indexed = T) %>% t()\n",
    "tau_res <- list()\n",
    "\n",
    "tau_res[[\"ADmouse_9494_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9494_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9494\"), \n",
    "                                   \"ADmouse_9494\", block_size, top_level_palette_list, \n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9494_plaque_all$ADmouse_9494_plaque_img)\n",
    "tau_res[[\"ADmouse_9723_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9723_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9723\"), \n",
    "                                   \"ADmouse_9723\", block_size, top_level_palette_list, \n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9723_plaque_all$ADmouse_9723_plaque_img)\n",
    "tau_res[[\"ADmouse_9723_2_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9723_2_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9723_2\"), \n",
    "                                   \"ADmouse_9723_2\", block_size, top_level_palette_list, \n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9723_2_plaque_all$ADmouse_9723_2_plaque_img)\n",
    "tau_res[[\"ADmouse_11346_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_11346_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_11346\"), \n",
    "                                   \"ADmouse_11346\", block_size, top_level_palette_list, \n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_11346_plaque_all$ADmouse_11346_plaque_img)\n",
    "\n",
    "#Export Heatmap\n",
    "pdf(file = \"Sep04_tau_labeled_heatmap.pdf\", width = 8, height = 6)\n",
    "draw(tau_res[[\"ADmouse_9723_res\"]][[\"plot_list\"]][[\"tau_labeled\"]])\n",
    "draw(tau_res[[\"ADmouse_9723_2_res\"]][[\"plot_list\"]][[\"tau_labeled\"]])\n",
    "draw(tau_res[[\"ADmouse_9494_res\"]][[\"plot_list\"]][[\"tau_labeled\"]])\n",
    "draw(tau_res[[\"ADmouse_11346_res\"]][[\"plot_list\"]][[\"tau_labeled\"]])\n",
    "dev.off()\n",
    "\n",
    "\n",
    "#Oligo/OPC subtype\n",
    "tau_res[[\"ADmouse_9494_Oligo_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9494_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9494\"), \n",
    "                                   \"ADmouse_9494\", block_size, cell_type_palette_list$Oligo, level = \"cell_type\",\n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9494_plaque_all$ADmouse_9494_plaque_img)\n",
    "tau_res[[\"ADmouse_9723_Oligo_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9723_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9723\"), \n",
    "                                   \"ADmouse_9723\", block_size, cell_type_palette_list$Oligo, level = \"cell_type\",\n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9723_plaque_all$ADmouse_9723_plaque_img)\n",
    "tau_res[[\"ADmouse_9723_2_Oligo_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9723_2_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9723_2\"), \n",
    "                                   \"ADmouse_9723_2\", block_size, cell_type_palette_list$Oligo, level = \"cell_type\",\n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9723_2_plaque_all$ADmouse_9723_2_plaque_img)\n",
    "tau_res[[\"ADmouse_11346_Oligo_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_11346_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_11346\"), \n",
    "                                   \"ADmouse_11346\", block_size, cell_type_palette_list$Oligo, level = \"cell_type\",\n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_11346_plaque_all$ADmouse_11346_plaque_img)\n",
    "#Astro subtype\n",
    "tau_res[[\"ADmouse_9494_Astro_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9494_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9494\"), \n",
    "                                   \"ADmouse_9494\", block_size, cell_type_palette_list$Astro, level = \"cell_type\",\n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9494_plaque_all$ADmouse_9494_plaque_img)\n",
    "tau_res[[\"ADmouse_9723_Astro_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9723_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9723\"), \n",
    "                                   \"ADmouse_9723\", block_size, cell_type_palette_list$Astro, level = \"cell_type\",\n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9723_plaque_all$ADmouse_9723_plaque_img)\n",
    "tau_res[[\"ADmouse_9723_2_Astro_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_9723_2_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_9723_2\"), \n",
    "                                   \"ADmouse_9723_2\", block_size, cell_type_palette_list$Astro, level = \"cell_type\",\n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_9723_2_plaque_all$ADmouse_9723_2_plaque_img)\n",
    "tau_res[[\"ADmouse_11346_Astro_res\"]] <- ad_tau_grid_analysis(starmap_raw@misc[[\"ADmouse_11346_morph\"]][[\"tau\"]],\n",
    "                                   subset(starmap_raw@meta.data, subset = sample == \"ADmouse_11346\"), \n",
    "                                   \"ADmouse_11346\", block_size, cell_type_palette_list$Astro, level = \"cell_type\",\n",
    "                                   plaque_img = starmap_raw@misc$ADmouse_11346_plaque_all$ADmouse_11346_plaque_img)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ce003b3",
   "metadata": {},
   "source": [
    "### Stat & Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce7fe55e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# Export Excel for tau analysis\n",
    "wb = createWorkbook()\n",
    "#top level\n",
    "sheet <- createSheet(wb, \"p-Tau\")\n",
    "\n",
    "#8mo\n",
    "tmp_df <- (tau_res[[\"AD9723_res\"]][[\"interval_stat_wide\"]][,5:16] + \n",
    "  tau_res[[\"AD9723_2_res\"]][[\"interval_stat_wide\"]][,5:16])/2 %>% as.matrix()\n",
    "tmp_df <- t(tmp_df)\n",
    "colnames(tmp_df) <- c(\"No Tau\",\"Low Tau\",\"High Tau\",\"High Tau w/ Aß\",\"High Tau w/o Aß\")\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=1, row.names=T)\n",
    "#13mo\n",
    "tmp_df <- (tau_res[[\"AD9494_res\"]][[\"interval_stat_wide\"]][,5:16] + \n",
    "  tau_res[[\"AD11346_res\"]][[\"interval_stat_wide\"]][,5:16])/2 %>% as.matrix()\n",
    "tmp_df <- t(tmp_df)\n",
    "colnames(tmp_df) <- c(\"No Tau\",\"Low Tau\",\"High Tau\",\"High Tau w/ Aß\",\"High Tau w/o Aß\")\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=8, startRow=1, row.names=T)\n",
    "#8mo Oligo\n",
    "tmp_df <- (tau_res[[\"AD9723_Oligo_res\"]][[\"interval_stat_wide\"]][,5:8] + \n",
    "  tau_res[[\"AD9723_2_Oligo_res\"]][[\"interval_stat_wide\"]][,5:8])/2 %>% as.matrix()\n",
    "tmp_df <- t(tmp_df)\n",
    "colnames(tmp_df) <- c(\"No Tau\",\"Low Tau\",\"High Tau\",\"High Tau w/ Aß\",\"High Tau w/o Aß\")\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=15, row.names=T)\n",
    "#13mo Oligo\n",
    "tmp_df <- (tau_res[[\"AD9494_Oligo_res\"]][[\"interval_stat_wide\"]][,5:8] + \n",
    "  tau_res[[\"AD11346_Oligo_res\"]][[\"interval_stat_wide\"]][,5:8])/2 %>% as.matrix()\n",
    "tmp_df <- t(tmp_df)\n",
    "colnames(tmp_df) <- c(\"No Tau\",\"Low Tau\",\"High Tau\",\"High Tau w/ Aß\",\"High Tau w/o Aß\")\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=8, startRow=15, row.names=T)\n",
    "#8mo Astro\n",
    "tmp_df <- (tau_res[[\"AD9723_Astro_res\"]][[\"interval_stat_wide\"]][,5:7] + \n",
    "  tau_res[[\"AD9723_2_Astro_res\"]][[\"interval_stat_wide\"]][,5:7])/2 %>% as.matrix()\n",
    "tmp_df <- t(tmp_df)\n",
    "colnames(tmp_df) <- c(\"No Tau\",\"Low Tau\",\"High Tau\",\"High Tau w/ Aß\",\"High Tau w/o Aß\")\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=15, row.names=T)\n",
    "#13mo Oligo\n",
    "tmp_df <- (tau_res[[\"AD9494_Astro_res\"]][[\"interval_stat_wide\"]][,5:7] + \n",
    "  tau_res[[\"AD11346_Astro_res\"]][[\"interval_stat_wide\"]][,5:7])/2 %>% as.matrix()\n",
    "tmp_df <- t(tmp_df)\n",
    "colnames(tmp_df) <- c(\"No Tau\",\"Low Tau\",\"High Tau\",\"High Tau w/ Aß\",\"High Tau w/o Aß\")\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=8, startRow=15, row.names=T)\n",
    "\n",
    "# Stat test for tau\n",
    "\n",
    "\n",
    "ad_tau_stat <- function(cell_stat, cell_type_vec){\n",
    "  tmp_df <- data.frame(cell_type = cell_type_vec)\n",
    "  for(i in c(\"Low\",\"High\",\"High w ABeta\",\"High w/o ABeta\")){\n",
    "  tmp_df[[i]] <- apply(tmp_df,1,\n",
    "                       FUN = function(x){\n",
    "                         if(i == \"High\") tmp_vec <- cell_stat[[x[[\"cell_type\"]]]][cell_stat$tau_group %in% c(\"High w ABeta\", \"High w/o ABeta\")]\n",
    "                         else tmp_vec <- cell_stat[[x[[\"cell_type\"]]]][cell_stat$tau_group == i]\n",
    "                         tmp_obj <- t.test(x = tmp_vec,\n",
    "                                           y = cell_stat[[x[[\"cell_type\"]]]][cell_stat$tau_group == \"Zero\"],\n",
    "                                           alternative = \"greater\")\n",
    "                         tmp_obj$p.value\n",
    "                       })\n",
    "  }\n",
    "  tmp_df\n",
    "}\n",
    "\n",
    "#top level\n",
    "sheet <- createSheet(wb, \"p-Tau-stat\")\n",
    "tmp_vec <- c('CA1','CA2','CA3','CTX-Ex','DG','Endo','Inh','Micro','Astro','OPC','Oligo','SMC')\n",
    "#8mo\n",
    "tmp_df <- ad_tau_stat(rbind(tau_res[[\"AD9723_res\"]][[\"cell_stat\"]],tau_res[[\"AD9723_2_res\"]][[\"cell_stat\"]]),tmp_vec)\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=1, row.names=T)\n",
    "#13mo\n",
    "tmp_df <- ad_tau_stat(rbind(tau_res[[\"AD9494_res\"]][[\"cell_stat\"]],tau_res[[\"AD11346_res\"]][[\"cell_stat\"]]),tmp_vec)\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=7, startRow=1, row.names=T)\n",
    "\n",
    "tmp_vec <- c('Oligo1','Oligo2','Oligo3','OPC')\n",
    "#8mo Oligo\n",
    "tmp_df <- ad_tau_stat(rbind(tau_res[[\"AD9723_Oligo_res\"]][[\"cell_stat\"]],\n",
    "                            tau_res[[\"AD9723_2_Oligo_res\"]][[\"cell_stat\"]]),tmp_vec)\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=15, row.names=T)\n",
    "#13mo Oligo\n",
    "tmp_df <- ad_tau_stat(rbind(tau_res[[\"AD9494_Oligo_res\"]][[\"cell_stat\"]],\n",
    "                            tau_res[[\"AD11346_Oligo_res\"]][[\"cell_stat\"]]),tmp_vec)\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=7, startRow=15, row.names=T)\n",
    "\n",
    "tmp_vec <- c('Astro1','Astro2','Astro3')\n",
    "#8mo Oligo\n",
    "tmp_df <- ad_tau_stat(rbind(tau_res[[\"AD9723_Astro_res\"]][[\"cell_stat\"]],\n",
    "                            tau_res[[\"AD9723_2_Astro_res\"]][[\"cell_stat\"]]),tmp_vec)\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=15, row.names=T)\n",
    "#13mo Astro\n",
    "tmp_df <- ad_tau_stat(rbind(tau_res[[\"AD9494_Astro_res\"]][[\"cell_stat\"]],\n",
    "                            tau_res[[\"AD11346_Astro_res\"]][[\"cell_stat\"]]),tmp_vec)\n",
    "addDataFrame(tmp_df, sheet=sheet, startColumn=7, startRow=15, row.names=T)\n",
    "saveWorkbook(wb, \"Jun17_p-Tau_cell_composition_stat.xlsx\")\n",
    "#Combined\n",
    "plot_list <- list()\n",
    "palette_list <- cell_type_palette_list$Oligo\n",
    "palette_list <- top_level_palette_list\n",
    "\n",
    "\n",
    "ad_combined_tau_density <- function(palette_list,stat_1,stat_2,title_text = \"\"){\n",
    "  tmp_df <- data.frame(cell_type = stat_1$variable,\n",
    "                     title = stat_1$title,\n",
    "                     sample1 = stat_1$value,\n",
    "                     sample2 = stat_2$value)\n",
    "  tmp_df$average <- (tmp_df$sample1+tmp_df$sample2)/2\n",
    "  tmp_df <- tmp_df %>% subset(subset = cell_type != \"DG\")\n",
    "  palette_list <- palette_list[names(palette_list) != \"DG\"]\n",
    "  tmp_df$title <- factor(tmp_df$title,levels = c(\"0%_any\",\"50%_any\",\"100%_any\",\"100%_with_plaque\",\"100%_without_plaque\"))\n",
    "  p <-\n",
    "    ggplot(data = tmp_df,aes(x = title, y = average, fill = cell_type, order = cell_type))+\n",
    "    geom_bar(stat=\"identity\") + \n",
    "    ggtitle(paste(\"p-Tau Cell Density\",title_text)) +\n",
    "    scale_fill_manual(\"\",values = palette_list ,\n",
    "                      breaks = names(palette_list),\n",
    "                      labels = names(palette_list)) +\n",
    "    theme_classic()\n",
    "}\n",
    "\n",
    "plot_list[[\"8mo_overall\"]] <- ad_combined_tau_density(top_level_palette_list,\n",
    "                                                       tau_res$AD9723_res$interval_stat,\n",
    "                                                       tau_res$AD9723_2_res$interval_stat,\"8mo overall\")\n",
    "plot_list[[\"13mo_overall\"]] <- ad_combined_tau_density(top_level_palette_list,\n",
    "                                                       tau_res$AD9494_res$interval_stat,\n",
    "                                                       tau_res$AD11346_res$interval_stat,\"13mo overall\")\n",
    "plot_list[[\"8mo_Oligo\"]] <- ad_combined_tau_density(cell_type_palette_list$Oligo,\n",
    "                                                       tau_res$AD9723_Oligo_res$interval_stat,\n",
    "                                                       tau_res$AD9723_2_Oligo_res$interval_stat,\"8mo Oligo\")\n",
    "plot_list[[\"13mo_Oligo\"]] <- ad_combined_tau_density(cell_type_palette_list$Oligo,\n",
    "                                                       tau_res$AD9494_Oligo_res$interval_stat,\n",
    "                                                       tau_res$AD11346_Oligo_res$interval_stat,\"13mo Oligo\")\n",
    "plot_list[[\"8mo_Astro\"]] <- ad_combined_tau_density(cell_type_palette_list$Astro,\n",
    "                                                       tau_res$AD9723_Astro_res$interval_stat,\n",
    "                                                       tau_res$AD9723_2_Astro_res$interval_stat,\"8mo Astro\")\n",
    "plot_list[[\"13mo_Astro\"]] <- ad_combined_tau_density(cell_type_palette_list$Astro,\n",
    "                                                       tau_res$AD9494_Astro_res$interval_stat,\n",
    "                                                       tau_res$AD11346_Astro_res$interval_stat,\"13mo Astro\")\n",
    "\n",
    "pdf(file = \"Jun03_Tau_stacked_barplot.pdf\",width = 16,height = 12,useDingbats = FALSE)\n",
    "ggarrange(plotlist = plot_list,ncol = 2, nrow = 2)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "#Tau Distribution Around hippocampus area\n",
    "tau_list <- list()\n",
    "\n",
    "for(sample_name in c('ADmouse_9723','ADmouse_9494','ADmouse_9723_2','ADmouse_11346')){\n",
    "  for(area_name in c(\"Cortex\",\"Hippocampus\")){\n",
    "    message(paste(sample_name,area_name))\n",
    "    tau_img <- Image(starmap_raw@misc[[paste0(sample_name,\"_morph\")]][[\"tau\"]],colormode = Grayscale)\n",
    "    tau_img <- tau_img > 0 #0/1\n",
    "    tmp_list <- list()\n",
    "    plaque_img <- starmap_raw@misc[[paste0(sample_name,\"_plaque_\",area_name)]][[paste0(sample_name,\"_plaque_img\")]]\n",
    "    dilate_img <- plaque_img > 0 \n",
    "    dilate_img_bk <- dilate_img\n",
    "    if(area_name == \"Cortex\") region_img <- starmap_raw@misc[[paste0(sample_name,\"_morph\")]]$region == 1\n",
    "    else if(area_name == \"Hippocampus\") region_img <- starmap_raw@misc[[paste0(sample_name,\"_morph\")]]$region == 3\n",
    "\n",
    "    plaque_vec <- starmap_raw@misc[[paste0(sample_name,\"_plaque_\",area_name)]][[paste0(sample_name,\"_plaque_dilate_area\")]]\n",
    "    voronoi_img <- propagate(seeds = plaque_img, x = Image(dim = dim(plaque_img)))\n",
    "    for(i in 1:5){\n",
    "      dilate_img_bk <- dilate_img\n",
    "      dilate_img <- dilate(plaque_img > 0 , kern = makeBrush(round(3.175*20*i), shape = \"disc\"))\n",
    "      dilate_img <- dilate_img * region_img\n",
    "      if(i != 1) dilate_img2 <- dilate_img * (1 - dilate_img_bk) else dilate_img2 <- dilate_img\n",
    "      dilate_img2 <- (dilate_img2 & tau_img) * voronoi_img\n",
    "      tmp_list[[paste0(\"dilate_\",i)]] <- table(dilate_img2)\n",
    "    }\n",
    "\n",
    "    tau_list[[paste(sample_name,area_name,sep = \"_\")]] <- tmp_list\n",
    "  }\n",
    "}\n",
    "# Export Violin plot & stat\n",
    "for(sample_name in c('ADmouse_9723','ADmouse_9494','ADmouse_9723_2','ADmouse_11346')){\n",
    "  for(area_name in c(\"Cortex\",\"Hippocampus\")){\n",
    "    tmp_list <- starmap_raw@misc[[paste0(sample_name,\"_plaque_\",area_name)]][[paste0(sample_name,\"_plaque_dilate_info\")]]\n",
    "    tmp_df <- data.frame(plaque_id = tmp_list$`10um_dilate_info`$plaque_id)\n",
    "    for(i in seq_len(5)){\n",
    "      tmp_df[[paste0(\"dilate_\",i,\"0um\")]] <- apply(tmp_df,1,\n",
    "                                                   FUN = function(x){\n",
    "         if(as.numeric(x[[\"plaque_id\"]]) %in%tmp_list[[paste0(i,\"0um_dilate_info\")]]$plaque_id){\n",
    "           tmp_list[[paste0(i,\"0um_dilate_info\")]]$s.area[tmp_list[[paste0(i,\"0um_dilate_info\")]]$plaque_id == as.numeric(x[[\"plaque_id\"]])]}else 0\n",
    "                                                   })\n",
    "      }\n",
    "    for(i in seq_len(5)){tmp_df[[paste0(\"tau_\",i,\"0um\")]] <- apply(tmp_df,1,\n",
    "                                                   FUN = function(x){\n",
    "         if(as.character(x[[\"plaque_id\"]]) %in% \n",
    "            names(tau_list[[paste(sample_name,area_name,sep = \"_\")]][[paste0(\"dilate_\",i)]])){\n",
    "           tau_list[[paste(sample_name,area_name,sep = \"_\")]][[paste0(\"dilate_\",i)]][[as.character(x[[\"plaque_id\"]])]]\n",
    "                                                        }else 0\n",
    "                                                   })\n",
    "    }\n",
    "    tmp_df[,7:11] <- tmp_df[,7:11]/tmp_df[,2:6]\n",
    "    tau_list[[paste(sample_name,area_name,\"df\",sep = \"_\")]] <- tmp_df\n",
    "  }\n",
    "}\n",
    "# Violin Plot \n",
    "plot_list <- list()\n",
    "\n",
    "sample_list <- c('ADmouse_9723','ADmouse_9494','ADmouse_9723_2','ADmouse_11346')\n",
    "for(i in 1:2){\n",
    "  for(area_name in c(\"Cortex\",\"Hippocampus\")){\n",
    "    plot_list[[paste(sep = \"_\",i,area_name)]] <- ggplot(data = rbind(tau_list[[paste(sample_list[i],area_name,\"df\",sep = \"_\")]], tau_list[[paste(sample_list[i+2],area_name,\"df\",sep = \"_\")]]) %>% \n",
    "                                                          .[,c(1,7:11)] %>% melt(id.vars = \"plaque_id\"),\n",
    "                         aes(x=variable, y=value)) +\n",
    "    theme_classic() +\n",
    "    geom_jitter(alpha = 0.5, color = \"darkred\") +\n",
    "\n",
    "\n",
    "    labs(title = paste(\"Tau\",switch (i,\"1\" = \"8mo\", \"2\" = \"13mo\"),area_name))\n",
    "  }\n",
    "}\n",
    "# One sample t-test\n",
    "i = 2\n",
    "area_name  = \"Cortex\"\n",
    "tmp_df <- rbind(tau_list[[paste(sample_list[i],area_name,\"df\",sep = \"_\")]], tau_list[[paste(sample_list[i+2],area_name,\"df\",sep = \"_\")]]) %>% .[,c(1,7:11)]\n",
    "tmp <- 0\n",
    "for(j in c(i,i+2)){\n",
    "  if(area_name == \"Cortex\"){\n",
    "    tmp_img <- (Image(starmap_raw@misc[[paste0(sample_list[j],\"_morph\")]][[\"tau\"]],colormode = Grayscale) > 0) *\n",
    "  (Image(starmap_raw@misc[[paste0(sample_list[j],\"_morph\")]][[\"region\"]],colormode = Grayscale) == 1)\n",
    "    tmp <- tmp + sum(tmp_img)/\n",
    "      sum(Image(starmap_raw@misc[[paste0(sample_list[j],\"_morph\")]][[\"region\"]],colormode = Grayscale) == 1)\n",
    "  }else if(area_name == \"Hippocampus\"){\n",
    "    tmp_img <- (Image(starmap_raw@misc[[paste0(sample_list[j],\"_morph\")]][[\"tau\"]],colormode = Grayscale) > 0) *\n",
    "  (Image(starmap_raw@misc[[paste0(sample_list[j],\"_morph\")]][[\"region\"]],colormode = Grayscale) == 3)\n",
    "    tmp <- tmp + sum(tmp_img)/\n",
    "      sum(Image(starmap_raw@misc[[paste0(sample_list[j],\"_morph\")]][[\"region\"]],colormode = Grayscale) == 3)\n",
    "  }\n",
    "}\n",
    "tmp <- tmp/2 #Average Percentge\n",
    "for(k in 1:5){\n",
    "  tmp_obj <- t.test(x = tmp_df[[paste0(\"tau_\",k,\"0um\")]], mu = tmp, alternative = \"greater\")\n",
    "  message(paste(area_name,ifelse(i == 1,\"8mo\",\"13mo\"),paste0(\"tau_\",k,\"0um\"),tmp_obj$p.value))\n",
    "}\n",
    "\n",
    "plot_list[[paste0(area_name,ifelse(i == 1,\"8mo\",\"13mo\"))]] <- ggplot(data = data.frame(group = c(colnames(tmp_df)[2:6],\"tau_overall\"),\n",
    "                         value = c(colMeans(tmp_df[,2:6],na.rm = T),overall = tmp)),\n",
    "       aes(x = group, y = value))+\n",
    "  geom_bar(fill = \"darkred\",stat = \"identity\") + \n",
    "  theme_classic() + labs(title = paste(area_name,ifelse(i == 1,\"8mo\",\"13mo\")))\n",
    "\n",
    "\n",
    "# Archived\n",
    "pdf(file = \"Jun01_tau_distr.pdf\",width = 8,height = 8,useDingbats = FALSE)\n",
    "plot_list %>% print()\n",
    "dev.off()\n",
    "#\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bff924f",
   "metadata": {},
   "source": [
    "## Pseudotime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b0ed995",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#External Embedding\n",
    "embed_ls <- list()\n",
    "embed_ls[[\"Oligo_diff\"]]<-read.csv(\"data/2022-02-27-Hu-AD-subclustering-embedding/embedding_Oligo_OPC_diffmap.csv\",header = F)\n",
    "embed_ls[[\"Oligo_umap\"]]<-read.csv(\"data/2022-02-27-Hu-AD-subclustering-embedding/embedding_Oligo_OPC_umap.csv\",header = F)\n",
    "embed_ls[[\"Astro_diff\"]]<-read.csv(\"data/2022-02-27-Hu-AD-subclustering-embedding/embedding_Astro_diffmap.csv\",header = F)\n",
    "embed_ls[[\"Astro_umap\"]]<-read.csv(\"data/2022-02-27-Hu-AD-subclustering-embedding/embedding_Astro_umap.csv\",header = F)\n",
    "embed_ls[[\"Micro_diff\"]]<-read.csv(\"data/2022-02-27-Hu-AD-subclustering-embedding/embedding_Micro_diffmap.csv\",header = F)\n",
    "embed_ls[[\"Micro_umap\"]]<-read.csv(\"data/2022-02-27-Hu-AD-subclustering-embedding/embedding_Micro_umap.csv\",header = F)\n",
    "\n",
    "#\n",
    "psdtime_res_list <- list()\n",
    "\n",
    "psdtime_res_list[[\"Micro\"]] <- psdtime_analysis(starmap_raw,\"Micro\", embed_df = as.data.frame(embed_ls[[\"Micro_diff\"]][,2:3]*100),\n",
    "                                                min_branch_len = 8, ncenter = 30, eps = 0)\n",
    "psdtime_res_list[[\"Astro\"]] <- psdtime_analysis(starmap_raw,\"Astro\", embed_df = as.data.frame(embed_ls[[\"Astro_diff\"]][,2:3]*100),\n",
    "                                                min_branch_len = 10, ncenter = 40, eps = 0)\n",
    "psdtime_res_list[[\"Oligo\"]] <- psdtime_analysis(starmap_raw,c(\"Oligo\",\"OPC\"), embed_df = as.data.frame(embed_ls[[\"Oligo_diff\"]][,2:3]*100),\n",
    "                                                min_branch_len = 10, ncenter = 80, eps = 0)\n",
    "\n",
    "psdtime_res_list[[\"DG\"]] <- psdtime_analysis(starmap_raw,\"DG\", embed_df = NULL,\n",
    "                                                min_branch_len = 8, ncenter = 40, eps = 0.5)\n",
    "\n",
    "#Export\n",
    "\n",
    "pdf(\"Mar15_Monocle.pdf\",width = 16,height = 4)\n",
    "ggarrange(plotlist = psdtime_res_list[[\"Micro\"]]$Plot_list, ncol = 3, nrow = 1)\n",
    "ggarrange(plotlist = psdtime_res_list[[\"Astro\"]]$Plot_list, ncol = 3, nrow = 1)\n",
    "ggarrange(plotlist = psdtime_res_list[[\"Oligo\"]]$Plot_list, ncol = 3, nrow = 1)\n",
    "dev.off()\n",
    "\n",
    "#Draw Figures\n",
    "source(\"~/Desktop/At_Broad/AD/R_pub_code/func/AD_Pseudo-timeFunction.R\")\n",
    "\n",
    "sample_convert <- function(cell_meta_df){\n",
    "  cell_meta_df$sample <- apply(cell_meta_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(x[[\"sample\"]] %in% c('ADmouse_9723','ADmouse_9723_2'))\n",
    "                              \"8mo_Disease\"\n",
    "                            else if(x[[\"sample\"]] %in% c('ADmouse_11346','ADmouse_9494'))\n",
    "                              \"13mo_Disease\" \n",
    "                            else if(x[[\"sample\"]] %in% c('ADmouse_9735','ADmouse_9707'))\n",
    "                              \"8mo_Control\" \n",
    "                            else if(x[[\"sample\"]] %in% c('ADmouse_9498','ADmouse_11351'))\n",
    "                              \"13mo_Control\" \n",
    "                              })\n",
    "  cell_meta_df$sample <- as.factor(cell_meta_df$sample)\n",
    "  return(cell_meta_df)\n",
    "}\n",
    "\n",
    "setwd(\"/Users/zhouhaowen/Desktop/At_Broad/AD\")\n",
    "pdf(file = \"Aug31_F3_Micro_Pseudotime.pdf\",width = 4,height = 3,useDingbats = FALSE)\n",
    "cell_meta <- sample_convert(psdtime_res_list[[\"Micro\"]][[\"cell_meta\"]])\n",
    "ad_psdtime_highlight(cell_meta, \"sample\",    psdtime_res_list[[\"Micro\"]][[\"cds\"]], T, \"Micro\") %>% print()\n",
    "ad_psdtime_highlight(cell_meta, \"cell_type\", psdtime_res_list[[\"Micro\"]][[\"cds\"]], T, \"Micro\") %>% print()\n",
    "ad_psdtime_gene_highlight(psdtime_res_list[[\"Micro\"]][[\"cell_meta\"]], \n",
    "                          psdtime_res_list[[\"Micro\"]][[\"cds\"]], c(\"Cst7\",\"Ctsb\",\"Trem2\",\"P2ry12\")) %>% print()\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"Aug31_F4_Astro_Pseudotime.pdf\",width = 4,height = 3,useDingbats = FALSE)\n",
    "cell_meta <- sample_convert(psdtime_res_list[[\"Astro\"]][[\"cell_meta\"]])\n",
    "ad_psdtime_highlight(cell_meta, \"sample\",    psdtime_res_list[[\"Astro\"]][[\"cds\"]], T, \"Astro\") %>% print()\n",
    "ad_psdtime_highlight(cell_meta, \"cell_type\", psdtime_res_list[[\"Astro\"]][[\"cds\"]], T, \"Astro\") %>% print()\n",
    "ad_psdtime_gene_highlight(psdtime_res_list[[\"Astro\"]][[\"cell_meta\"]], \n",
    "                          psdtime_res_list[[\"Astro\"]][[\"cds\"]], c(\"Gfap\",\"Cst3\",\"Vim\",\"Aqp4\")) %>% print()\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"Aug31_F5_Oligo_Pseudotime.pdf\",width = 4,height = 3,useDingbats = FALSE)\n",
    "cell_meta <- sample_convert(psdtime_res_list[[\"Oligo\"]][[\"cell_meta\"]])\n",
    "ad_psdtime_highlight(cell_meta, \"sample\",    psdtime_res_list[[\"Oligo\"]][[\"cds\"]], T, \"Oligo\") %>% print()\n",
    "ad_psdtime_highlight(cell_meta, \"cell_type\", psdtime_res_list[[\"Oligo\"]][[\"cds\"]], T, \"Oligo\") %>% print()\n",
    "ad_psdtime_gene_highlight(psdtime_res_list[[\"Oligo\"]][[\"cell_meta\"]], \n",
    "                          psdtime_res_list[[\"Oligo\"]][[\"cds\"]], c(\"Plp1\",\"Cldn11\",\"Klk6\",\"C4b\")) %>% print()\n",
    "dev.off()\n",
    "\n",
    "# Gene Scatter Plot \n",
    "gene_list <- c(\"TOP2A\",\"MKI67\",\"ASPM\",\"PTPRZ1\",\"PDGFRA\",\"PCDH15\",\"NXPH1\",\"CSPG4\",\"SERPINC5\",\"PMP22\",\"PLP1\",\n",
    "               \"MBP\",\"KLK6\",\"APOD\",\"MOG\",\"SIRT2\",\"RASGEF1B\",\"TMEM108\",\"GSN\",\"MOBP\") %>% str_to_title() %>%\n",
    "  intersect(row.names(starmap_raw@assays$RNA))\n",
    "plot_list <- list()\n",
    "for(i in 1:4){\n",
    "  plot_list[[i]]<-plot_genes_in_pseudotime(psdtime_res_list$Oligo$cds[gene_list[(5*i -4):min(5*i,length(gene_list))],],\n",
    "                         color_cells_by=\"cell_type\",\"~ splines::ns(pseudotime, df=3)\",\n",
    "                         min_expr=0.5) + scale_color_manual(values = cell_type_palette_list$Oligo)\n",
    "}\n",
    "\n",
    "for(i in c(\"Pcdh15\",\"Pdgfra\",\"Mbp\",\"Plp1\",\"Klk6\",\"Mobp\")){\n",
    "  plot_list[[i]] <- ad_plot_cells(psdtime_res_list$Oligo$cds, genes=c(i),\n",
    "           show_trajectory_graph=T,norm_method = \"log\", min_expr = 0.1, \n",
    "           alpha = 0.5, cell_size = 0.7,\n",
    "           label_cell_groups=F,\n",
    "           label_branch_points = F,\n",
    "           label_roots = T,\n",
    "           label_leaves=F) + scale_color_steps(low = \"red\", high = \"red\",na.value = \"grey80\") \n",
    "}\n",
    "\n",
    "\n",
    "pdf(\"Jun01_pseudotime_diffmap.pdf\", width = 6, height = 6)\n",
    "plot_list%>% print()\n",
    "dev.off()\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9b5e0a0",
   "metadata": {},
   "source": [
    "## Gene Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a979292",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(ComplexHeatmap)\n",
    "gene_cl_res <- list()\n",
    "\n",
    "gene_cl_res[[\"ADmouse_9723\"]] <- ad_s_gene_cl(starmap_raw, \n",
    "                                              starmap_raw@misc$ADmouse_9723_plaque_all$ADmouse_9723_cell_meta,\n",
    "                                              \"ADmouse_9723\",   p_filter = T, p_val = 0.05)\n",
    "gene_cl_res[[\"ADmouse_9723_2\"]] <- ad_s_gene_cl(starmap_raw, \n",
    "                                              starmap_raw@misc$ADmouse_9723_2_plaque_all$ADmouse_9723_2_cell_meta,\n",
    "                                              \"ADmouse_9723_2\", p_filter = T, p_val = 0.05)\n",
    "gene_cl_res[[\"ADmouse_9494\"]] <- ad_s_gene_cl(starmap_raw, \n",
    "                                              starmap_raw@misc$ADmouse_9494_plaque_all$ADmouse_9494_cell_meta,\n",
    "                                              \"ADmouse_9494\",   p_filter = T, p_val = 0.05)\n",
    "gene_cl_res[[\"ADmouse_11346\"]] <- ad_s_gene_cl(starmap_raw, \n",
    "                                              starmap_raw@misc$ADmouse_11346_plaque_all$ADmouse_11346_cell_meta,\n",
    "                                              \"ADmouse_11346\",  p_filter = T, p_val = 0.05)\n",
    "\n",
    "pdf(\"Mar07_gene_cl_p_05.pdf\",height = 6,width = 4)\n",
    "gene_cl_res[[\"ADmouse_9723\"]]$combined_heatmap %>% draw()\n",
    "gene_cl_res[[\"ADmouse_9723_2\"]]$combined_heatmap %>% draw()\n",
    "gene_cl_res[[\"ADmouse_9494\"]]$combined_heatmap %>% draw()\n",
    "gene_cl_res[[\"ADmouse_11346\"]]$combined_heatmap %>% draw()\n",
    "dev.off()\n",
    "# For Combined Spatial Analysis\n",
    "p_val <- 0.05\n",
    "sample_list <- list(\"8mo\" = c(\"ADmouse_9723\",\"ADmouse_9723_2\"), \"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\"))\n",
    "gene_list <- union(union(gene_cl_res[[\"ADmouse_9723\"]][[\"stat_test\"]][[\"gene\"]],\n",
    "                         gene_cl_res[[\"ADmouse_9723_2\"]][[\"stat_test\"]][[\"gene\"]]),\n",
    "                   union(gene_cl_res[[\"ADmouse_9494\"]][[\"stat_test\"]][[\"gene\"]],\n",
    "                         gene_cl_res[[\"ADmouse_11346\"]][[\"stat_test\"]][[\"gene\"]]))\n",
    "expr_df <- data.frame(gene = rep(gene_list, times = length(sample_list)),\n",
    "                      sample = \"-\",\n",
    "                      R10 = 0,R20 = 0,R30 = 0,R40 = 0,\"R40+\" = 0,\n",
    "                      RA10 = 0,RA20 = 0,RA30 = 0,RA40 = 0,\"RA40+\" = 0,\n",
    "                      RB10 = 0,RB20 = 0,RB30 = 0,RB40 = 0,\"RB40+\" = 0)\n",
    "expr_df[[\"sample\"]] <- rep(sample_list, each = length(gene_list))\n",
    "expr_df[[\"group\"]] <- rep(names(sample_list), each = length(gene_list))\n",
    "\n",
    "colnames(expr_df)[7] <- \"R40+\"\n",
    "colnames(expr_df)[12] <- \"RA40+\"\n",
    "colnames(expr_df)[17] <- \"RB40+\"\n",
    "\n",
    "\n",
    "sample_vec <- unlist(sample_list)\n",
    "\n",
    "for(i in 1:length(sample_vec)){\n",
    "  if(i == 1){\n",
    "    tmp_mat <- starmap_raw@assays$scaled@data[,starmap_raw@meta.data$sample == sample_vec[1]]\n",
    "    cell_meta <- data.frame(sample = sample_vec[1],\n",
    "                            starmap_raw@misc[[paste0(sample_vec[1],\"_plaque_all\")]][[paste0(sample_vec[1],\"_cell_meta\")]]) \n",
    "  }else{\n",
    "    tmp_mat <- cbind(tmp_mat,\n",
    "                     starmap_raw@assays$scaled@data[,starmap_raw@meta.data$sample == sample_vec[i]])\n",
    "    cell_meta <- rbind(cell_meta,\n",
    "                       data.frame(sample = sample_vec[i],\n",
    "                       starmap_raw@misc[[paste0(sample_vec[i],\"_plaque_all\")]][[paste0(sample_vec[i],\"_cell_meta\")]]) )\n",
    "  }\n",
    "}\n",
    "#c(\"R10\",\"R20\",\"R30\",\"R40\",\"R40+\")\n",
    "for(i in 1:5){\n",
    "  expr_df[[i+2]] <- apply(expr_df,1,\n",
    "                        FUN = function(x){\n",
    "                          if(i == 5){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                           cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                          }else{\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                           cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                          }\n",
    "                        })\n",
    "  expr_df[[i+7]] <- apply(expr_df,1,\n",
    "                        FUN = function(x){\n",
    "                          if(i == 5){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                           cell_meta$sample %in% unlist(x[[\"sample\"]])[1] ])\n",
    "                          }else{\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                           cell_meta$sample == unlist(x[[\"sample\"]])[1] ])\n",
    "                          }\n",
    "                        })\n",
    "  expr_df[[i+12]] <- apply(expr_df,1,\n",
    "                        FUN = function(x){\n",
    "                          if(i == 5){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                           cell_meta$sample %in% unlist(x[[\"sample\"]])[2] ])\n",
    "                          }else{\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                           cell_meta$sample == unlist(x[[\"sample\"]])[2] ])\n",
    "                          }\n",
    "                        })\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "library(factoextra)\n",
    "#fviz_nbclust(t(scale(t(as.matrix(expr_df[,3:7])))),kmeans,k.max = 15,method = \"wss\")\n",
    "\n",
    "seq_num <- 6\n",
    "set.seed(2021)\n",
    "tmp_df.km_res <- kmeans(t(scale(t(as.matrix(expr_df[,3:7])))), seq_num, nstart = 25, iter.max = 100)\n",
    "max_int_vec <- rep(0,seq_num)\n",
    "for(i in 1:seq_num){\n",
    "  tmp_vec <- expr_df[tmp_df.km_res$cluster == i,3:7] %>% colSums() %>% .[1:4]\n",
    "  max_int_vec[i] <- which(tmp_vec == max(tmp_vec))\n",
    "}\n",
    "names(max_int_vec) <- as.character(1:seq_num)\n",
    "\n",
    "#Visualization\n",
    "library(umap)\n",
    "tmp_df <- data.frame(expr_df,cluster = tmp_df.km_res$cluster)\n",
    "tmp_df.pca <- prcomp(t(tmp_df[,8:17]), rank. = 3)\n",
    "custom.config = umap.defaults\n",
    "custom.config$n_neighbors = 15\n",
    "tmp_df.umap <- umap(tmp_df[,8:17], config = custom.config)\n",
    "\n",
    "pdf(\"Apr17_gene_cl_pca.pdf\",width = 6, height = 6)\n",
    "ggplot(data = data.frame(as.matrix(tmp_df[,8:17]) %*% tmp_df.pca$x, group = tmp_df$group, cl = as.factor(tmp_df$cluster)),\n",
    "       aes(x = PC1, y = PC2, color = group)) + \n",
    "  geom_point() + theme_classic() + xlim(-10,10) + ylim(-10,10)\n",
    "\n",
    "ggplot(data = data.frame(as.matrix(tmp_df[,8:17]) %*% tmp_df.pca$x, group = tmp_df$group, cl = as.factor(tmp_df$cluster)),\n",
    "       aes(x = PC1, y = PC2, color = cl)) + \n",
    "  geom_point() + theme_classic() + xlim(-10,10) + ylim(-10,10)\n",
    "dev.off()\n",
    "#Stat test\n",
    "SDEG_ls <- list()\n",
    "logfc <- 0.1\n",
    "for(i in 1:2){\n",
    "  group_name <- names(sample_list)[i]\n",
    "  \n",
    "  gene_df <- data.frame(gene = expr_df$gene[expr_df$group == group_name], \n",
    "                        cluster = tmp_df.km_res$cluster[expr_df$group == group_name], \n",
    "                        max_int = 0, p_val = 1, logfc = 0,\n",
    "                        expr_df[expr_df$group == group_name,3:17])\n",
    "  \n",
    "  gene_df$max_int <- apply(gene_df,1,FUN = function(x){max_int_vec[x[[\"cluster\"]] ]})\n",
    "  gene_df$p_val <- apply(gene_df,1,\n",
    "                             FUN = function(x){\n",
    "                               i = x[[\"max_int\"]]\n",
    "                               tmp_vec1 <- tmp_mat[x[[\"gene\"]], cell_meta$sample %in% sample_list[[group_name]] &\n",
    "                                                     cell_meta$interval == paste(i,\"0um\",sep = \"\")]\n",
    "                               tmp_vec2 <- tmp_mat[x[[\"gene\"]], cell_meta$sample %in% sample_list[[group_name]] &\n",
    "                                                     cell_meta$interval != paste(i,\"0um\",sep = \"\")]\n",
    "                               a <- wilcox.test(tmp_vec1,tmp_vec2, alternative = \"greater\")$p.value\n",
    "                               return(a)\n",
    "                             })\n",
    "  \n",
    "  \n",
    "  gene_df$p_val_adj <- p.adjust(gene_df$p_val)\n",
    "  \n",
    "  SDEG_ls[[group_name]]  <- gene_df\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5721416c",
   "metadata": {},
   "source": [
    "### Export cl res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ce8f899",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "group_name <- \"13mo\"\n",
    "tmp_df <- data.frame(expr_df[expr_df$group == group_name,c(1,8:17)],\n",
    "                     p_val = SDEG_ls[[group_name]]$p_val,\n",
    "                     cl = SDEG_ls[[group_name]]$cluster)\n",
    "seq_vec <- c(2,6,4,3,5,1)\n",
    "tmp_df$cluster <- apply(tmp_df,1,\n",
    "                        FUN = function(x){ paste0(\"Cluster_\",which(seq_vec == as.numeric(x[[\"cl\"]])))})\n",
    "tmp_df <- arrange(tmp_df,cluster)\n",
    "\n",
    "write_csv(tmp_df,paste0(\"May26_gene_cl_\",group_name,\".csv\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db149a9b",
   "metadata": {},
   "source": [
    "### Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25b8c98b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "group_name <- \"13mo\"\n",
    "min_pct = 0.05\n",
    "v_heatmap <- NULL\n",
    "v_heatmap2 <- NULL\n",
    "v_heatmap3 <- NULL\n",
    "tmp_df <- SDEG_ls[[group_name]][,6:20]\n",
    "row.names(tmp_df) <- SDEG_ls[[group_name]]$gene\n",
    "seq_vec <- c(2,6,4,3,5,1)\n",
    "for(i in seq_vec){\n",
    "  tmp_mat2 <- tmp_df[tmp_list[[group_name]]$cluster == i, 6:15]\n",
    "  tmp_mat2[,1:5] <- t(scale(t(tmp_mat2[,1:5])))\n",
    "  tmp_mat2[,6:10] <- t(scale(t(tmp_mat2[,6:10])))\n",
    "  v_heatmap <- v_heatmap %v% Heatmap(tmp_mat2,\n",
    "                            cluster_columns = F,cluster_rows  = T, show_row_names = F, show_row_dend = F,\n",
    "                            row_title = paste(\"C\",i),border_gp = gpar(col = \"black\", lty = 1, lwd = 2),\n",
    "                            column_title = paste(\"PCT:\",min_pct,\"P-val\",p_val)) \n",
    "}\n",
    "\n",
    "seq_vec <- c(5,1)\n",
    "for(i in seq_vec){\n",
    "  if(sum(SDEG_ls[[group_name]]$cluster == i & SDEG_ls[[group_name]]$p_val < p_val) > 0){\n",
    "    tmp_mat2 <- tmp_df[SDEG_ls[[group_name]]$cluster == i & SDEG_ls[[group_name]]$p_val < p_val, 6:15]\n",
    "    tmp_mat2[,1:5] <- t(scale(t(tmp_mat2[,1:5])))\n",
    "    tmp_mat2[,6:10] <- t(scale(t(tmp_mat2[,6:10])))\n",
    "    v_heatmap3 <- v_heatmap3 %v% Heatmap(as.matrix(tmp_mat2),\n",
    "                              cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = F,\n",
    "                              row_title = paste(\"C\",i),border_gp = gpar(col = \"black\", lty = 1, lwd = 2),\n",
    "                              column_title = paste(group_name,\"Averaged PCT:\",min_pct,\"P-val\",p_val)) \n",
    "  }\n",
    "}\n",
    "#2,6,4,3,5,1\n",
    "seq_vec <- c(5,1)\n",
    "for(i in seq_vec){\n",
    "  if(sum(SDEG_ls[[group_name]]$cluster == i & SDEG_ls[[group_name]]$p_val_adj < p_val) > 0){\n",
    "    tmp_mat2 <- tmp_df[SDEG_ls[[group_name]]$cluster == i & SDEG_ls[[group_name]]$p_val_adj < p_val, 6:15]\n",
    "    tmp_mat2[,1:5] <- t(scale(t(tmp_mat2[,1:5])))\n",
    "    tmp_mat2[,6:10] <- t(scale(t(tmp_mat2[,6:10])))\n",
    "    v_heatmap3 <- v_heatmap3 %v% Heatmap(as.matrix(tmp_mat2),\n",
    "                              cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = F,\n",
    "                              row_title = paste(\"C\",i),border_gp = gpar(col = \"black\", lty = 1, lwd = 2),\n",
    "                              column_title = paste(group_name,\"Averaged PCT:\",min_pct,\"P-val-adj\",p_val)) \n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "pdf(paste(\"Apr19\",paste(group_name, collapse = \"_\"),\"gene_cl_heatmap.pdf\", sep = \"_\"),width = 8,height = 16)\n",
    "#v_heatmap %>% draw()\n",
    "#v_heatmap2 %>% draw()\n",
    "v_heatmap3 %>% draw()\n",
    "dev.off()\n",
    "\n",
    "PIG <- c('APOE','AXL','CD63','CD63-PS','CD9','CTSB','CTSD','CTSL','CTSZ','H2-K1','HEXA','LGALS3BP','LYZ2',\n",
    "         'NPC2','TREM2','TYROBP','H2-D1','B2M','C4B','GFAP','SERPINA3N','ARPC1B','C1QA','C1QB','C1QC','C4A',\n",
    "         'CLU','CSF1R','CST3','CTSA','CTSS','CTSH','CX3CR1','CYBA','FCER1G','FCGR3','FCRLS','GRN','GUSB','GNS',\n",
    "         'GPX4','HEXB','IGFBP5','ITGB5','ITM2B','LAPTM5','LGMN','LY86','MAN2B1','MPEG1','OLFML3','PLEK','PRDX6',\n",
    "         'GPX4-PS','S100A6','RPL18A','VSIR')\n",
    "library(VennDiagram)\n",
    "venn.diagram(\n",
    "        x = list(PIG,\n",
    "                 tmp_list$`8mo`$gene[tmp_list$`8mo`$cluster %in% c(2,4,3,6) & \n",
    "                                       tmp_list$`8mo`$p_val_adj < 0.05] %>% \n",
    "                   as.character() %>% str_to_upper(),\n",
    "                 tmp_list$`13mo`$gene[tmp_list$`13mo`$cluster %in% c(2,4,3,6) &\n",
    "                                        tmp_list$`13mo`$p_val_adj < 0.05] %>% \n",
    "                   as.character() %>% str_to_upper()),\n",
    "        category.names = c(\"PIG\",\"SDEG_8\",\"SDEG_13\"),filename = 'Mar13_SDEG_venn_sep_p_05.png',\n",
    "        output=TRUE,imagetype = \"png\")\n",
    "\n",
    "venn.diagram(\n",
    "        x = list(PIG,\n",
    "                 tmp_list$`8mo`$gene[tmp_list$`8mo`$cluster %in% c(2,4,3,6) & \n",
    "                                       tmp_list$`8mo`$p_val_A < 0.05 & tmp_list$`8mo`$p_val_B < 0.05] %>% \n",
    "                   as.character() %>% str_to_upper(),\n",
    "                 tmp_list$`13mo`$gene[tmp_list$`13mo`$cluster %in% c(2,4,3,6) &\n",
    "                                        tmp_list$`13mo`$p_val_A < 0.05 & tmp_list$`13mo`$p_val_B < 0.05] %>% \n",
    "                   as.character() %>% str_to_upper()),\n",
    "        category.names = c(\"PIG\",\"SDEG_8\",\"SDEG_13\"),filename = 'Mar13_SDEG_venn_sep_p_05.png',\n",
    "        output=TRUE,imagetype = \"png\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfdf0614",
   "metadata": {},
   "source": [
    "### Gene Cl GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e593466d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#gprofiler\n",
    "plot_list <- list()\n",
    "gostres_ls <- list()\n",
    "for(i in c(\"8mo\",\"13mo\")){\n",
    "  tmp_vec <- tmp_list[[i]]$gene[tmp_list[[i]]$cluster%in% c(2,4,3,6) & tmp_list[[i]]$p_val_adj < p_val] %>% as.character()\n",
    "  gostres_ls[[paste0(i,\"_filtered\")]] <- gost(query = tmp_vec,organism = \"mmusculus\", custom_bg = row.names(starmap_raw))\n",
    "}\n",
    "\n",
    "gostres_ls[[paste0(\"PIG_filtered\")]] <- gost(query = PIG,organism = \"hsapiens\",\n",
    "                                             custom_bg = str_to_upper(row.names(starmap_raw)))\n",
    "\n",
    "tmp_vec <- c()\n",
    "for(i in c(\"8mo\",\"13mo\",\"PIG\")){\n",
    "  tmp_vec <- c(tmp_vec, gostres_ls[[paste0(i,\"_filtered\")]]$result %>% \n",
    "                 subset(p_value < 0.05 & term_size < 1000 & term_size > 10 & source %in% c(\"GO:BP\",\"GO:CC\")) %>%\n",
    "                 .$term_name) %>%\n",
    "    unique()\n",
    "  plot_list[[paste0(i,\"_filtered\")]] <- ggplot(gostres_ls[[paste0(i,\"_filtered\")]]$result %>% \n",
    "           subset(p_value < 0.05 & term_size < 1000 & term_size > 10 & source %in% c(\"GO:BP\",\"GO:CC\")) %>%\n",
    "           arrange(desc(p_value)) %>% \n",
    "           mutate(new_term_name = term_name %>% factor(levels = term_name)),\n",
    "       aes(x = new_term_name, y = -log10(p_value))) + labs(title = i) +\n",
    "    geom_bar(stat = \"identity\", fill = \"darkblue\") +theme_classic() + coord_flip()\n",
    "}\n",
    "\n",
    "pdf(paste0(\"May12_SDEG_GO_barplot_bg.pdf\"),width = 6,height = 6)\n",
    "plot_list %>% print()\n",
    "dev.off()\n",
    "\n",
    "\n",
    "tmp_df <- data.frame(term_name = rep(tmp_vec,each = 3), \n",
    "                     term_source = rep(c(\"8mo\",\"13mo\",\"PIG\"),length(tmp_vec)),\n",
    "                     log2p = 0)\n",
    "\n",
    "for(i in 1:dim(tmp_df)[1]){\n",
    "  if(tmp_df$term_name[i] %in% gostres_ls[[paste0(tmp_df$term_source[i],\"_filtered\")]]$result$term_name)\n",
    "  tmp_df$log2p[i] <- gostres_ls[[paste0(tmp_df$term_source[i],\"_filtered\")]]$result %>% \n",
    "    subset(term_name == tmp_df$term_name[i]) %>% .$p_value %>% -log2(.)\n",
    "  else 0\n",
    "}\n",
    "\n",
    "ggplot(data = tmp_df %>% subset(log2p > 0), aes(x = term_name, y = log2p, fill = term_source)) + \n",
    "  geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "  theme_classic() + coord_flip()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da508f37",
   "metadata": {},
   "source": [
    "## Cell type SDEG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceb14c7d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# For Combined Spatial Analysis\n",
    "group_name <- \"13mo\"\n",
    "p_val <- 0.05\n",
    "sample_list <- list(\"8mo\" = c(\"ADmouse_9723\",\"ADmouse_9723_2\"), \"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\"))\n",
    "tmp_cell_type_list <- list(\"Micro\" = c(\"Micro\"), \"Astro\" = c(\"Astro\"), \"Oligo\" = c(\"Oligo\"),\n",
    "                           \"Neuron\" = c(\"CTX-Ex\",\"CA1\",\"CA2\",\"CA3\",\"DG\",\"Inh\"), \"CTX-Ex\" = c(\"CTX-Ex\"), \"DG\" = c(\"DG\"))\n",
    "gene_list <- row.names(starmap_raw)\n",
    "expr_df <- data.frame(gene = rep(gene_list, times = length(tmp_cell_type_list)),\n",
    "                      cell_type = \"-\", sample = \"-\",\n",
    "                      R10 = 0,R20 = 0,R30 = 0,R40 = 0,\"R40+\" = 0,\n",
    "                      RA10 = 0,RA20 = 0,RA30 = 0,RA40 = 0,\"RA40+\" = 0,\n",
    "                      RB10 = 0,RB20 = 0,RB30 = 0,RB40 = 0,\"RB40+\" = 0)\n",
    "expr_df[[\"cell_type\"]] <- rep(tmp_cell_type_list, each = length(gene_list))\n",
    "#1 8mo 2 13mo\n",
    "expr_df[[\"sample\"]] <- rep(sample_list[which(names(sample_list) == group_name)],\n",
    "                           each = length(gene_list)*length(tmp_cell_type_list))\n",
    "\n",
    "colnames(expr_df)[8] <- \"R40+\"\n",
    "colnames(expr_df)[13] <- \"RA40+\"\n",
    "colnames(expr_df)[18] <- \"RB40+\"\n",
    "\n",
    "sample_vec <- sample_list %>% unlist()\n",
    "cell_type_vec <- unlist(tmp_cell_type_list)\n",
    "\n",
    "for(i in 1:length(sample_vec)){\n",
    "  if(i == 1){\n",
    "    tmp_mat <- starmap_raw@assays$raw@data[,starmap_raw@meta.data$sample == sample_vec[1]]\n",
    "    cell_meta <- data.frame(sample = sample_vec[1],\n",
    "                            starmap_raw@misc[[paste0(sample_vec[1],\"_plaque_all\")]][[paste0(sample_vec[1],\"_cell_meta\")]]) \n",
    "  }else{\n",
    "    tmp_mat <- cbind(tmp_mat,\n",
    "                     starmap_raw@assays$raw@data[,starmap_raw@meta.data$sample == sample_vec[i]])\n",
    "    cell_meta <- rbind(cell_meta,\n",
    "                       data.frame(sample = sample_vec[i],\n",
    "                       starmap_raw@misc[[paste0(sample_vec[i],\"_plaque_all\")]][[paste0(sample_vec[i],\"_cell_meta\")]]) )\n",
    "  }\n",
    "}\n",
    "#c(\"R10\",\"R20\",\"R30\",\"R40\",\"R40+\")\n",
    "for(i in 1:5){\n",
    "  expr_df[[i+3]] <- apply(expr_df,1,\n",
    "                        FUN = function(x){\n",
    "                          if(i == 5){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                           cell_meta$top_level %in% unlist(x[[\"cell_type\"]]) ])\n",
    "                          }else{\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                           cell_meta$top_level %in% unlist(x[[\"cell_type\"]])])\n",
    "                          }\n",
    "                        })\n",
    "  expr_df[[i+8]] <- apply(expr_df,1,\n",
    "                        FUN = function(x){\n",
    "                          if(i == 5){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                           cell_meta$sample %in% unlist(x[[\"sample\"]])[1] &\n",
    "                                           cell_meta$top_level %in% unlist(x[[\"cell_type\"]])])\n",
    "                          }else{\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                           cell_meta$sample == unlist(x[[\"sample\"]])[1] &\n",
    "                                           cell_meta$top_level %in% unlist(x[[\"cell_type\"]]) ])\n",
    "                          }\n",
    "                        })\n",
    "  expr_df[[i+13]] <- apply(expr_df,1,\n",
    "                        FUN = function(x){\n",
    "                          if(i == 5){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                           cell_meta$sample %in% unlist(x[[\"sample\"]])[2] &\n",
    "                                           cell_meta$top_level %in% unlist(x[[\"cell_type\"]])])\n",
    "                          }else{\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                           cell_meta$sample == unlist(x[[\"sample\"]])[2] &\n",
    "                                           cell_meta$top_level %in% unlist(x[[\"cell_type\"]])])\n",
    "                          }\n",
    "                        })\n",
    "}\n",
    "\n",
    "\n",
    "expr_df[[\"max_int\"]] <- apply(expr_df,1,\n",
    "                              FUN = function(x){\n",
    "                                R10 <- as.numeric(x[[\"R10\"]])\n",
    "                                R20 <- as.numeric(x[[\"R20\"]])\n",
    "                                R30 <- as.numeric(x[[\"R30\"]])\n",
    "                                R40 <- as.numeric(x[[\"R40\"]])\n",
    "                                R50 <- as.numeric(x[[\"R40+\"]])\n",
    "                                if( R10 > max(R20,R30,R40,R50)) 1\n",
    "                                else if( R20 > max(R10,R30,R40,R50)) 2\n",
    "                                else if( R30 > max(R20,R10,R40,R50)) 3\n",
    "                                else 4\n",
    "                              })\n",
    "\n",
    "expr_df[[\"p_val\"]] <- apply(expr_df,1,\n",
    "                           FUN = function(x){\n",
    "                             i = x[[\"max_int\"]]\n",
    "                             if(i == 4) return(1)\n",
    "                             tmp_vec1 <- tmp_mat[x[[\"gene\"]],\n",
    "                                                   cell_meta$top_level %in% unlist(x[[\"cell_type\"]]) &\n",
    "                                                   cell_meta$interval == paste(i,\"0um\",sep = \"\")]\n",
    "                             tmp_vec2 <- tmp_mat[x[[\"gene\"]],\n",
    "                                                   cell_meta$top_level %in% unlist(x[[\"cell_type\"]]) & \n",
    "                                                   cell_meta$interval != paste(i,\"0um\",sep = \"\")]\n",
    "                             a <- wilcox.test(tmp_vec1,tmp_vec2, alternative = \"greater\")$p.value\n",
    "                             return(a)\n",
    "                           })\n",
    "\n",
    "\n",
    "expr_df[[\"p_val_adj\"]] <- 1\n",
    "expr_df[[\"group\"]] <- rep(names(tmp_cell_type_list), each = length(gene_list))\n",
    "\n",
    "for(i in names(tmp_cell_type_list)){\n",
    "  expr_df[[\"p_val_adj\"]][expr_df$group == i] <- p.adjust(expr_df[[\"p_val\"]][expr_df$group == i])\n",
    "  #expr_df[[\"p_val_adj_B\"]][expr_df$group == i] <- p.adjust(expr_df[[\"p_val_B\"]][expr_df$group == i])\n",
    "}\n",
    "\n",
    "\n",
    "SDEG_ls[[paste0(group_name,\"_cell_type\")]] <- expr_df\n",
    "\n",
    "#Print Astro/Micro/Oligo SDEG\n",
    "expr_df <- SDEG_ls[[\"13mo_cell_type\"]] \n",
    "for(i in c(\"Astro\",\"Micro\",\"Oligo\")){\n",
    "  if(i == \"Astro\" | i == \"Oligo\"){\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$group == i & expr_df[[\"max_int\"]] != 1 \n",
    "  }else if(i == \"Micro\"){\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$group == i \n",
    "  }else{\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$group == i\n",
    "  }\n",
    "  print(paste(i,paste(expr_df$gene[tmp_vec], collapse = \" \")))\n",
    "}\n",
    "\n",
    "\n",
    "#Export xlsx\n",
    "wb <- createWorkbook()\n",
    "for(i in c(\"Astro\",\"Micro\",\"Oligo\")){\n",
    "  sheet <- createSheet(wb,sheetName = paste(i,\"SDEG\"))\n",
    "  if(i == \"Astro\" | i == \"Oligo\"){\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$group == i & expr_df[[\"max_int\"]] != 1 \n",
    "  }else if(i == \"Micro\"){\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$group == i \n",
    "  }else{\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$group == i\n",
    "  }\n",
    "  tmp_df <- expr_df[tmp_vec, c(1,4:18,20:21)]\n",
    "  addDataFrame(tmp_df, sheet=sheet, \n",
    "               startColumn=1, startRow=1, row.names=FALSE)\n",
    "}\n",
    "saveWorkbook(wb,file = \"Sep04_cell_type_SDEG.xlsx\")\n",
    "\n",
    "plot_list <- list()\n",
    "\n",
    "k = \"13mo\"\n",
    "expr_df <- SDEG_ls[[k]]\n",
    "p_val = 0.01\n",
    "for(i in c(\"Astro\",\"Oligo\",\"Micro\")){\n",
    "  gene_vec <- c()\n",
    "  for(j in cell_type_list[[i]]){\n",
    "    if(j != \"OPC\"){\n",
    "      tmp_vec <- starmap_raw@meta.data$group == \"disease\" & starmap_raw@meta.data$time == paste0(k,\"nths\") &\n",
    "        starmap_raw@meta.data$cell_type == j\n",
    "      tmp_vec2 <- starmap_raw@meta.data$group == \"disease\" & starmap_raw@meta.data$time == paste0(k,\"nths\") &\n",
    "        starmap_raw@meta.data$cell_type != j\n",
    "      gene_vec <- unique(c(gene_vec,\n",
    "                           row.names(starmap_raw@assays$RNA)[starmap_raw@assays[[\"RNA\"]]@meta.features[[\"max_counts\"]] >= 4 &\n",
    "                                                    rowSums(starmap_raw@assays$RNA@counts[,tmp_vec] > 0)/sum(tmp_vec) > 0.15 &\n",
    "                                                    rowSums(starmap_raw@assays$RNA@counts[,tmp_vec2] > 0)/sum(tmp_vec2) < 0.85]))\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  \n",
    "  \n",
    "  if(i == \"Astro\" | i == \"Oligo\"){\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$gene %in% gene_vec &\n",
    "    expr_df$group == i & expr_df[[\"max_int\"]] != 1 &  expr_df[[\"max_int\"]] != 1 &\n",
    "      abs(expr_df[[\"max_int\"]] - expr_df[[\"max_int\"]]) < 2\n",
    "  }else if(k == \"8mo\" & i == \"Micro\"){\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$gene %in% gene_vec &\n",
    "      expr_df$group == i & expr_df[[\"max_int\"]] != 3 &  expr_df[[\"max_int\"]] != 3 &\n",
    "      expr_df$group == i & abs(expr_df[[\"max_int\"]] - expr_df[[\"max_int\"]]) < 2\n",
    "  }else{\n",
    "    tmp_vec <- expr_df[[\"p_val\"]] < p_val & expr_df$gene %in% gene_vec &\n",
    "    expr_df$group == i & abs(expr_df[[\"max_int\"]] - expr_df[[\"max_int\"]]) < 2\n",
    "  }\n",
    "  tmp_mat2 <- expr_df[tmp_vec, 4:18]\n",
    "  row.names(tmp_mat2) <- expr_df$gene[tmp_vec]\n",
    "  tmp_mat2[,1:5] <- t(scale(t(tmp_mat2[,1:5])))\n",
    "  tmp_mat2[,6:10] <- t(scale(t(tmp_mat2[,6:10])))\n",
    "  tmp_mat2[,11:15] <- t(scale(t(tmp_mat2[,11:15])))\n",
    "  tmp_mat2 <- tmp_mat2[!is.na(tmp_mat2 %>% rowSums()), ]\n",
    "\n",
    "  plot_list[[paste0(k,\"_\",i)]] <-  Heatmap(as.matrix(tmp_mat2),\n",
    "                              cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = F,\n",
    "                              border_gp = gpar(col = \"black\", lty = 1, lwd = 2),\n",
    "                              rect_gp = gpar(col = \"black\", lty = 0.5, lwd = 0.5),\n",
    "                              row_title = paste(k,i),column_split = rep(c(\"Combined\",\"A\",\"B\"), each = 5)) \n",
    "}\n",
    "\n",
    "pdf(paste0(\"Apr13_SDEG_heatmap_pval.pdf\"),width = 6,height = 10)\n",
    "plot_list %>% print()\n",
    "dev.off()\n",
    "\n",
    "#Micro\n",
    "i = \"Micro\"\n",
    "tmp_vec2 <- c(\"Cst3\",\"Ctss\",\"Selplg\",\"P2ry12\",\"Kctd12\",\"Lpcat2\",\"Siglech\",\"Gpr34\",\"Sparc\",\"Srrm2\",\"Tgfbr1\",\"Cst7\",\"Ctsb\",\"Apoe\",\"Cd9\",\"Trem2\",\"Myo9a\",\"Lyz2\",\"Clec7a\",\"Gfap\",\"Cd63\",\"Ly6e\",\"Ctsd\",\"Serpinb1b\",\"C1qa\")\n",
    "#Astro\n",
    "i = \"Astro\"\n",
    "tmp_vec2 <- c(\"Slc1a3\",\"Aldoc\",\"Gpr37l1\",\"Gjb6\",\"Vsnl1\",\"Atp1a3\",\"Nrgn\",\"Gja1\",\"Id4\",\"Cst3\",\"Clu\",\"Gfap\",\"Aqp4\",\"Cd63\",\"Vim\",\"Tmem176b\",\"Stat3\",\"Fxyd1\",\"Cd63\")\n",
    "#Oligo\n",
    "i = \"Oligo\"\n",
    "tmp_vec2 <- c(\"Pdgfra\",\"Cacng4\",\"Pcdh15\",\"Ccnd1\",\"Serpine2\",\"Plp1\",\"Tspan2\",\"Aplp1\",\"Bin1\",\"Aspa\",\"Trf\",\"Klk6\",\"Mobp\",\"Cldn11\",\"Tubb4a\",\"App\",\"Stmn4\",\"Dnm2\",\"C4b\",\"Cd9\",\"Gfap\")\n",
    "\n",
    "tmp_vec <- expr_df$group == i & expr_df$gene %in% tmp_vec2\n",
    "tmp_mat2 <- expr_df[tmp_vec, 9:18]\n",
    "row.names(tmp_mat2) <- expr_df$gene[tmp_vec]\n",
    "tmp_mat2[,1:5] <- t(scale(t(tmp_mat2[,1:5])))\n",
    "tmp_mat2[,6:10] <- t(scale(t(tmp_mat2[,6:10])))\n",
    "tmp_mat2 <- tmp_mat2[!is.na(tmp_mat2 %>% rowSums()), ]\n",
    "\n",
    "plot_list[[paste(\"13mo\",i)]] <- Heatmap(as.matrix(tmp_mat2),\n",
    "                                        cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = F,\n",
    "                                        border_gp = gpar(col = \"black\", lty = 1, lwd = 2),\n",
    "                                        rect_gp = gpar(col = \"black\", lty = 1, lwd = 1),\n",
    "                                        row_title = paste(\"13mo\",i)) \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef3f5f35",
   "metadata": {},
   "source": [
    "## DE\n",
    "\n",
    "### Data Prepare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "059c88c7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "myGO_gmt <- suppressMessages(GSA::GSA.read.gmt(\"/Users/zhouhaowen/Desktop/At_Broad/AD/gprofiler_full_mmusculus.name.gmt\"))\n",
    "myGO <- data.frame(GOID = myGO_gmt$geneset.names, \n",
    "                   Description = myGO_gmt$geneset.descriptions, \n",
    "                   Term_size = lengths(myGO_gmt$genesets),\n",
    "                   row.names = myGO_gmt$geneset.names)\n",
    "\n",
    "ad_cell_type_median_norm <- function(starmap_obj){\n",
    "  starmap_scaled <- starmap_obj\n",
    "  median_vec <- rep(0,length(levels(starmap_scaled@meta.data[[\"sample\"]])))\n",
    "  names(median_vec) <- levels(starmap_scaled@meta.data[[\"sample\"]])\n",
    "  for(cell_type in levels(starmap_scaled@meta.data[[\"top_level\"]])){\n",
    "    for(i in levels(starmap_scaled@meta.data[[\"sample\"]])){\n",
    "      median_vec[[i]] <- median(starmap_scaled@meta.data[[\"n_counts\"]][starmap_scaled@meta.data[[\"sample\"]] == i & starmap_scaled@meta.data[[\"top_level\"]] == cell_type])\n",
    "    }\n",
    "    \n",
    "    scale_factor <- mean(na.omit(median_vec))\n",
    "    \n",
    "    for(i in levels(starmap_scaled@meta.data[[\"sample\"]])){\n",
    "      starmap_scaled@assays[[\"RNA\"]]@data[,starmap_scaled@meta.data[[\"sample\"]] == i  & starmap_scaled@meta.data[[\"top_level\"]] == cell_type] <- \n",
    "        scale_factor*starmap_scaled@assays[[\"RNA\"]]@data[,starmap_scaled@meta.data[[\"sample\"]] == i & starmap_scaled@meta.data[[\"top_level\"]] == cell_type]/\n",
    "               median(starmap_scaled@meta.data[[\"n_counts\"]][starmap_scaled@meta.data[[\"sample\"]] == i & starmap_scaled@meta.data[[\"top_level\"]] == cell_type])\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  starmap_scaled@misc <- list()\n",
    "  return(starmap_scaled)\n",
    "}\n",
    "#Use Raw Counts + Median Norm\n",
    "#0: \"ADmouse_9723_2\",\"ADmouse_9707\", \"ADmouse_11346\", \"ADmouse_11351\"\n",
    "#1: \"ADmouse_9723\",\"ADmouse_9494\",\"ADmouse_9735\",\"ADmouse_9498\"\n",
    "#2: all samples\n",
    "function(starmap_obj,\n",
    "                                 top_level_list,\n",
    "                                 time_name,\n",
    "                                 fig_title = NULL,\n",
    "                                 FC_cutoff = 0.1,\n",
    "                                 p_cutoff = 0.05,\n",
    "                                 Pcap = NULL,\n",
    "                                 FCcap = NULL,\n",
    "                                 draw_connector = F,\n",
    "                                 xrange = NULL,\n",
    "                                 yrange = NULL){\n",
    "  \n",
    "  tmp <- subset(starmap_obj,subset = top_level %in% top_level_list & time %in% time_name)\n",
    "  tmp <- SetIdent(object = tmp, value  = tmp@meta.data[[\"group\"]])\n",
    "  require(tibble)\n",
    "  tmp_df <- as.data.frame(AverageExpression(tmp)[[1]]) %>% rownames_to_column()\n",
    "  marker_list <- FindMarkers(tmp,ident.1 = \"disease\",ident.2 = \"control\",logfc.threshold = 0,min.pct = 0.05) %>% rownames_to_column()\n",
    "  \n",
    "  tmp_df <- left_join(marker_list,tmp_df,by = \"rowname\")\n",
    "  require(EnhancedVolcano)\n",
    "  if(!is.null(Pcap)) marker_list$p_val[marker_list$p_val < Pcap] <- Pcap\n",
    "  if(!is.null(FCcap)){\n",
    "    marker_list$avg_logFC[marker_list$avg_logFC > FCcap] <- FCcap\n",
    "    marker_list$avg_logFC[marker_list$avg_logFC < -FCcap] <- -FCcap\n",
    "  }\n",
    "  tmp_col <- ifelse(marker_list$p_val>p_cutoff,\"grey\",\n",
    "                    ifelse(abs(marker_list$avg_logFC) < FC_cutoff,\n",
    "                           \"darkgreen\",ifelse(marker_list$avg_logFC < 0, \"darkblue\", \"darkred\")))\n",
    "  require(stringr)\n",
    "  row.names(marker_list) <- str_to_title(row.names(marker_list))\n",
    "  names(tmp_col) <- row.names(marker_list)\n",
    "  marker_list$rowname <- str_to_title(marker_list$rowname)\n",
    "  tmp_plot <- EnhancedVolcano(marker_list,\n",
    "                              lab = marker_list$rowname,\n",
    "                              x=\"avg_logFC\",y=\"p_val\",\n",
    "                              title = paste(fig_title,time_name,\"Disease vs. Control\"),\n",
    "                              subtitle = \" \",\n",
    "                              xlim = ifelse(c(is.null(xrange),is.null(xrange)),\n",
    "                                            c(-max(abs(marker_list$avg_logFC)) - 0.1,\n",
    "                                              max(abs(marker_list$avg_logFC)) + 0.1), xrange),\n",
    "                              ylim = ifelse(c(is.null(yrange),is.null(yrange)),\n",
    "                                            c(0, max(-log10(marker_list$p_val), na.rm=TRUE) + 0.1)\n",
    "                                            ,yrange),\n",
    "                              FCcutoff = FC_cutoff,pCutoff = p_cutoff,\n",
    "                              colCustom = tmp_col,\n",
    "                              #Disable caption, grid and lines\n",
    "                              caption = \"\",\n",
    "                              captionLabSize = 0,\n",
    "                              legendPosition = \"none\",\n",
    "                              gridlines.major = F,\n",
    "                              gridlines.minor = F,\n",
    "                              drawConnectors = draw_connector,\n",
    "                              widthConnectors = 0.5,\n",
    "                              endsConnectors = 'none',\n",
    "                              colConnectors = 'grey50',\n",
    "                              arrowhead = F,\n",
    "                              shape = 16,\n",
    "                              pointSize = 1.5,\n",
    "                              colAlpha = 0.9)\n",
    "  \n",
    "  return(tmp_plot)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca588231",
   "metadata": {},
   "source": [
    "### Run DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53bc41fb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "DE_13mo <- list()\n",
    "DE_8mo <- list()\n",
    "\n",
    "for(i in c(\"combined\")){\n",
    "  starmap_scaled <- starmap_raw\n",
    "  starmap_scaled@assays$RNA@data <- starmap_scaled@assays$raw@data\n",
    "  message(i)\n",
    "  if(i == 1){\n",
    "    starmap_scaled <- starmap_scaled %>%  subset(subset = sample %in% c(\"ADmouse_9723\",\"ADmouse_9494\",\"ADmouse_9735\",\"ADmouse_9498\"))\n",
    "  }else if(i == 2){\n",
    "    starmap_scaled <- starmap_scaled %>%  subset(subset = sample %in% c(\"ADmouse_9723_2\",\"ADmouse_9707\", \"ADmouse_11346\", \"ADmouse_11351\"))\n",
    "  }else if(i == \"combined\"){\n",
    "    #do nothing\n",
    "  }else break\n",
    "  \n",
    "  #starmap_scaled <- ad_cell_type_median_norm(starmap_scaled)\n",
    "  starmap_scaled <- ad_starmap_normalize(starmap_scaled)\n",
    "  \n",
    "  DE_13mo[[paste0(\"Micro\",i)]] <- ad_D_vs_C_DE(starmap_scaled,\"Micro\",\"13months\",logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_13mo[[paste0(\"Astro\",i)]] <- ad_D_vs_C_DE(starmap_scaled,\"Astro\",\"13months\",logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_13mo[[paste0(\"Oligo\",i)]] <- ad_D_vs_C_DE(starmap_scaled,\"Oligo\",\"13months\",logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_13mo[[paste0(\"Neuron\",i)]]<- ad_D_vs_C_DE(starmap_scaled,c(\"CTX-Ex\",\"CA1\",\"CA2\",\"CA3\",\"DG\",\"Inh\"),\"13months\",\n",
    "                                     logfc_cutoff = 0.1,p_cutoff = 0.1)\n",
    "  DE_13mo[[paste0(\"CTX-Ex\",i)]]<- ad_D_vs_C_DE(starmap_scaled,c(\"CTX-Ex\"),\"13months\",\n",
    "                                     logfc_cutoff = 0.1,p_cutoff = 0.1)\n",
    "  DE_13mo[[paste0(\"Inh\",i)]]   <- ad_D_vs_C_DE(starmap_scaled,c(\"Inh\"),\"13months\",\n",
    "                                     logfc_cutoff = 0.1,p_cutoff = 0.1)\n",
    "  DE_13mo[[paste0(\"DG\",i)]]<- ad_D_vs_C_DE(starmap_scaled,c(\"DG\"),\"13months\",\n",
    "                                     logfc_cutoff = 0.1,p_cutoff = 0.1)\n",
    "  #\n",
    "  #for(j in c(\"CA1\",\"CA2\",\"CA3\",\"DG\",\"Inh\",\"CTX-Ex\")){\n",
    "  #  DE_13mo[[paste0(j,i)]] <- ad_D_vs_C_DE(starmap_scaled,j,\"13months\",logfc_cutoff = 0.05,p_cutoff = 0.05)\n",
    "  #  #DE_8mo[[paste0(j,i)]] <- ad_D_vs_C_DE(starmap_scaled,j,\"8months\",logfc_cutoff = 0.05,p_cutoff = 0.05,sub_type = T)\n",
    "  #}\n",
    "  \n",
    "  DE_8mo[[paste0(\"Micro\",i)]] <- ad_D_vs_C_DE(starmap_scaled,\"Micro\",\"8months\",logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_8mo[[paste0(\"Astro\",i)]] <- ad_D_vs_C_DE(starmap_scaled,\"Astro\",\"8months\",logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_8mo[[paste0(\"Oligo\",i)]] <- ad_D_vs_C_DE(starmap_scaled,\"Oligo\",\"8months\",logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_8mo[[paste0(\"Neuron\",i)]]<- ad_D_vs_C_DE(starmap_scaled,c(\"CTX-Ex\",\"CA1\",\"CA2\",\"CA3\",\"DG\",\"Inh\"),\"8months\",\n",
    "                                     logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_8mo[[paste0(\"CTX-Ex\",i)]]<- ad_D_vs_C_DE(starmap_scaled,c(\"CTX-Ex\"),\"8months\",\n",
    "                                     logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_8mo[[paste0(\"Inh\",i)]]   <- ad_D_vs_C_DE(starmap_scaled,c(\"Inh\"),\"8months\",\n",
    "                                     logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "  DE_8mo[[paste0(\"DG\",i)]]<- ad_D_vs_C_DE(starmap_scaled,c(\"DG\"),\"8months\",\n",
    "                                     logfc_cutoff = 0.1,p_cutoff = 0.05)\n",
    "}\n",
    "#Overlap Stat\n",
    "tmp_df <- data.frame(cell_type = c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\"))\n",
    "tmp_df[[\"rep1\"]] <- apply(tmp_df,1,FUN = function(x){length(DE_8mo[[paste0(x[[\"cell_type\"]],1)]]$rowname) })\n",
    "tmp_df[[\"rep2\"]] <- apply(tmp_df,1,FUN = function(x){length(DE_8mo[[paste0(x[[\"cell_type\"]],2)]]$rowname) })\n",
    "tmp_df[[\"overlap\"]] <- apply(tmp_df,1,\n",
    "                             FUN = function(x){\n",
    "                               length(intersect(DE_8mo[[paste0(x[[\"cell_type\"]],1)]]$rowname,DE_8mo[[paste0(x[[\"cell_type\"]],2)]]$rowname)) \n",
    "                               })"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3b88927",
   "metadata": {},
   "source": [
    "### Tau+ CA1 vs tau-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba5cea42",
   "metadata": {
    "lines_to_next_cell": 2,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "for(i in c(\"8months\",\"13months\")){\n",
    "  tmp <- subset(starmap_raw, time == i & group == \"disease\"& top_level == \"CA1\")\n",
    "  tmp@meta.data[[\"tau_group\"]] <- sapply(tmp@meta.data$tau,function(x){ifelse(as.numeric(x) > 7, \"pos\", \"neg\")})\n",
    "  tmp <- SetIdent(object = tmp, value  = tmp@meta.data[[\"tau_group\"]])\n",
    "  tmp_df <- FindMarkers(tmp,ident.1 = \"pos\",ident.2 = \"neg\",logfc.threshold = 0.05,min.pct = 0.05) %>% subset(p_val < 0.05) \n",
    "  if(i == \"8months\"){\n",
    "    DE_8mo[[\"p_Tau_CA1\"]] <- tmp_df\n",
    "  }else if(i == \"13months\"){\n",
    "    DE_13mo[[\"p_Tau_CA1\"]] <- tmp_df\n",
    "  }\n",
    "  \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e674690",
   "metadata": {},
   "source": [
    "### Export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b70b7320",
   "metadata": {
    "lines_to_next_cell": 2,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#Export xlsx\n",
    "library(xlsx)\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\")){\n",
    "  k <- max(dim(DE_8mo[[paste0(i,1)]])[1],dim(DE_8mo[[paste0(i,2)]])[1],dim(DE_8mo[[paste0(i,\"combined\")]])[1])\n",
    "  tmp_df <- data.frame(sample1 = \"\",rbind(DE_8mo[[paste0(i,1)]],\n",
    "                                          data.frame(rowname = rep(\"\",1+k - dim(DE_8mo[[paste0(i,1)]])[1]), \n",
    "                                                     p_val = \"\", avg_logFC = \"\", pct.1 = \"\",pct.2 = \"\",p_val_adj = \"\")),\n",
    "                       sample2 = \"\",rbind(DE_8mo[[paste0(i,2)]],\n",
    "                                          data.frame(rowname = rep(\"\",1+k - dim(DE_8mo[[paste0(i,2)]])[1]), \n",
    "                                                     p_val = \"\", avg_logFC = \"\", pct.1 = \"\",pct.2 = \"\",p_val_adj = \"\")),\n",
    "                       combined = \"\",rbind(DE_8mo[[paste0(i,\"combined\")]],\n",
    "                                          data.frame(rowname = rep(\"\",1+k - dim(DE_8mo[[paste0(i,\"combined\")]])[1]), \n",
    "                                                     p_val = \"\", avg_logFC = \"\", pct.1 = \"\",pct.2 = \"\",p_val_adj = \"\")))\n",
    "  tmp_df <- tmp_df[-dim(tmp_df)[1],]\n",
    "  write.xlsx(tmp_df, file=\"Mar30_DEG.xlsx\", sheetName=paste0(i,\"_8mo\"), append = T, row.names=FALSE)\n",
    "  k <- max(dim(DE_13mo[[paste0(i,1)]])[1],dim(DE_13mo[[paste0(i,2)]])[1],dim(DE_13mo[[paste0(i,\"combined\")]])[1])\n",
    "  tmp_df <- data.frame(sample1 = \"\",rbind(DE_13mo[[paste0(i,1)]],\n",
    "                                          data.frame(rowname = rep(\"\",1+k - dim(DE_13mo[[paste0(i,1)]])[1]), \n",
    "                                                     p_val = \"\", avg_logFC = \"\", pct.1 = \"\",pct.2 = \"\",p_val_adj = \"\")),\n",
    "                       sample2 = \"\",rbind(DE_13mo[[paste0(i,2)]],\n",
    "                                          data.frame(rowname = rep(\"\",1+k - dim(DE_13mo[[paste0(i,2)]])[1]), \n",
    "                                                     p_val = \"\", avg_logFC = \"\", pct.1 = \"\",pct.2 = \"\",p_val_adj = \"\")),\n",
    "                       combined = \"\",rbind(DE_13mo[[paste0(i,\"combined\")]],\n",
    "                                          data.frame(rowname = rep(\"\",1+k - dim(DE_13mo[[paste0(i,\"combined\")]])[1]), \n",
    "                                                     p_val = \"\", avg_logFC = \"\", pct.1 = \"\",pct.2 = \"\",p_val_adj = \"\")))\n",
    "  tmp_df <- tmp_df[-dim(tmp_df)[1],]\n",
    "  write.xlsx(tmp_df, file=\"Mar30_DEG.xlsx\", sheetName=paste0(i,\"_13mo\"), append = T, row.names=FALSE)\n",
    "}\n",
    "\n",
    "#####For Combined Result#####\n",
    "wb = createWorkbook()\n",
    "#Set cell style\n",
    "cs <- CellStyle(wb)+Alignment(h = \"ALIGN_CENTER\")\n",
    "#sheet name vec\n",
    "tmp_vec <- c(\"Micro\"= \"Microglia\",\"Astro\" = \"Astrocyte\",\n",
    "             \"Oligo\" = \"Oligodendrocyte\",\"Neuron\" = \"All Neurons\",\n",
    "             \"CTX-Ex\" = \"Excitatory Neurons\",\"Inh\" = \"Inhibitory Neurons\",\"DG\" = \"DG\")\n",
    "\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"CTX-Ex\",\"Inh\",\"DG\")){\n",
    "  sheet <- createSheet(wb, tmp_vec[[i]])\n",
    "  a_row <- createRow(sheet = sheet, rowIndex = 1)\n",
    "  a_cell <- createCell(row = a_row, colIndex = 1:13)\n",
    "  setCellValue(a_cell[[1,1]], \"8 months\")\n",
    "  setCellValue(a_cell[[1,8]], \"13 months\")\n",
    "  setCellStyle(a_cell[[1,1]], cs)\n",
    "  setCellStyle(a_cell[[1,8]], cs)\n",
    "\n",
    "\n",
    "  addMergedRegion(sheet,1,1,1,6)\n",
    "  addMergedRegion(sheet,1,1,8,13)\n",
    "  addDataFrame(DE_8mo[[paste0(i,\"combined\")]] %>% rename(\"rowname\" = \"gene\"), sheet=sheet, \n",
    "               startColumn=1, startRow=2, row.names=FALSE)\n",
    "  \n",
    "  addDataFrame(DE_13mo[[paste0(i,\"combined\")]] %>% rename(\"rowname\" = \"gene\"), sheet=sheet, \n",
    "               startColumn=8, startRow=2, row.names=FALSE)\n",
    "}\n",
    "\n",
    "saveWorkbook(wb, \"Jun08_DEG_raw.xlsx\")\n",
    "\n",
    "#Export Volcano\n",
    "library(EnhancedVolcano)\n",
    "#starmap_scaled <- starmap_raw\n",
    "#starmap_scaled@assays$RNA@data <- starmap_scaled@assays$raw@data\n",
    "#starmap_scaled <- ad_starmap_normalize(starmap_scaled)\n",
    "plot_list <- list()\n",
    "for(k in c(\"8months\",\"13months\")){\n",
    "  for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\")){\n",
    "    if(i == \"CA\") plot_list[[paste(i,k,sep=\"_\")]] <- ad_D_vs_C_DE_volcano(starmap_scaled,c(\"CA1\",\"CA2\",\"CA3\"),k,Pcap = 10e-50,\n",
    "                                                                          fig_title = i,draw_connector = T)\n",
    "    else if(i == \"Cortex\") plot_list[[paste(i,k,sep=\"_\")]] <- ad_D_vs_C_DE_volcano(starmap_scaled,c(\"CTX-Ex\",\"Inhi\"),\n",
    "                                                                                   k,Pcap = 10e-50, fig_title = i,\n",
    "                                                                                   draw_connector = T)\n",
    "    else if(i == \"Neuron\") plot_list[[paste(i,k,sep=\"_\")]] <- ad_D_vs_C_DE_volcano(starmap_scaled,c(\"CTX-Ex\",\"CA1\",\"CA2\",\"CA3\",\"DG\",\"Inh\"),k, Pcap = 10e-50, \n",
    "                                                                                   fig_title = i,draw_connector = T)\n",
    "    else plot_list[[paste(i,k,sep=\"_\")]] <- ad_D_vs_C_DE_volcano(starmap_scaled,i,k,Pcap = 10e-50, \n",
    "                                                                 fig_title = i,draw_connector = T)\n",
    "    \n",
    "  }\n",
    "}\n",
    "pdf(file = \"Jun07_DE_volcano.pdf\",width = 24,height = 12,useDingbats = FALSE)\n",
    "ggarrange(plotlist = plot_list[c(1:7)],ncol = 4,nrow = 2)\n",
    "ggarrange(plotlist = plot_list[c(8:14)],ncol = 4,nrow = 2)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "#####Find Intersect#####\n",
    "DEG_ls <- list()\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\")){\n",
    "  DEG_ls[[paste0(i,\"_8mo\")]]  <- data.frame(gene = \n",
    "                                 c(intersect(DE_8mo[[paste0(i,1)]] %>% subset(subset = avg_logFC > 0 & p_val < 0.05) %>% .$rowname,\n",
    "                                             DE_8mo[[paste0(i,2)]] %>% subset(subset = avg_logFC > 0 & p_val < 0.05) %>% .$rowname),\n",
    "                                   intersect(DE_8mo[[paste0(i,1)]] %>% subset(subset = avg_logFC < 0 & p_val < 0.05) %>% .$rowname,\n",
    "                                             DE_8mo[[paste0(i,2)]] %>% subset(subset = avg_logFC < 0 & p_val < 0.05) %>% .$rowname)),\n",
    "                                 logFC = 0)\n",
    "  DEG_ls[[paste0(i,\"_8mo\")]]$logFC <- apply(DEG_ls[[paste0(i,\"_8mo\")]],1,\n",
    "                                            FUN = function(x){\n",
    "                                              (DE_8mo[[paste0(i,1)]]$avg_logFC[DE_8mo[[paste0(i,1)]]$rowname == x[[\"gene\"]]] +\n",
    "                                                DE_8mo[[paste0(i,1)]]$avg_logFC[DE_8mo[[paste0(i,1)]]$rowname == x[[\"gene\"]]])/2\n",
    "                                            })\n",
    "  DEG_ls[[paste0(i,\"_13mo\")]] <- data.frame(gene = \n",
    "                                c(intersect(DE_13mo[[paste0(i,1)]] %>% subset(subset = avg_logFC > 0 & p_val < 0.05) %>% .$rowname,\n",
    "                                            DE_13mo[[paste0(i,2)]] %>% subset(subset = avg_logFC > 0 & p_val < 0.05) %>% .$rowname),\n",
    "                                  intersect(DE_13mo[[paste0(i,1)]] %>% subset(subset = avg_logFC < 0 & p_val < 0.05) %>% .$rowname,\n",
    "                                            DE_13mo[[paste0(i,2)]] %>% subset(subset = avg_logFC < 0 & p_val < 0.05) %>% .$rowname)),\n",
    "                                logFC = 0)\n",
    "  DEG_ls[[paste0(i,\"_13mo\")]]$logFC <- apply(DEG_ls[[paste0(i,\"_13mo\")]],1,\n",
    "                                            FUN = function(x){\n",
    "                                              (DE_13mo[[paste0(i,1)]]$avg_logFC[DE_13mo[[paste0(i,1)]]$rowname == x[[\"gene\"]]] +\n",
    "                                                DE_13mo[[paste0(i,1)]]$avg_logFC[DE_13mo[[paste0(i,1)]]$rowname == x[[\"gene\"]]])/2\n",
    "                                            })\n",
    "}\n",
    "# Export rnk\n",
    "# No LogFC filter\n",
    "DE_13mo[[paste0(\"Neuron\",\"combined\")]]<- ad_D_vs_C_DE(starmap_scaled,c(\"CTX-Ex\",\"CA1\",\"CA2\",\"CA3\",\"DG\",\"Inhi\"),\"13months\",\n",
    "                                     logfc_cutoff = 0.05,p_cutoff = 0.05)\n",
    "DE_13mo[[paste0(\"DG\",\"combined\")]]<- ad_D_vs_C_DE(starmap_scaled,c(\"DG\"),\"13months\",\n",
    "                                         logfc_cutoff = 0.05,p_cutoff = 0.05)\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"DG\")){#\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\"\n",
    "  print(i)\n",
    "  tmp_df <- ad_D_vs_C_DE(starmap_scaled,i,\"13months\",p_cutoff = 1, logfc_cutoff = 0)\n",
    "  write_tsv(tmp_df[,c(1,3)] %>% arrange(desc(avg_logFC)),\n",
    "  path = paste(i,\"_13mo_DE_C.rnk\",sep = \"\"),col_names = F)\n",
    "  #write_tsv(DE_8mo[[paste0(i,\"combined\")]][,c(1,3)] %>% arrange(desc(avg_logFC)),\n",
    "  #path = paste(i,\"_8mo_DE_C.rnk\",sep = \"\"),col_names = F)\n",
    "}\n",
    "\n",
    "#plot Heatmap\n",
    "gsea_res <- list()\n",
    "\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"DG\")){#\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\"\n",
    "  for(j in c(13)){\n",
    "    group_name <- \"C\"\n",
    "      tmp_vec <- list.dirs(paste0(\"/Users/zhouhaowen/Desktop/At_Broad/AD/Pathway_analysis/Jun08_GSEA_\",group_name),recursive = F)\n",
    "      k <- tmp_vec[str_detect(tmp_vec,paste0(i,\"_\",j))] %>% str_split(pattern = \"\\\\.\") %>% unlist %>% .[3]\n",
    "    gsea_res[[group_name]][[paste(i,j,sep = \"_\")]] <- rbind(read_tsv(file = paste0(tmp_vec[str_detect(tmp_vec,paste0(i,\"_\",j))],\"/gsea_report_for_na_pos_\",k,\".tsv\")),\n",
    "                             read_tsv(file = paste0(tmp_vec[str_detect(tmp_vec,paste0(i,\"_\",j))],\"/gsea_report_for_na_neg_\",k,\".tsv\"))) %>% subset(`NOM p-val` < 0.05) %>% arrange(desc(NES))\n",
    "    gsea_res[[group_name]][[paste(i,j,sep = \"_\")]][[\"term_size\"]] <- apply(gsea_res[[group_name]][[paste(i,j,sep = \"_\")]],\n",
    "                                                                           1,\n",
    "                                                                           FUN = function(x){\n",
    "                                                                             if(x[[\"NAME\"]] %in% myGO$GOID){\n",
    "                                                                             myGO$Term_size[myGO$GOID == x[[\"NAME\"]]]\n",
    "                                                                             }else -1\n",
    "                                                                           })\n",
    "    \n",
    "  }\n",
    "}\n",
    "\n",
    "top_GO <- c()\n",
    "#,\"Cortex\",\"CA\",\"DG\"\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"DG\")){\n",
    "  tmp_vec1 <- gsea_res[[\"C\"]][[paste(i,13,sep = \"_\")]] %>% subset(subset = `NOM p-val` < 0.05 & term_size < 1000\n",
    "                                                                  & term_size > 10) %>% \n",
    "    arrange(`NOM p-val`) %>% .$NAME\n",
    "  #tmp_vec2 <- gsea_res[[\"C\"]][[paste(i,8,sep = \"_\")]] %>% subset(subset = `NOM p-val` < 0.05) %>% .$NAME\n",
    "  #top_GO <- c(top_GO,union(tmp_vec1,tmp_vec2)) %>% unique()\n",
    "  top_GO <- c(top_GO,tmp_vec1 %>% .[1:min(20,length(.))] ) %>% unique()\n",
    "}\n",
    "\n",
    "top_GO_df <- data.frame(name = top_GO)\n",
    "\n",
    "for (k in seq_len(nrow(top_GO_df))) {\n",
    "  for(i in c(\"CTX-Ex\",\"Inh\",\"CA1\",\"CA2\",\"CA3\",\"DG\")){\n",
    "    #if(sum(as.character(gsea_res[[\"C\"]][[paste(i,8,sep = \"_\")]]$NAME) == top_GO_df$name[k]) == 0){\n",
    "    #  top_GO_df[[paste0(i,8)]][k] <- 0\n",
    "    #}else{\n",
    "    #  top_GO_df[[paste0(i,8)]][k] <- gsea_res[[\"C\"]][[paste(i,8,sep = \"_\")]]$NES[as.character(gsea_res[[\"C\"]][[paste(i,8,sep = \"_\")]]$NAME) == top_GO_df$name[k]]\n",
    "    #}\n",
    "    \n",
    "    if(sum(as.character(gsea_res[[\"C\"]][[paste0(i,\"_13\")]]$NAME) == top_GO_df$name[k]) == 0){\n",
    "      top_GO_df[[paste0(i,\"_13\")]][k] <- 0\n",
    "    }else{\n",
    "      top_GO_df[[paste0(i,\"_13\")]][k] <- gsea_res[[\"C\"]][[paste0(i,\"_13\")]]$NES[as.character(gsea_res[[\"C\"]][[paste0(i,\"_13\")]]$NAME) == top_GO_df$name[k]]\n",
    "    }\n",
    "    \n",
    "  }\n",
    "}\n",
    "\n",
    "top_GO_df[[\"term\"]] <- apply(top_GO_df,1,\n",
    "                             FUN = function(x){\n",
    "                               myGO$Description[myGO$GOID == x[[\"name\"]]]\n",
    "                             })\n",
    "\n",
    "Heatmap(as.matrix(as.data.frame(top_GO_df[,2:7],row.names = as.character(top_GO_df$term))),\n",
    "        cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = T,\n",
    "        border_gp = gpar(col = \"black\", lty = 1, lwd = 2),rect_gp = gpar(col = \"black\", lty = 1, lwd = 1),\n",
    "        column_title = paste(8,13,\"mo\"), \n",
    "        col = colorRamp2(c(quantile(top_GO_df[,2:7] %>% unlist,probs = c(0.01,0.99)),0), c(\"blue\", \"red\", \"white\")),\n",
    "        row_names_max_width = max_text_width(as.character(top_GO_df$term), gp = gpar(fontsize = 12))) \n",
    "\n",
    "#For overall SDEG\n",
    "gostres_ls[[\"All_8mo_SDEG\"]] <-  gost(query = SDEG_ls[[\"8mo_all\"]] %>% \n",
    "                                                 subset(p_val_adj_A < 0.01 & p_val_adj_B < 0.01 & \n",
    "                                                          max_int <= 3) %>% .$gene, organism = \"mmusculus\")\n",
    "gostres_ls[[\"All_13mo_SDEG\"]] <-  gost(query = SDEG_ls[[\"13mo_all\"]] %>% \n",
    "                                                 subset(p_val_adj_A < 0.01 & p_val_adj_B < 0.01 & \n",
    "                                                          max_int <= 3) %>% .$gene, organism = \"mmusculus\")\n",
    "\n",
    "plot_list <- list()\n",
    "for(i in c(\"8mo\",\"13mo\")){\n",
    "  plot_list[[paste0(i,\"_all\")]] <- ggplot(data = gostres_ls[[paste0(\"All_\",i,\"_SDEG\")]][[\"result\"]] %>% \n",
    "                                   subset(p_value < 0.01 & term_size < 1000 & term_size > 10 & source == \"GO:BP\") %>%\n",
    "                                   arrange(desc(p_value)) %>% \n",
    "                                   mutate(term_name = factor(term_name,levels = term_name)),\n",
    "                                 aes(y = term_name, x = -log10(p_value))) +\n",
    "    geom_bar(stat = \"identity\", fill = \"darkblue\") + \n",
    "    theme_classic() + labs(title = i)\n",
    "}\n",
    "\n",
    "#Export For multi-group \n",
    "setwd(\"Mar02_rerun/GSEA/rnk/\")\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\")){\n",
    "  print(i)\n",
    "  write_tsv(data.frame(names = DE_13mo[[i]] %>% \n",
    "  arrange(desc(avg_logFC)) %>% .$rowname %>% str_to_title(),\n",
    "                     logfc = DE_13mo[[i]] %>% arrange(desc(avg_logFC)) %>% .$avg_logFC),\n",
    "  path = paste(i,\"_13mo_DE_2M.rnk\",sep = \"\"),col_names = F)\n",
    "  write_tsv(data.frame(names = DE_8mo[[i]] %>% \n",
    "  arrange(desc(avg_logFC)) %>% .$rowname %>% str_to_title(),\n",
    "                     logfc = DE_8mo[[i]] %>% arrange(desc(avg_logFC)) %>% .$avg_logFC),\n",
    "  path = paste(i,\"_8mo_DE_2M.rnk\",sep = \"\"),col_names = F)\n",
    "}\n",
    "\n",
    "#GSEA Heatmap\n",
    "gsea_res <- list()\n",
    "\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\")){\n",
    "  for(j in c(8,13)){\n",
    "    for(group_name in c(\"0MS\",\"1MS\",\"2MS\")){\n",
    "      tmp_vec <- list.dirs(paste0(\"/Users/zhouhaowen/Desktop/At_Broad/AD/Mar02_rerun/GSEA/res_\",group_name),recursive = F)\n",
    "      k <- tmp_vec[str_detect(tmp_vec,paste0(i,\"_\",j))] %>% str_split(pattern = \"\\\\.\") %>% unlist %>% .[3]\n",
    "    gsea_res[[group_name]][[paste(i,j,sep = \"_\")]] <- rbind(read_tsv(file = paste0(tmp_vec[str_detect(tmp_vec,paste0(i,\"_\",j))],\"/gsea_report_for_na_pos_\",k,\".tsv\")),\n",
    "                             read_tsv(file = paste0(tmp_vec[str_detect(tmp_vec,paste0(i,\"_\",j))],\"/gsea_report_for_na_neg_\",k,\".tsv\"))) %>% subset(`NOM p-val` < 0.05) %>% arrange(desc(NES))\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\",\"Cortex\",\"CA\",\"DG\")){\n",
    "  for(j in c(8,13)){\n",
    "    tmp_vec1 <- gsea_res[[\"1MS\"]][[paste(i,j,sep = \"_\")]] %>% subset(subset = `NOM p-val` < 0.05) %>% .$NAME\n",
    "    tmp_vec2 <- gsea_res[[\"0MS\"]][[paste(i,j,sep = \"_\")]] %>% subset(subset = `NOM p-val` < 0.05) %>% .$NAME\n",
    "    tmp_vec3 <- gsea_res[[\"2MS\"]][[paste(i,j,sep = \"_\")]] %>% subset(subset = `NOM p-val` < 0.05) %>% .$NAME\n",
    "    paste(i,j,\"|\",length(tmp_vec1),length(tmp_vec2),length(tmp_vec3),\"|\",\n",
    "          length(intersect(tmp_vec1,intersect(tmp_vec2,tmp_vec3))),length(intersect(tmp_vec1,tmp_vec2))) %>% message()\n",
    "  }\n",
    "}\n",
    "\n",
    "top_GO <- c()\n",
    "j = 13\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\")){\n",
    "  \n",
    "  #tmp_vec1 <- gsea_res[[\"2MS\"]][[paste(i,j,sep = \"_\")]] %>% subset(subset = `NOM p-val` < 0.05) %>% .$NAME\n",
    "  #tmp_vec2 <- gsea_res[[\"2MS\"]][[paste(i,j,sep = \"_\")]] %>% subset(subset = `NOM p-val` < 0.05) %>% .$NAME\n",
    "  #tmp_vec3 <- gsea_res[[\"2MS\"]][[paste(i,j,sep = \"_\")]] %>% subset(subset = `NOM p-val` < 0.05) %>% .$NAME\n",
    "  tmp_df <- gsea_res[[\"C\"]][[paste(i,j,sep = \"_\")]] #%>% \n",
    "    #subset(subset = NAME %in% intersect(tmp_vec1,tmp_vec2)) %>% arrange(`NOM p-val`)\n",
    "  #if(dim(tmp_df)[1] > 20){tmp_df <- tmp_df[1:20,]}\n",
    "  top_GO <- c(top_GO,tmp_df$NAME) %>% unique()\n",
    "}\n",
    "\n",
    "top_GO_df <- data.frame(name = top_GO,\n",
    "                        Micro1 = 0,Micro2 = 0,Astro1 = 0,Astro2 = 0,\n",
    "                        Oligo1 = 0,Oligo2 = 0,Neuron1 = 0,Neuron2 = 0,\n",
    "                        Cortex1 = 0, Cortex2 = 0, CA1 = 0, CA2 = 0, DG1 = 0, DG2 = 0)\n",
    "for(k in 1:dim(top_GO_df)[1]) {\n",
    "  for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\")){\n",
    "    if(sum(as.character(gsea_res[[\"C\"]][[paste(i,j,sep = \"_\")]]$NAME) == top_GO_df$name[k]) == 0){\n",
    "      top_GO_df[[paste0(i,1)]][k] <- 0\n",
    "    }else{\n",
    "      top_GO_df[[paste0(i,1)]][k] <- gsea_res[[\"C\"]][[paste(i,j,sep = \"_\")]]$NES[as.character(gsea_res[[\"C\"]][[paste(i,j,sep = \"_\")]]$NAME) == top_GO_df$name[k]]\n",
    "    }\n",
    "\n",
    "    \n",
    "  }\n",
    "}\n",
    "\n",
    "top_GO_df[[\"term\"]] <- apply(top_GO_df,1,\n",
    "                             FUN = function(x){\n",
    "                               myGO$Description[myGO$GOID == x[[\"name\"]]]\n",
    "                             })\n",
    "\n",
    "require(gplots)\n",
    "require(viridis)\n",
    "require(RColorBrewer)\n",
    "require(circlize)\n",
    "pdf(file = \"F7S_8mo_GO_heatmap.pdf\",width = 8,height = 12,useDingbats = FALSE)\n",
    "Heatmap(as.matrix(as.data.frame(top_GO_df[,c(2,4,6,8)],row.names = as.character(top_GO_df$term))),\n",
    "        cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = T,\n",
    "        border_gp = gpar(col = \"black\", lty = 1, lwd = 2),rect_gp = gpar(col = \"black\", lty = 1, lwd = 1),\n",
    "        column_title = paste(j,\"mo\"), \n",
    "        col = colorRamp2(c(quantile(top_GO_df[,c(2,4,6,8)] %>% unlist,probs = c(0.01,0.99)),0), c(\"blue\", \"red\", \"white\")),\n",
    "        row_names_max_width = max_text_width(as.character(top_GO_df$term), gp = gpar(fontsize = 12))) \n",
    "\n",
    "\n",
    "dev.off()\n",
    "\n",
    "#Export bar plot\n",
    "plot_list <- list()\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"DG\")){\n",
    "  tmp_df <- gsea_res[[\"C\"]][[paste0(i,\"_13\")]] %>% subset(subset = `NOM p-val` < 0.01  & term_size < 1000\n",
    "                                                                  & term_size > 10)\n",
    "  tmp_df$NAME <- apply(tmp_df,1,FUN = function(x){myGO$Description[myGO$GOID == x[[\"NAME\"]]]})\n",
    "  plot_list[[i]] <- ggplot(data = tmp_df,aes(x = reorder(NAME,ES), y = ES, fill = ES < 0))+\n",
    "    geom_bar(stat = \"identity\")+\n",
    "    coord_flip() +\n",
    "    labs(x = \"GO Terms\", y = \"Enrichment Score\",\n",
    "         title = paste(\"GO:BP\",i)) +\n",
    "    theme_minimal()\n",
    "}\n",
    "\n",
    "pdf(file = \"Jun08_MAOD_GSEA_Barplot_0_01.pdf\",width = 8,height = 12,useDingbats = FALSE)\n",
    "plot_list %>% print()\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bde7642b",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## gProfiler\n",
    "### DEG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e564c87e",
   "metadata": {
    "lines_to_next_cell": 0,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "######Using gProfiler######\n",
    "\n",
    "library(gprofiler2)\n",
    "gostres_ls <- list()\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"DG\")){\n",
    "  message(i)\n",
    "  gostres_ls[[paste0(i,\"_8mo_up\")]] <- gost(query = intersect(DE_8mo[[paste0(i,\"1\")]] %>% subset(avg_logFC > 0) %>% .$rowname,\n",
    "                                                              DE_8mo[[paste0(i,\"2\")]] %>% subset(avg_logFC > 0) %>% .$rowname),\n",
    "                                            organism = \"mmusculus\", custom_bg = row.names(starmap_raw))\n",
    "  gostres_ls[[paste0(i,\"_8mo_down\")]] <- gost(query = intersect(DE_8mo[[paste0(i,\"1\")]] %>% subset(avg_logFC < 0) %>% .$rowname,\n",
    "                                                                DE_8mo[[paste0(i,\"2\")]] %>% subset(avg_logFC < 0) %>% .$rowname),\n",
    "                                            organism = \"mmusculus\", custom_bg = row.names(starmap_raw))\n",
    "  gostres_ls[[paste0(i,\"_13mo_up\")]] <- gost(query = intersect(DE_13mo[[paste0(i,\"1\")]] %>% subset(avg_logFC > 0) %>% .$rowname,\n",
    "                                                               DE_13mo[[paste0(i,\"2\")]] %>% subset(avg_logFC > 0) %>% .$rowname),\n",
    "                                             organism = \"mmusculus\", custom_bg = row.names(starmap_raw))\n",
    "  gostres_ls[[paste0(i,\"_13mo_down\")]] <- gost(query = intersect(DE_13mo[[paste0(i,\"1\")]] %>% subset(avg_logFC < 0) %>% .$rowname,\n",
    "                                                                 DE_13mo[[paste0(i,\"2\")]] %>% subset(avg_logFC < 0) %>% .$rowname),\n",
    "                                             organism = \"mmusculus\", custom_bg = row.names(starmap_raw))\n",
    "}\n",
    "\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"DG\")){\n",
    "  message(i)\n",
    "  #gostres_ls[[paste0(i,\"_8mo_up\")]] <- gost(query = DE_8mo[[paste0(i,\"combined\")]] %>% subset(avg_logFC > 0) %>% .$rowname,\n",
    "  #                                          organism = \"mmusculus\", custom_bg = row.names(starmap_raw))\n",
    "  #gostres_ls[[paste0(i,\"_8mo_down\")]] <- gost(query = DE_8mo[[paste0(i,\"combined\")]] %>% subset(avg_logFC < 0) %>% .$rowname,\n",
    "  #                                          organism = \"mmusculus\", custom_bg = row.names(starmap_raw))\n",
    "  gostres_ls[[paste0(i,\"_13mo\")]] <- gost(query = DE_13mo[[paste0(i,\"combined\")]] %>% .$rowname, organism = \"mmusculus\")\n",
    "  #gostres_ls[[paste0(i,\"_13mo_down\")]] <- gost(query = DE_13mo[[paste0(i,\"combined\")]] %>% subset(avg_logFC < 0) %>% .$rowname,\n",
    "  #                                           organism = \"mmusculus\")\n",
    "}\n",
    "#Barplot\n",
    "top_GO <- c()\n",
    "\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"DG\")){\n",
    "  for(k in c(\"\")){\n",
    "      j = paste0(\"_13mo\",k)\n",
    "      if(!is.null(gostres_ls[[paste0(i,j)]])){\n",
    "        tmp_vec <- gostres_ls[[paste0(i,j)]][[\"result\"]] %>% arrange(p_value,precision) %>%\n",
    "                      subset(p_value < 0.01 & term_size < 1000 & term_size > 10 & source == \"GO:BP\") %>% .$term_id\n",
    "        message(paste(paste0(i,j), length(tmp_vec)))\n",
    "        if(length(tmp_vec) > 0) top_GO <- c(top_GO, tmp_vec %>% .[1:min(30,length(.))]) %>% unique()\n",
    "      }\n",
    "  }\n",
    "}\n",
    "top_GO_df <- data.frame(GOID = top_GO)\n",
    "top_GO_df$name <- apply(top_GO_df,1,FUN = function(x){myGO$Description[myGO$GOID == x[[\"GOID\"]]]})\n",
    "\n",
    "for(i in c(\"CTX-Ex\",\"Inh\",\"CA1\",\"CA2\",\"CA3\",\"DG\")){\n",
    "  for(k in c(\"up\",\"down\")){\n",
    "    j = paste0(\"_13mo_\",k)\n",
    "    top_GO_df[[paste0(i,j)]] <- 0\n",
    "    top_GO_df[[paste0(i,j)]] <- apply(top_GO_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(x[[\"GOID\"]] %in% gostres_ls[[paste0(i,j)]][[\"result\"]][[\"term_id\"]]){\n",
    "                              if(k == \"up\") -log10(gostres_ls[[paste0(i,j)]][[\"result\"]]$p_value[gostres_ls[[paste0(i,j)]][[\"result\"]][[\"term_id\"]] == x[[\"GOID\"]]])\n",
    "                              else log10(gostres_ls[[paste0(i,j)]][[\"result\"]]$p_value[gostres_ls[[paste0(i,j)]][[\"result\"]][[\"term_id\"]] == x[[\"GOID\"]]])\n",
    "                            }else 0#as.numeric(x[[paste0(i,j)]])\n",
    "                          })\n",
    "  }\n",
    "\n",
    "}\n",
    "require(circlize)\n",
    "\n",
    "plot_list[[paste0(\"_\",k,\"mo_SDEG\")]] <- \n",
    "  Heatmap(as.matrix(as.data.frame(top_GO_df[,3:14],row.names = as.character(top_GO_df$name))),\n",
    "        cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = T,\n",
    "        border_gp = gpar(col = \"black\", lty = 1, lwd = 2),rect_gp = gpar(col = \"black\", lty = 1, lwd = 1),\n",
    "        column_title = paste(k,\"mo\"), \n",
    "        col = colorRamp2(c(quantile(top_GO_df[,3:14] %>% unlist,probs = c(0.01, 0.99)),0), c(\"blue\", \"red\", \"white\")),\n",
    "        row_names_max_width = max_text_width(as.character(top_GO_df$name), gp = gpar(fontsize = 12))) \n",
    "\n",
    "#Barplot\n",
    "plot_list <- list()\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"DG\")){\n",
    "  \n",
    "  #tmp_df <- rbind(gostres_ls[[paste0(i,\"_13mo\")]]$result %>% subset(p_value < 0.01 & term_size < 1000 & \n",
    "  #                                                                       term_size > 10 & source == \"GO:BP\") %>%\n",
    "  #                  arrange(p_value) %>% mutate(logp = -1 * log10(p_value))) %>% arrange(logp)\n",
    "  #tmp_df <- tmp_df[!is.na(tmp_df$query),]\n",
    "  #write.csv(tmp_df[,c(3,4,5,6,11)],paste0(\"Jun08_\",i,\"_gprofiler_res.csv\"))\n",
    "  #Load External\n",
    "  tmp_df <- read.csv(file = paste0(\"/Users/zhouhaowen/Dropbox/mAD/GO/\",i,\"_gprofiler.csv\"))\n",
    "  tmp_df$logp <- -log10(tmp_df$p_value)\n",
    "  plot_list[[i]] <- ggplot(data = tmp_df,aes(x = reorder(term_name,logp), y = logp, fill = logp < 0))+\n",
    "    geom_bar(stat = \"identity\")+\n",
    "    coord_flip() +\n",
    "    labs(x = \"GO Terms\", y = \"-log10 P-value\",\n",
    "         title = paste(\"GO:BP\",i)) +\n",
    "    theme_minimal()\n",
    "}\n",
    "\n",
    "pdf(file = \"Jun08_MAOD_gprofiler_Barplot_no_bg.pdf\",width = 8,height = 12,useDingbats = FALSE)\n",
    "plot_list %>% print()\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8a0d306",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### SDEG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16447481",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#SDEG gprofiler\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  message(i)\n",
    "  gostres_ls[[paste0(i,\"_8mo_SDEG\")]] <-  gost(query = SDEG_ls[[\"8mo\"]] %>% \n",
    "                                                 subset(p_val < 0.01 & \n",
    "                                                          max_int <= 3 & group == i &\n",
    "                                                          (group %in% c(\"Micro\") & max_int != 3  | (group %in% c(\"Astro\",\"Oligo\") & max_int != 1))) %>% .$gene, organism = \"mmusculus\")\n",
    "  \n",
    "  gostres_ls[[paste0(i,\"_13mo_SDEG\")]] <- gost(query = SDEG_ls[[\"13mo\"]] %>% \n",
    "                                                 subset(p_val < 0.01 &\n",
    "                                                          max_int <= 3 & group == i &\n",
    "                                                          (group %in% c(\"Micro\") | (group %in% c(\"Astro\",\"Oligo\") & max_int != 1))) %>% .$gene, organism = \"mmusculus\")\n",
    "}\n",
    "#Barplot \n",
    "\n",
    "plot_list <- list()\n",
    "\n",
    "top_GO <- c()\n",
    "k = 8\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  j = paste0(\"_\",k,\"mo_SDEG\")\n",
    "  if(!is.null(gostres_ls[[paste0(i,j)]])){\n",
    "    tmp_vec <- gostres_ls[[paste0(i,j)]][[\"result\"]] %>% arrange(p_value,precision) %>%\n",
    "                  subset(p_value < 0.01 & term_size < 1000 & term_size > 10 & source == \"GO:BP\") %>% .$term_id\n",
    "    message(paste(paste0(i,j), length(tmp_vec)))\n",
    "    if(length(tmp_vec) > 0) top_GO <- c(top_GO, tmp_vec %>% .[1:min(30,length(.))]) %>% unique()\n",
    "  }\n",
    "}\n",
    "\n",
    "top_GO_df <- data.frame(GOID = top_GO)\n",
    "top_GO_df$name <- apply(top_GO_df,1,FUN = function(x){myGO$Description[myGO$GOID == x[[\"GOID\"]]]})\n",
    "\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  j = paste0(\"_\",k,\"mo_SDEG\")\n",
    "  top_GO_df[[i]] <- apply(top_GO_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(x[[\"GOID\"]] %in% gostres_ls[[paste0(i,j)]][[\"result\"]][[\"term_id\"]]){\n",
    "                              -log10(gostres_ls[[paste0(i,j)]][[\"result\"]]$p_value[gostres_ls[[paste0(i,j)]][[\"result\"]][[\"term_id\"]] == x[[\"GOID\"]]])\n",
    "                            }else 0\n",
    "                          })\n",
    "}\n",
    "require(circlize)\n",
    "\n",
    "plot_list[[paste0(\"_\",k,\"mo_SDEG\")]] <- \n",
    "  Heatmap(as.matrix(as.data.frame(top_GO_df[,3:5],row.names = as.character(top_GO_df$name))),\n",
    "        cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = T,\n",
    "        border_gp = gpar(col = \"black\", lty = 1, lwd = 2),rect_gp = gpar(col = \"black\", lty = 1, lwd = 1),\n",
    "        column_title = paste(k,\"mo\"), \n",
    "        col = colorRamp2(c(quantile(top_GO_df[,3:5] %>% unlist,probs = c(0.99)),0), c( \"red\", \"white\")),\n",
    "        row_names_max_width = max_text_width(as.character(top_GO_df$name), gp = gpar(fontsize = 12))) \n",
    "\n",
    "#using gProfiler on gene clustering \n",
    "\n",
    "\n",
    "for(i in c(\"8mo\",\"13mo\")){\n",
    "  for(j in tmp_list[[i]]$cluster%>% as.factor() %>% levels()){\n",
    "    message(paste(i,j))\n",
    "    gostres_ls[[paste0(i,\"_\",j)]] <- gost(query = tmp_list[[i]]$gene[tmp_list[[i]]$cluster == j],\n",
    "                                            organism = \"mmusculus\")\n",
    "  }\n",
    "}\n",
    "\n",
    "plot_list <- list()\n",
    "\n",
    "\n",
    "tmp_vec1 <- c(\"8mo\",\"13mo\")\n",
    "tmp_vec2 <- seq_vec[1:4] %>% as.character()\n",
    "size_limit <- F\n",
    "for(i in tmp_vec1){\n",
    "  top_GO <- c()\n",
    "  for(j in tmp_vec2){\n",
    "    if(!is.null(gostres_ls[[paste0(i,\"_\",j)]])){\n",
    "      tmp_vec <- gostres_ls[[paste0(i,\"_\",j)]][[\"result\"]] %>% arrange(p_value,precision) %>%\n",
    "                    subset(p_value < 0.01 & term_size < 1000 & term_size > 10 & source == \"GO:BP\") %>% .$term_id\n",
    "      message(paste(paste0(i,\"_\",j), length(tmp_vec)))\n",
    "      if(length(tmp_vec) > 0 & size_limit) top_GO <- c(top_GO, tmp_vec %>% .[1:min(30,length(.))]) %>% unique()\n",
    "      else if(length(tmp_vec) > 0) top_GO <- c(top_GO, tmp_vec) %>% unique()\n",
    "    }\n",
    "  }\n",
    "  top_GO_df <- data.frame(GOID = top_GO)\n",
    "  top_GO_df$name <- apply(top_GO_df,1,FUN = function(x){myGO$Description[myGO$GOID == x[[\"GOID\"]]]})\n",
    "  for(j in tmp_vec2){\n",
    "    top_GO_df[[j]] <- apply(top_GO_df,1,\n",
    "                            FUN = function(x){\n",
    "                              if(x[[\"GOID\"]] %in% gostres_ls[[paste0(i,\"_\",j)]][[\"result\"]][[\"term_id\"]]){\n",
    "                              -log10(gostres_ls[[paste0(i,\"_\",j)]][[\"result\"]]$p_value[gostres_ls[[paste0(i,\"_\",j)]][[\"result\"]][[\"term_id\"]] == x[[\"GOID\"]]])\n",
    "                              }else 0\n",
    "                          })\n",
    "  }\n",
    "  require(circlize)\n",
    "  tmp <- quantile(top_GO_df[,3:(2+length(tmp_vec2))] %>% unlist,probs = c(0.99))\n",
    "  plot_list[[paste0(i,\"_\",j,\"_GO\")]] <- \n",
    "  Heatmap(as.matrix(as.data.frame(top_GO_df[,3:(2+length(tmp_vec2))],row.names = as.character(top_GO_df$name))),\n",
    "        cluster_columns = F,cluster_rows  = T, show_row_names = size_limit, show_row_dend = T,\n",
    "        border_gp = gpar(col = \"black\", lty = 1, lwd = 2),#rect_gp = gpar(col = \"black\", lty = 1, lwd = 1),\n",
    "        column_title = paste(i), \n",
    "        col = colorRamp2(c(tmp,tmp*0.6,-log10(0.001),0), c(\"#fcffa4\",\"#ed6925\",\"#781c6d\", \"#000004\")),\n",
    "        row_names_max_width = max_text_width(as.character(top_GO_df$name), gp = gpar(fontsize = 12))) \n",
    "}\n",
    "\n",
    "\n",
    "pdf(paste(\"Apr13\",\"gene_cl_GO.pdf\", sep = \"_\"),width = 8,height = 16)\n",
    "plot_list %>% draw()\n",
    "dev.off()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f629a34",
   "metadata": {},
   "source": [
    "## Stat. Test\n",
    "\n",
    "### Func."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "315d9bcb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#####Test Para\n",
    "res_ls <- starmap_raw@misc$ADmouse_9494_plaque_Cortex\n",
    "sample_name <- \"ADmouse_9494\"\n",
    "region_size <- sum(starmap_raw@misc$ADmouse_9494_morph$region == 3)\n",
    "region <- \"Hippocampus\"\n",
    "top_level_list <- c(\"CTX-Ex\", \"Endo\", \"Inh\", \"Micro\", \"Astro\", \"OPC\",\"Oligo\", \"SMC\") \n",
    "\n",
    "res_ls <- starmap_64@misc$AD_mouse9721_plaque_all\n",
    "sample_name <- \"AD_mouse9721\"\n",
    "region_size <-  dim(starmap_64@misc[[paste0(sample_name,\"_morph\")]]$plaque)[1] *\n",
    "  dim(starmap_64@misc[[paste0(sample_name,\"_morph\")]]$plaque)[2]\n",
    "region <- \"all\"\n",
    "top_level_list <- c(\"CTX-Ex\", \"CA1\",\"CA2\",\"CA3\", \"DG\",\"Endo\", \"Inh\", \"Micro\", \"Astro\", \"OPC\",\"Oligo\", \"SMC\")\n",
    "#####Func1\n",
    "ad_distr_stat <- function(res_ls, sample_name, region_size, cell_type_list, top_level_list, region = \"all\"){\n",
    "  dilate_ls <- res_ls[[paste0(sample_name,\"_plaque_dilate_info\")]]\n",
    "  cell_meta <- res_ls[[paste0(sample_name,\"_cell_meta\")]]\n",
    "  plaque_meta <- res_ls[[paste0(sample_name,\"_plaque_meta\")]]\n",
    "  dilate_area <- res_ls[[paste0(sample_name,\"_plaque_dilate_area\")]]\n",
    "  cell_meta <- subset(cell_meta, top_level %in% top_level_list)\n",
    "  return_ls <- list()\n",
    "  return_ls[[\"avg_stat\"]] <- list()\n",
    "  #Top level\n",
    "  tmp_vec <- setNames(sapply(top_level_list,function(x){1000000* sum(cell_meta$top_level == x)/(region_size/3.3^2)}), \n",
    "                      top_level_list)#avg density\n",
    "\n",
    "  tmp_df <- data.frame(plaque_id = rep(plaque_meta$plaque_id, each = length(top_level_list)),\n",
    "                       top_level = rep(top_level_list, nrow(plaque_meta)))\n",
    "  for(i in 1:5){\n",
    "    tmp_df[[paste0(\"R\",i*10)]] <- apply(tmp_df,1,\n",
    "                                        FUN = function(x){\n",
    "                                          if(!as.numeric(x[[\"plaque_id\"]]) %in%\n",
    "                                             dilate_ls[[paste0(i,\"0um_dilate_info\")]]$plaque_id){ NA\n",
    "                                          }else{\n",
    "                                            1000000 * sum(cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                                cell_meta$nearest_plaque == as.numeric(x[[\"plaque_id\"]]) &\n",
    "                                                cell_meta$top_level == as.character(x[[\"top_level\"]])) /                                            (dilate_ls[[paste0(i,\"0um_dilate_info\")]]$s.area[dilate_ls[[paste0(i,\"0um_dilate_info\")]]$plaque_id ==  as.numeric(x[[\"plaque_id\"]])]/3.3^2)\n",
    "                                          }\n",
    "                                        })\n",
    "\n",
    "  }\n",
    "  return_ls[[\"top_level\"]] <- tmp_df\n",
    "  return_ls[[\"avg_stat\"]][[\"top_level\"]] <- tmp_vec\n",
    "  #Sub type\n",
    "  for(i in names(cell_type_list)){\n",
    "      tmp_df <- data.frame(plaque_id = rep(plaque_meta$plaque_id, each = length(cell_type_list[[i]])),\n",
    "                           subtype = rep(cell_type_list[[i]], nrow(plaque_meta)))\n",
    "      for(j in 1:5){\n",
    "        tmp_df[[paste0(\"R\",j*10)]] <- apply(tmp_df,1,\n",
    "                                            FUN = function(x){\n",
    "                                               if(!as.numeric(x[[\"plaque_id\"]]) %in%\n",
    "                                             dilate_ls[[paste0(j,\"0um_dilate_info\")]]$plaque_id) NA\n",
    "                                          else{\n",
    "                                            1000000 * sum(cell_meta$interval == paste0(j,\"0um\") & \n",
    "                                                  cell_meta$nearest_plaque == as.numeric(x[[\"plaque_id\"]]) &\n",
    "                                                  cell_meta$cell_type == as.character(x[[\"subtype\"]])) /                                          (dilate_ls[[paste0(j,\"0um_dilate_info\")]]$s.area[dilate_ls[[paste0(j,\"0um_dilate_info\")]]$plaque_id ==  as.numeric(x[[\"plaque_id\"]])]/3.3^2)\n",
    "                                          }\n",
    "                                            }) %>% unlist\n",
    "      }\n",
    "      \n",
    "      tmp_vec <- setNames(sapply(cell_type_list[[i]],function(x){1000000* sum(cell_meta$cell_type == x)/(region_size/3.3^2)}), \n",
    "                      cell_type_list[[i]])#avg density\n",
    "      return_ls[[i]] <- tmp_df\n",
    "      return_ls[[\"avg_stat\"]][[i]] <- tmp_vec\n",
    "  }\n",
    "  return_ls[[\"cell_meta\"]] <- cell_meta\n",
    "  return_ls[[\"dilate_vec\"]] <- dilate_area\n",
    "  return_ls\n",
    "}\n",
    "#####Test Para\n",
    "sample_list <- c(\"ADmouse_9494\",\"ADmouse_11346\") \n",
    "res_ls <- distr_stat_ls\n",
    "region = \"Hippocampus\"\n",
    "\n",
    "sample_list <- c(\"AD_mouse9721\") \n",
    "res_ls <- distr_stat_ls\n",
    "region = \"all\"\n",
    "\n",
    "#####Func2\n",
    "ad_distr_combined_analysis <- function(sample_list, res_ls, region, cell_type_list){\n",
    "  return_ls <- list()\n",
    "  return_ls[[\"plot_ls\"]] <- list()\n",
    "  return_ls[[\"stat_ls\"]] <- list()\n",
    "  stacked_vln <- function(df, palette_list, title, level, mean_vec){\n",
    "    colnames(df)[colnames(df) == level] <- \"cell_type\"\n",
    "    ggplot(df,aes(variable, value, fill = cell_type)) +\n",
    "          geom_violin(scale = \"width\", adjust = 1, trim = TRUE) +\n",
    "          scale_y_continuous(expand = c(0, 0), position=\"right\", labels = function(x)\n",
    "                             c(rep(x = \"\", times = length(x)-2), x[length(x) - 1], \"\")) +\n",
    "          facet_grid(rows = vars(cell_type) , scales = \"free\", switch = \"y\") +\n",
    "          geom_hline(data = data.frame(cell_type = factor(names(mean_vec)), value = tmp_vec), \n",
    "                     aes(yintercept = value),color = \"green\") + \n",
    "          theme_cowplot(font_size = 12) +\n",
    "          theme(legend.position = \"none\", panel.spacing = unit(0, \"lines\"),\n",
    "                panel.background = element_rect(fill = NA, color = \"black\"),\n",
    "                strip.background = element_blank(),\n",
    "                strip.text = element_text(face = \"bold\"),\n",
    "                strip.text.y.left = element_text(angle = 0),\n",
    "                axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +\n",
    "          scale_fill_manual(values = palette_list,\n",
    "                      breaks = names(palette_list)) + \n",
    "          ggtitle(paste(paste(sample_list,collapse = \" \"), title)) + xlab(\"Interval\") + ylab(\"Density\")\n",
    "  }\n",
    "  \n",
    "  cell_meta <- res_ls[[paste(sample_list[1],region,sep = \"_\")]][[\"cell_meta\"]]\n",
    "  dilate_vec <- res_ls[[paste(sample_list[1],region,sep = \"_\")]][[\"dilate_vec\"]]\n",
    "  tmp_df <- as.data.frame(res_ls[[paste(sample_list[1],region,sep = \"_\")]][[\"top_level\"]])\n",
    "  tmp_vec <- res_ls[[paste(sample_list[1],region,sep = \"_\")]][[\"avg_stat\"]][[\"top_level\"]]\n",
    "  if(length(sample_list) > 1){\n",
    "      for(i in 2:length(sample_list)){\n",
    "        cell_meta <- rbind(cell_meta,res_ls[[paste(sample_list[i],region,sep = \"_\")]][[\"cell_meta\"]])\n",
    "      }\n",
    "      \n",
    "      for(i in 2:length(sample_list)){\n",
    "        dilate_vec <- dilate_vec + res_ls[[paste(sample_list[i],region,sep = \"_\")]][[\"dilate_vec\"]]\n",
    "      }\n",
    "      #Top level\n",
    "      \n",
    "      for(i in 2:length(sample_list)){\n",
    "        tmp_df <- rbind(tmp_df,as.data.frame(res_ls[[paste(sample_list[i],region,sep = \"_\")]][[\"top_level\"]]))\n",
    "      }\n",
    "      \n",
    "      for(i in 2:length(sample_list)){\n",
    "        tmp_vec <- tmp_vec+res_ls[[paste(sample_list[i],region,sep = \"_\")]][[\"avg_stat\"]][[\"top_level\"]]\n",
    "      }\n",
    "  }\n",
    "\n",
    "  tmp_vec <- tmp_vec/length(sample_list)\n",
    "  tmp_list <- list()\n",
    "  tmp_list[[\"perc_df\"]] <- data.frame(top_level = levels(as.factor(tmp_df$top_level)))\n",
    "  tmp_list[[\"dens_df\"]] <- data.frame(top_level = levels(as.factor(tmp_df$top_level)))\n",
    "  tmp_list[[\"stat_c_df\"]] <- data.frame(top_level = levels(as.factor(tmp_df$top_level)))#Chi-sq test\n",
    "  tmp_list[[\"stat_t_df\"]] <- data.frame(top_level = levels(as.factor(tmp_df$top_level)))#one sample t-test\n",
    "  for(i in 1:5){\n",
    "    tmp_list[[\"perc_df\"]][[paste0(\"R\",i*10)]] <- apply(tmp_list[[\"perc_df\"]],1,\n",
    "                                                        FUN = function(x){\n",
    "                                                          sum(cell_meta$top_level == x[[\"top_level\"]] &\n",
    "                                                                cell_meta$interval == paste0(i,\"0um\"))/\n",
    "                                                            sum(cell_meta$interval == paste0(i,\"0um\"))})\n",
    "    tmp_list[[\"dens_df\"]][[paste0(\"R\",i*10)]]  <- apply(tmp_list[[\"dens_df\"]],1,\n",
    "                                                        FUN = function(x){\n",
    "                                                          1000000 * sum(cell_meta$top_level == x[[\"top_level\"]] &\n",
    "                                                                cell_meta$interval == paste0(i,\"0um\"))/\n",
    "                                                            (dilate_vec[[paste0(i,\"0um\")]]/3.3^2) })\n",
    "    \n",
    "    tmp_list[[\"stat_t_df\"]][[paste0(\"R\",i*10)]] <- apply(tmp_list[[\"stat_t_df\"]],1,\n",
    "                                         FUN = function(x){\n",
    "                                           tmp_vec2 <- as.numeric(tmp_df[[paste0(\"R\",i*10)]]) %>% \n",
    "                                             .[tmp_df$top_level == x[[\"top_level\"]]] %>% na.omit()\n",
    "                                           \n",
    "                                           tmp_obj <- t.test(x = tmp_vec2,\n",
    "                                                             mu = tmp_vec[[x[[\"top_level\"]]]], \n",
    "                                                             alternative = ifelse(mean(tmp_vec2) > tmp_vec[[x[[\"top_level\"]]]],\n",
    "                                                                                  \"greater\", \"less\"))\n",
    "                                           #ifelse(mean(tmp_vec2) > tmp_vec[[x[[\"top_level\"]]]], tmp_obj$p.value, \"NA\")\n",
    "                                            tmp_obj$p.value\n",
    "                                         })\n",
    "    \n",
    "    tmp_list[[\"stat_c_df\"]][[paste0(\"R\",i*10)]] <- apply(tmp_list[[\"stat_c_df\"]],1,\n",
    "                                         FUN = function(x){\n",
    "                                           tmp_vec2 <- as.numeric(tmp_df[[paste0(\"R\",i*10)]]) %>% \n",
    "                                             .[tmp_df$top_level == x[[\"top_level\"]]] %>% na.omit()\n",
    "                                           a = sum(cell_meta$top_level == x[[\"top_level\"]] & cell_meta$interval == paste0(i,\"0um\"))\n",
    "                                           b = sum(cell_meta$top_level != x[[\"top_level\"]] & cell_meta$interval == paste0(i,\"0um\"))\n",
    "                                           c = sum(cell_meta$top_level == x[[\"top_level\"]] & cell_meta$interval != paste0(i,\"0um\"))\n",
    "                                           d = sum(cell_meta$top_level != x[[\"top_level\"]] & cell_meta$interval != paste0(i,\"0um\"))\n",
    "                                           \n",
    "                                           tmp_obj <- suppressWarnings(chisq.test(x = matrix(c(a,c,b,d), nrow = 2)))\n",
    "                                           #ifelse(mean(tmp_vec2) > tmp_vec[[x[[\"top_level\"]]]], tmp_obj$p.value, \"NA\")\n",
    "                                            tmp_obj$p.value\n",
    "                                         })\n",
    "  }\n",
    "  #Add overall \n",
    "  tmp_list[[\"dens_df\"]] <- left_join(tmp_list[[\"dens_df\"]],\n",
    "                                     data.frame(top_level = names(tmp_vec),overall = tmp_vec), by = \"top_level\")\n",
    "  tmp_list[[\"perc_df\"]][[\"overall\"]] <- apply(tmp_list[[\"perc_df\"]],1,\n",
    "                                                        FUN = function(x){\n",
    "                                                          sum(cell_meta$top_level == x[[\"top_level\"]])/nrow(cell_meta)})\n",
    "  return_ls[[\"stat_ls\"]][[\"combined\"]] <- tmp_list\n",
    "  #Stacked Barplot\n",
    "  require(reshape2)\n",
    "  require(cowplot)\n",
    "  require(magrittr)\n",
    "  tmp_df %<>% melt(id.vars = c(\"plaque_id\",\"top_level\")) %>% subset(!is.na(value))\n",
    "  for(i in levels(factor(tmp_df$top_level))){\n",
    "    if(sum(tmp_df$value == 0 & tmp_df$top_level == i) / sum(tmp_df$value == 0 & tmp_df$top_level == i) < 0.99){\n",
    "      tmp_df <- tmp_df[tmp_df$top_level != i|tmp_df$value < quantile(tmp_df$value[tmp_df$top_level == i], probs = 0.99),]\n",
    "    }\n",
    "  }\n",
    "  return_ls[[\"plot_ls\"]][[\"combined\"]] <- stacked_vln(tmp_df, top_level_palette_list, \"Top Level\", \"top_level\", tmp_vec)\n",
    "  \n",
    "  for(j in names(cell_type_list)){\n",
    "    \n",
    "    tmp_df <- as.data.frame(res_ls[[paste(sample_list[1],region,sep = \"_\")]][[j]])\n",
    "    tmp_vec <- res_ls[[paste(sample_list[1],region,sep = \"_\")]][[\"avg_stat\"]][[j]]\n",
    "    if(length(sample_list) > 1){\n",
    "          for(i in 2:length(sample_list)){\n",
    "            tmp_df <- rbind(tmp_df,as.data.frame(res_ls[[paste(sample_list[i],region,sep = \"_\")]][[j]]))\n",
    "          }\n",
    "          \n",
    "          for(i in 2:length(sample_list)){\n",
    "            tmp_vec <- tmp_vec+res_ls[[paste(sample_list[i],region,sep = \"_\")]][[\"avg_stat\"]][[j]]\n",
    "          }\n",
    "    }\n",
    "\n",
    "    tmp_vec <- tmp_vec/length(sample_list)\n",
    "    tmp_list <- list()\n",
    "    tmp_list[[\"perc_df\"]]  <- data.frame(subtype = levels(as.factor(tmp_df$subtype)))\n",
    "    tmp_list[[\"dens_df\"]]  <- data.frame(subtype = levels(as.factor(tmp_df$subtype)))\n",
    "    tmp_list[[\"stat_c_df\"]] <- data.frame(subtype = levels(as.factor(tmp_df$subtype)))#Chi-sq test\n",
    "    tmp_list[[\"stat_t_df\"]] <- data.frame(subtype = levels(as.factor(tmp_df$subtype)))#one sample t-test\n",
    "    for(i in 1:5){\n",
    "      tmp_list[[\"perc_df\"]][[paste0(\"R\",i*10)]] <- apply(tmp_list[[\"perc_df\"]],1,\n",
    "                                                        FUN = function(x){\n",
    "                                                          sum(cell_meta$cell_type == x[[\"subtype\"]] &\n",
    "                                                                cell_meta$interval == paste0(i,\"0um\"))/\n",
    "                                                            sum(cell_meta$interval == paste0(i,\"0um\") &\n",
    "                                                                  cell_meta$cell_type %in% cell_type_list[[j]])})\n",
    "      tmp_list[[\"dens_df\"]][[paste0(\"R\",i*10)]]  <- apply(tmp_list[[\"dens_df\"]],1,\n",
    "                                                        FUN = function(x){\n",
    "                                                          1000000 * sum(cell_meta$cell_type == x[[\"subtype\"]] &\n",
    "                                                                cell_meta$interval == paste0(i,\"0um\"))/\n",
    "                                                            (dilate_vec[[paste0(i,\"0um\")]]/3.3^2) })\n",
    "    \n",
    "      tmp_list[[\"stat_t_df\"]][[paste0(\"R\",i*10)]] <- apply(tmp_list[[\"stat_t_df\"]],1,\n",
    "                                         FUN = function(x){\n",
    "                                           tmp_vec2 <- as.numeric(tmp_df[[paste0(\"R\",i*10)]]) %>% \n",
    "                                             .[tmp_df$subtype == x[[\"subtype\"]]] %>% na.omit()\n",
    "                                           \n",
    "                                           tmp_obj <- t.test(x = tmp_vec2,\n",
    "                                                             mu = tmp_vec[[x[[\"subtype\"]]]], \n",
    "                                                             alternative = ifelse(mean(tmp_vec2) > tmp_vec[[x[[\"subtype\"]]]],\n",
    "                                                                                  \"greater\", \"less\"))\n",
    "                                           #ifelse(mean(tmp_vec2) > tmp_vec[[x[[\"subtype\"]]]], tmp_obj$p.value, \"NA\")\n",
    "                                            tmp_obj$p.value\n",
    "                                           \n",
    "                                         })\n",
    "    \n",
    "      tmp_list[[\"stat_c_df\"]][[paste0(\"R\",i*10)]] <- apply(tmp_list[[\"stat_c_df\"]],1,\n",
    "                                         FUN = function(x){\n",
    "                                           tmp_vec2 <- as.numeric(tmp_df[[paste0(\"R\",i*10)]]) %>% \n",
    "                                             .[tmp_df$subtype == x[[\"subtype\"]]] %>% na.omit()\n",
    "                                           a = sum(cell_meta$cell_type == x[[\"subtype\"]] & cell_meta$interval == paste0(i,\"0um\"))\n",
    "                                           b = sum(cell_meta$cell_type != x[[\"subtype\"]] & cell_meta$interval == paste0(i,\"0um\"))\n",
    "                                           c = sum(cell_meta$cell_type == x[[\"subtype\"]] & cell_meta$interval != paste0(i,\"0um\"))\n",
    "                                           d = sum(cell_meta$cell_type != x[[\"subtype\"]] & cell_meta$interval != paste0(i,\"0um\"))\n",
    "                                           \n",
    "                                           tmp_obj <- suppressWarnings(chisq.test(x = matrix(c(a,c,b,d), nrow = 2)))\n",
    "                                           #ifelse(mean(tmp_vec2) > tmp_vec[[x[[\"subtype\"]]]], tmp_obj$p.value, \"NA\")\n",
    "                                            tmp_obj$p.value\n",
    "                                         })\n",
    "    }\n",
    "      #Add overall \n",
    "    tmp_list[[\"dens_df\"]] <- left_join(tmp_list[[\"dens_df\"]],\n",
    "                                     data.frame(subtype = names(tmp_vec),overall = tmp_vec), by = \"subtype\")\n",
    "    tmp_list[[\"dens_df\"]][[\"overall\"]] <- tmp_vec\n",
    "    tmp_list[[\"perc_df\"]][[\"overall\"]] <- apply(tmp_list[[\"perc_df\"]],1,\n",
    "                                                        FUN = function(x){\n",
    "                                                          sum(cell_meta$cell_type == x[[\"subtype\"]])/\n",
    "                                                            sum(cell_meta$cell_type %in% cell_type_list[[j]])})\n",
    "    return_ls[[\"stat_ls\"]][[j]] <- tmp_list\n",
    "    #Stacked Barplot\n",
    "    tmp_df %<>% melt(id.vars = c(\"plaque_id\",\"subtype\")) %>% subset(!is.na(value))\n",
    "    #Skip for CADG\n",
    "    for(i in levels(factor(tmp_df$subtype))){\n",
    "      if(sum(tmp_df$value == 0 & tmp_df$subtype == i)/sum(tmp_df$subtype == i) < 0.99){\n",
    "        tmp_df <- tmp_df[tmp_df$subtype != i|tmp_df$value < quantile(tmp_df$value[tmp_df$subtype == i], probs = 0.99),]\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    return_ls[[\"plot_ls\"]][[j]] <- stacked_vln(tmp_df, cell_type_palette_list[[j]], j, \"subtype\", tmp_vec)\n",
    "  }\n",
    "  return_ls\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c82c50",
   "metadata": {},
   "source": [
    "### Export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ce6f592",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Draw Violin Plot\n",
    "distr_stat_ls <- list()\n",
    "for(i in c(\"ADmouse_9723\",\"ADmouse_9723_2\",\"ADmouse_9494\",\"ADmouse_11346\")){\n",
    "  for(j in c(\"Cortex\",\"Hippocampus\",\"all\")){#\"Cortex\",\"Hippocampus\",\"all\"\n",
    "    message(paste(i,j))\n",
    "    if(j == \"Cortex\"){\n",
    "      top_level_list <- c(\"CTX-Ex\", \"Endo\", \"Inh\", \"Micro\", \"Astro\", \"OPC\",\"Oligo\", \"SMC\")\n",
    "      region_size <- sum(starmap_raw@misc[[paste0(i,\"_morph\")]]$region == 1)\n",
    "    }else if(j == \"Hippocampus\"){\n",
    "      top_level_list <- c(\"CA1\",\"CA2\",\"CA3\", \"Endo\", \"Inh\", \"Micro\", \"Astro\", \"OPC\",\"Oligo\", \"SMC\")\n",
    "      region_size <- sum(starmap_raw@misc[[paste0(i,\"_morph\")]]$region == 3)\n",
    "    }else {\n",
    "        top_level_list <- c(\"CTX-Ex\", \"CA1\",\"CA2\",\"CA3\", \"DG\",\"Endo\", \"Inh\", \"Micro\", \"Astro\", \"OPC\",\"Oligo\", \"SMC\")\n",
    "        region_size <- dim(starmap_raw@misc[[paste0(i,\"_morph\")]]$region)[1] * \n",
    "          dim(starmap_raw@misc[[paste0(i,\"_morph\")]]$region)[2]\n",
    "      }\n",
    "    distr_stat_ls[[paste(i,j,sep = \"_\")]] <- ad_distr_stat(res_ls = starmap_raw@misc[[paste0(i,\"_plaque_\",j)]],\n",
    "                                                           sample_name = i,\n",
    "                                                           region_size = region_size,\n",
    "                                                           cell_type_list = cell_type_list,\n",
    "                                                           top_level_list = top_level_list, region = j)\n",
    "  }\n",
    "}\n",
    "#Combined Analysis\n",
    "comb_al_ls <- list()\n",
    "comb_al_ls[[\"8mo_Cortex\"]] <- ad_distr_combined_analysis(sample_list = c(\"ADmouse_9723\",\"ADmouse_9723_2\"), \n",
    "                                                         res_ls = distr_stat_ls, \n",
    "                                                         region = \"Cortex\", cell_type_list = cell_type_list)\n",
    "\n",
    "comb_al_ls[[\"8mo_Hippocampus\"]] <- ad_distr_combined_analysis(sample_list = c(\"ADmouse_9723\",\"ADmouse_9723_2\"), \n",
    "                                                              res_ls = distr_stat_ls, \n",
    "                                                              region = \"Hippocampus\", cell_type_list = cell_type_list)\n",
    "\n",
    "comb_al_ls[[\"13mo_Cortex\"]] <- ad_distr_combined_analysis(sample_list = c(\"ADmouse_9494\",\"ADmouse_11346\"), \n",
    "                                                          res_ls = distr_stat_ls, \n",
    "                                                          region = \"Cortex\", cell_type_list = cell_type_list)\n",
    "\n",
    "comb_al_ls[[\"13mo_Hippocampus\"]] <- ad_distr_combined_analysis(sample_list = c(\"ADmouse_9494\",\"ADmouse_11346\"), \n",
    "                                                               res_ls = distr_stat_ls, \n",
    "                                                               region = \"Hippocampus\", cell_type_list = cell_type_list)\n",
    "\n",
    "comb_al_ls[[\"8mo_all\"]] <- ad_distr_combined_analysis(sample_list = c(\"ADmouse_9723\",\"ADmouse_9723_2\"), \n",
    "                                                              res_ls = distr_stat_ls, \n",
    "                                                              region = \"all\", cell_type_list = cell_type_list)\n",
    "\n",
    "comb_al_ls[[\"13mo_all\"]] <- ad_distr_combined_analysis(sample_list = c(\"ADmouse_9494\",\"ADmouse_11346\"), \n",
    "                                                          res_ls = distr_stat_ls, \n",
    "                                                          region = \"all\", cell_type_list = cell_type_list)\n",
    "\n",
    "ggarrange(plotlist = comb_al_ls$`8mo_Cortex`$plot_ls, nrow = 2, ncol = 4)\n",
    "ggarrange(plotlist = comb_al_ls$`8mo_Hippocampus`$plot_ls, nrow = 2, ncol = 4)\n",
    "ggarrange(plotlist = comb_al_ls$`13mo_Cortex`$plot_ls, nrow = 2, ncol = 4)\n",
    "ggarrange(plotlist = comb_al_ls$`13mo_Hippocampus`$plot_ls, nrow = 2, ncol = 4)\n",
    "\n",
    "#Export xlsx\n",
    "library(xlsx)\n",
    "wb = createWorkbook()\n",
    "#top level\n",
    "sheet <- createSheet(wb, \"TauPS2APP(2766 genes, majot type)\")\n",
    "#8mo \n",
    "for(i in c(\"8mo\",\"13mo\")){\n",
    "  tmp_df <- cbind(rbind(data.frame(region = \"all\",comb_al_ls[[paste0(i,\"_all\")]]$stat_ls$combined$perc_df),\n",
    "                       data.frame(region = \"Cortex\",comb_al_ls[[paste0(i,\"_Cortex\")]]$stat_ls$combined$perc_df),\n",
    "                       data.frame(region = \"Hippocampus\",comb_al_ls[[paste0(i,\"_Hippocampus\")]]$stat_ls$combined$perc_df)))\n",
    "\n",
    "  tmp_df[3:8] <- tmp_df[3:8] * 100#portion to percentage\n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=ifelse(i == \"8mo\",1, nrow(tmp_df)+3), row.names=FALSE)\n",
    "  \n",
    "  tmp_df <- cbind(rbind(data.frame(comb_al_ls[[paste0(i,\"_all\")]]$stat_ls$combined$dens_df),\n",
    "                        data.frame(comb_al_ls[[paste0(i,\"_Cortex\")]]$stat_ls$combined$dens_df),\n",
    "                        data.frame(comb_al_ls[[paste0(i,\"_Hippocampus\")]]$stat_ls$combined$dens_df)))\n",
    "  \n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=10, startRow=ifelse(i == \"8mo\",1, nrow(tmp_df)+3), row.names=FALSE)\n",
    "\n",
    "  tmp_df <- cbind(rbind(data.frame(region = \"all\",comb_al_ls[[paste0(i,\"_all\")]]$stat_ls$combined$stat_c_df),\n",
    "                       data.frame(region = \"Cortex\",comb_al_ls[[paste0(i,\"_Cortex\")]]$stat_ls$combined$stat_c_df),\n",
    "                       data.frame(region = \"Hippocampus\",comb_al_ls[[paste0(i,\"_Hippocampus\")]]$stat_ls$combined$stat_c_df)))\n",
    "\n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=ifelse(i == \"8mo\",nrow(tmp_df)*2+5, nrow(tmp_df)*3+7), row.names=FALSE)\n",
    "  \n",
    "  tmp_df <- cbind(rbind(data.frame(comb_al_ls[[paste0(i,\"_all\")]]$stat_ls$combined$stat_t_df),\n",
    "                        data.frame(comb_al_ls[[paste0(i,\"_Cortex\")]]$stat_ls$combined$stat_t_df),\n",
    "                        data.frame(comb_al_ls[[paste0(i,\"_Hippocampus\")]]$stat_ls$combined$stat_t_df)))\n",
    "\n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=9, startRow=ifelse(i == \"8mo\",nrow(tmp_df)*2+5, nrow(tmp_df)*3+7), row.names=FALSE)\n",
    "}\n",
    "#subtype\n",
    "for(j in c(\"Ex\",\"Micro\", \"Oligo\", \"Astro\", \"Inhi\", \"CADG\")){#\"Ex\",\"Micro\", \"Oligo\", \"Astro\", \"Inhi\", \"CADG\"\n",
    "  sheet <- createSheet(wb, paste(j,\"Subtype\"))\n",
    "  #8mo \n",
    "  for(i in c(\"8mo\",\"13mo\")){\n",
    "    tmp_df <- cbind(rbind(data.frame(region = \"all\",comb_al_ls[[paste0(i,\"_all\")]]$stat_ls[[j]]$perc_df),\n",
    "                          data.frame(region = \"Cortex\",comb_al_ls[[paste0(i,\"_Cortex\")]]$stat_ls[[j]]$perc_df),\n",
    "                          data.frame(region = \"Hippocampus\",comb_al_ls[[paste0(i,\"_Hippocampus\")]]$stat_ls[[j]]$perc_df)))\n",
    "  \n",
    "    tmp_df[3:8] <- tmp_df[3:8] * 100#portion to percentage\n",
    "    addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=ifelse(i == \"8mo\",1, nrow(tmp_df)+3), row.names=FALSE)\n",
    "    \n",
    "    tmp_df <- cbind(rbind(data.frame(comb_al_ls[[paste0(i,\"_all\")]]$stat_ls[[j]]$dens_df),\n",
    "                          data.frame(comb_al_ls[[paste0(i,\"_Cortex\")]]$stat_ls[[j]]$dens_df),\n",
    "                          data.frame(comb_al_ls[[paste0(i,\"_Hippocampus\")]]$stat_ls[[j]]$dens_df))[,-1])\n",
    "    \n",
    "    addDataFrame(tmp_df, sheet=sheet, startColumn=10, startRow=ifelse(i == \"8mo\",1, nrow(tmp_df)+3), row.names=FALSE)\n",
    "  \n",
    "    tmp_df <- cbind(rbind(data.frame(region = \"all\",comb_al_ls[[paste0(i,\"_all\")]]$stat_ls[[j]]$stat_c_df),\n",
    "                          data.frame(region = \"Cortex\",comb_al_ls[[paste0(i,\"_Cortex\")]]$stat_ls[[j]]$stat_c_df),\n",
    "                          data.frame(region = \"Hippocampus\",comb_al_ls[[paste0(i,\"_Hippocampus\")]]$stat_ls[[j]]$stat_c_df)))\n",
    "  \n",
    "    addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=ifelse(i == \"8mo\",nrow(tmp_df)*2+5, nrow(tmp_df)*3+7), row.names=FALSE)\n",
    "    \n",
    "    tmp_df <- cbind(rbind(data.frame(comb_al_ls[[paste0(i,\"_all\")]]$stat_ls[[j]]$stat_t_df),\n",
    "                          data.frame(comb_al_ls[[paste0(i,\"_Cortex\")]]$stat_ls[[j]]$stat_t_df),\n",
    "                          data.frame(comb_al_ls[[paste0(i,\"_Hippocampus\")]]$stat_ls[[j]]$stat_t_df))[,-1])\n",
    "  \n",
    "    addDataFrame(tmp_df, sheet=sheet, startColumn=9, startRow=ifelse(i == \"8mo\",nrow(tmp_df)*2+5, nrow(tmp_df)*3+7), row.names=FALSE)\n",
    "  }\n",
    "}\n",
    "\n",
    "saveWorkbook(wb, \"May27_cell_composition_stat.xlsx\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c995934d",
   "metadata": {},
   "source": [
    "## 64-gene Validation\n",
    "\n",
    "### Stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f8390e8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "starmap_64@meta.data$cell_type <- starmap_64@meta.data$predicted_cell_type\n",
    "starmap_64@meta.data$top_level <- as.character(starmap_64@meta.data$top_level)\n",
    "starmap_64@meta.data$top_level[starmap_64@meta.data$top_level == \"Ex\"] <- \"CTX-Ex\"\n",
    "starmap_64@meta.data$top_level[starmap_64@meta.data$top_level == \"Inhi\"] <- \"Inh\"\n",
    "\n",
    "starmap_64 <- ad_calc_plaque_dist2(starmap_64,starmap_64@misc[[\"AD_mouse9721_morph\"]], 400, \n",
    "                                   \"AD_mouse9721\", top_level_palette_list, \"all\")\n",
    "starmap_64 <- ad_calc_plaque_dist2(starmap_64,starmap_64@misc[[\"AD_mouse9919_morph\"]], 400, \n",
    "                                   \"AD_mouse9919\", top_level_palette_list, \"all\")\n",
    "\n",
    "\n",
    "for(i in c(\"AD_mouse9721\",\"AD_mouse9919\")){\n",
    "  top_level_list <- c(\"CTX-Ex\", \"CA1\",\"CA2\",\"CA3\", \"DG\",\"Endo\", \"Inh\", \"Micro\", \"Astro\", \"OPC\",\"Oligo\", \"SMC\")\n",
    "  region_size <- dim(starmap_64@misc[[paste0(i,\"_morph\")]]$plaque)[1] * dim(starmap_64@misc[[paste0(i,\"_morph\")]]$plaque)[2]\n",
    "  distr_stat_ls[[paste(i,\"all\",sep = \"_\")]] <- ad_distr_stat(res_ls = starmap_64@misc[[paste0(i,\"_plaque_\",\"all\")]],\n",
    "                                                           sample_name = i,\n",
    "                                                           region_size = region_size,\n",
    "                                                           cell_type_list = cell_type_list[2:4],\n",
    "                                                           top_level_list = top_level_list, region = \"all\")\n",
    "\n",
    "}\n",
    "#Combined Analysis\n",
    "comb_al_ls[[\"8mo_all_val\"]] <- ad_distr_combined_analysis(sample_list = c(\"AD_mouse9721\"), \n",
    "                                                              res_ls = distr_stat_ls, \n",
    "                                                              region = \"all\", cell_type_list = cell_type_list[2:4])\n",
    "\n",
    "comb_al_ls[[\"13mo_all_val\"]] <- ad_distr_combined_analysis(sample_list = c(\"AD_mouse9919\"), \n",
    "                                                          res_ls = distr_stat_ls, \n",
    "                                                          region = \"all\", cell_type_list = cell_type_list[2:4])\n",
    "wb = createWorkbook()\n",
    "#top level\n",
    "sheet <- createSheet(wb, \"Major types validation\")\n",
    "#8mo \n",
    "for(i in c(\"8mo\",\"13mo\")){\n",
    "  tmp_df <- data.frame(region = \"all\",comb_al_ls[[paste0(i,\"_all_val\")]]$stat_ls$combined$perc_df)\n",
    "\n",
    "  tmp_df[3:8] <- tmp_df[3:8] * 100#portion to percentage\n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=ifelse(i == \"8mo\",1, nrow(tmp_df)+3), row.names=FALSE)\n",
    "  \n",
    "  tmp_df <- data.frame(comb_al_ls[[paste0(i,\"_all_val\")]]$stat_ls$combined$dens_df)\n",
    "  \n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=10, startRow=ifelse(i == \"8mo\",1, nrow(tmp_df)+3), row.names=FALSE)\n",
    "\n",
    "  tmp_df <- data.frame(region = \"all\",comb_al_ls[[paste0(i,\"_all_val\")]]$stat_ls$combined$stat_c_df)\n",
    "\n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=ifelse(i == \"8mo\",nrow(tmp_df)*2+5, nrow(tmp_df)*3+7), row.names=FALSE)\n",
    "  \n",
    "  tmp_df <- data.frame(comb_al_ls[[paste0(i,\"_all_val\")]]$stat_ls$combined$stat_t_df)\n",
    "\n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=9, startRow=ifelse(i == \"8mo\",nrow(tmp_df)*2+5, nrow(tmp_df)*3+7), row.names=FALSE)\n",
    "}\n",
    "\n",
    "saveWorkbook(wb, \"May31_cell_composition_val_stat.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebe852e1",
   "metadata": {},
   "source": [
    "### Stacked Barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50824ff6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Export Stacked barplot\n",
    "sample_list <- c('AD_mouse9721','AD_mouse9919')\n",
    "plot_list <- list()\n",
    "area_name = \"all\"\n",
    "\n",
    "for(i in 1:2){\n",
    "    cell_meta <- starmap_64@misc[[paste(sep = \"_\",sample_list[i],\"plaque\", area_name)]][[paste(sep = \"_\", sample_list[i],\"cell_meta\")]]\n",
    "    palette_list <- top_level_palette_list[-5]\n",
    "    j = sum(starmap_64@misc[[paste(sep = \"_\",sample_list[i],\"morph\")]][[\"plaque\"]] >= 0)\n",
    "      \n",
    "    plot_list[[paste(i,area_name,sep = \"_\")]] <- ad_plot_stacked_bar(cell_meta$interval,\n",
    "                                          cell_meta$top_level,\n",
    "                                          palette_list,\n",
    "                                          paste(\"Top Level\",sample_list[i],area_name),\n",
    "                                          starmap_64@misc[[paste(sep = \"_\",sample_list[i],\"plaque\",area_name)]][[paste(sep = \"_\",sample_list[i],\"plaque_dilate_area\")]] , img_size = j)\n",
    "  }\n",
    "\n",
    "#Export A/M/O/OPC overall density\n",
    "tmp_df <- data.frame(cell_type = rep(c(\"Astro\",\"Micro\",\"Oligo\",\"OPC\"),each = 8),\n",
    "                     group = rep(c(\"exp\",\"val\"), each = 4, time = 4),\n",
    "                     time = rep(c(\"8mo\",\"13mo\"), each = 2, time = 8),\n",
    "                     condition = rep(c(\"AD\",\"Ctrl\"), time = 16),\n",
    "                     sample = rep(c(\"ADmouse_9723,ADmouse_9723_2\",\"ADmouse_9735,ADmouse_9707\",\n",
    "                                    \"ADmouse_9494,ADmouse_11346\",\"ADmouse_9498,ADmouse_11351\",\n",
    "                                    \"AD_mouse9721\",\"AD_mouse9781\",\n",
    "                                    \"AD_mouse9919\",\"AD_mouse9930\"), time = 4),\n",
    "                     density = 0)\n",
    "# All Stat\n",
    "tmp_df$label <- paste(tmp_df$group,tmp_df$condition,tmp_df$time,sep = \"_\")\n",
    "tmp_df$label <- factor(tmp_df$label, \n",
    "                       levels = c(\"exp_Ctrl_8mo\",\"exp_AD_8mo\",\"exp_Ctrl_13mo\",\"exp_AD_13mo\",\n",
    "                                  \"val_Ctrl_8mo\",\"val_AD_8mo\",\"val_Ctrl_13mo\",\"val_AD_13mo\"))\n",
    "tmp_df$cell_type <- as.character(tmp_df$cell_type)\n",
    "\n",
    "tmp_df$density <- apply(tmp_df,1,\n",
    "                        FUN = function(x){\n",
    "                          region_size = 0\n",
    "                          cell_count = 0\n",
    "                          sample_list <- str_split(x[[\"sample\"]],\",\") %>% unlist\n",
    "                          for(i in sample_list){\n",
    "                            cell_count <- cell_count + ifelse(x[[\"group\"]] == \"exp\",\n",
    "                                                              sum(starmap_raw@meta.data$sample == i & starmap_raw@meta.data$top_level == x[[\"cell_type\"]]),\n",
    "                                                              sum(starmap_64@meta.data$sample == i & starmap_64@meta.data$top_level == x[[\"cell_type\"]]))\n",
    "                            if(x[[\"group\"]] == \"exp\") tmp_img <- starmap_raw@misc[[paste0(i,\"_morph\")]]$plaque\n",
    "                            else tmp_img <- starmap_64@misc[[paste0(i,\"_morph\")]]$plaque\n",
    "                            region_size <- region_size + dim(tmp_img)[1] * dim(tmp_img)[2]\n",
    "                          }\n",
    "                          cell_count/region_size*3.3^2*1000000\n",
    "                        })\n",
    "\n",
    "plot_list[[\"Density_comparison\"]] <- ggplot(tmp_df,aes(x = label, y = density, fill = cell_type)) + \n",
    "  geom_bar(stat = \"identity\") + \n",
    "  theme_classic() + \n",
    "  theme(axis.text.x = element_text(angle = 45, hjust = 0.9, vjust = 0.9)) + \n",
    "  scale_fill_manual(values = top_level_palette_list)\n",
    "\n",
    "pdf(\"May31_FigS10_barplot.pdf\", width = 8, height = 6)\n",
    "plot_list %>% print()\n",
    "dev.off()\n",
    "\n",
    "# Alveus Stat\n",
    "# Assign region label for validation\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "706fb0bd",
   "metadata": {},
   "source": [
    "### DEG/SDEG Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f10dd2e8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "plot_list <- list()\n",
    "#SDEG\n",
    "#8mo\n",
    "gene_list <- SDEG_ls$`8mo`$gene[SDEG_ls$`8mo`$p_val < 0.01 & SDEG_ls$`8mo`$cluster %in% c(2,6)] %>%\n",
    "  str_to_upper() %>% intersect(row.names(starmap_64@assays$RNA@counts))\n",
    "#13mo\n",
    "gene_list <- SDEG_ls$`13mo`$gene[SDEG_ls$`13mo`$p_val < 0.01 & SDEG_ls$`13mo`$cluster %in% c(2,6)] %>%\n",
    "  str_to_upper() %>% intersect(row.names(starmap_64@assays$RNA@counts))\n",
    "\n",
    "group_name = \"13_months\"\n",
    "sample_list <- c(\"8_months\" = \"AD_mouse9721\",\"13_months\" = \"AD_mouse9919\")\n",
    "\n",
    "tmp_df <- matrix(0, nrow = length(gene_list), ncol = 5) %>% as.data.frame(row.names = gene_list)\n",
    "colnames(tmp_df) <- paste(\"R\",1:5,\"0\",sep = \"\")\n",
    "      \n",
    "tmp_mat <- starmap_64@assays[[\"RNA\"]]@counts[gene_list,\n",
    "                                              starmap_64@meta.data$time == group_name]\n",
    "cell_meta <- starmap_64@misc[[paste0(sample_list[[group_name]],\"_plaque_all\")]][[paste0(sample_list[[group_name]],\"_cell_meta\")]]\n",
    "      \n",
    "for (i in 1:4) {\n",
    "        tmp_df[,i] <- rowSums(tmp_mat[,cell_meta$interval == paste(i,\"0um\",sep = \"\")])  /\n",
    "          sum(cell_meta$interval == paste(i,\"0um\",sep = \"\"))\n",
    "\n",
    "}\n",
    "tmp_df[,5] <- rowSums(tmp_mat[,cell_meta$interval %in% c(\"other\",\"50um\")])  /\n",
    "          sum(cell_meta$interval %in% c(\"other\",\"50um\"))\n",
    "require(ComplexHeatmap)\n",
    "row.names(tmp_df) <- str_to_title(row.names(tmp_df))\n",
    "plot_list[[paste0(group_name,\"_SDEG\")]] <- Heatmap(t(scale(t(tmp_df[,1:5]))), \n",
    "                                             cluster_columns = F,rect_gp = gpar(col = \"black\", lwd = 1),\n",
    "                                             cluster_rows  = T, show_row_names = T,\n",
    "                                             row_title = paste(group_name))\n",
    "## Average Expr\n",
    "gene_list <- list()\n",
    "#Micro\n",
    "gene_list[[\"Micro\"]] <- c(\"P2RY12\",\"CALM1\",\"GPR34\",\"TGFBR1\",\"CST7\", \"APOE\", \"CTSB\", \"CD9\",  \"C1QA\", \"TREM2\",\n",
    "               \"CD83\",\"CTSS\")\n",
    "#Astro\n",
    "gene_list[[\"Astro\"]] <- c(\"ALDOC\", \"CST3\", \"GFAP\", \"VIM\", \"AQP4\",\"CD63\")\n",
    "#Oligo\n",
    "gene_list[[\"Oligo\"]] <- c('PLP1','CLDN11','KLK6',\"SERPINA3N\",\"C4B\",\"CDKN1A\",\"CD9\")\n",
    "#CTX-Ex\n",
    "gene_list[[\"CTX-Ex\"]] <- c(\"GPM6A\",\"DDIT3\",\"CCNB2\")\n",
    "#Inh\n",
    "gene_list[[\"Inh\"]] <- c(\"GPM6A\",\"DDIT3\",\"CCNB2\")\n",
    "#DG\n",
    "gene_list[[\"DG\"]] <- c(\"GPM6A\",\"DDIT3\",\"CCNB2\")\n",
    "\n",
    "sample_list <- c(\"exp_Ctrl_8mo\" =\"ADmouse_9735,ADmouse_9707\", \"exp_AD_8mo\" =\"ADmouse_9723,ADmouse_9723_2\",\n",
    "                 \"exp_Ctrl_13mo\"=\"ADmouse_9498,ADmouse_11351\",\"exp_AD_13mo\"=\"ADmouse_9494,ADmouse_11346\",\n",
    "                 \"val_Ctrl_8mo\" =\"AD_mouse9781\",\"val_AD_8mo\" =\"AD_mouse9721\",\n",
    "                 \"val_Ctrl_13mo\"=\"AD_mouse9930\",\"val_AD_13mo\"=\"AD_mouse9919\")\n",
    "\n",
    "tmp_vec <- c(\"Micro\",\"Astro\",\"Oligo\",\"CTX-Ex\",\"Inh\",\"DG\")\n",
    "v_heatmap <- NULL\n",
    "for(cell_type in tmp_vec){\n",
    "  tmp_df <- data.frame(gene = gene_list[[cell_type]],\n",
    "                       exp_Ctrl_8mo = 0,exp_AD_8mo = 0,\n",
    "                       exp_Ctrl_13mo = 0,exp_AD_13mo = 0,\n",
    "                       val_Ctrl_8mo = 0,val_AD_8mo = 0,\n",
    "                       val_Ctrl_13mo = 0,val_AD_13mo = 0)\n",
    "  for(i in colnames(tmp_df)){\n",
    "    tmp_df[[i]] <- apply(tmp_df,1,\n",
    "                           FUN = function(x){\n",
    "                             if(str_detect(i,\"exp\")){\n",
    "                               sample_vec <- str_split(sample_list[[i]],\",\") %>% unlist\n",
    "                               tmp_vec <- starmap_raw@meta.data$sample %in% sample_vec & \n",
    "                                 starmap_raw@meta.data$top_level == cell_type\n",
    "                               mean(starmap_raw@assays$RNA@counts[str_to_title(x[[\"gene\"]]), tmp_vec])\n",
    "                             }else if(str_detect(i,\"val\")){\n",
    "                               sample_vec <- str_split(sample_list[[i]],\",\") %>% unlist\n",
    "                               tmp_vec <- starmap_64@meta.data$sample %in% sample_vec & \n",
    "                                 starmap_64@meta.data$top_level == cell_type\n",
    "                               mean(starmap_64@assays$RNA@counts[x[[\"gene\"]], tmp_vec])\n",
    "                             }else x[[i]]\n",
    "                           })\n",
    "  }\n",
    "  tmp_df[,6:9] <- t(scale(t(as.matrix(tmp_df[,6:9]))))\n",
    "  tmp_df[,2:5] <- t(scale(t(as.matrix(tmp_df[,2:5]))))\n",
    "  tmp_mat <- as.matrix(tmp_df[2:9])\n",
    "  row.names(tmp_mat) <- str_to_title(tmp_df$gene)\n",
    "  v_heatmap <- v_heatmap %v% Heatmap(tmp_mat, column_split = rep(c(\"exp\",\"val\"), each = 4), \n",
    "          rect_gp = gpar(col = \"black\", lwd = 1),\n",
    "          cluster_rows = F, cluster_columns = F, row_title = cell_type, show_row_names = T)\n",
    "}\n",
    "\n",
    "plot_list[[\"avg_expr\"]] <- v_heatmap\n",
    "\n",
    "pdf(\"May31_FigS10_heatmap.pdf\", width = 6, height = 12)\n",
    "plot_list[[\"avg_expr\"]] %>% print()\n",
    "plot_list$`8_months_SDEG` %v% plot_list$`13_months_SDEG` %>% print()\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c322ef71",
   "metadata": {},
   "source": [
    "## Alveus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b76dc17",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "starmap <- LoadH5Seurat(\"~/Desktop/At_Broad/AD/data/2022-08-28-Hu-AD-stardist-scaled.h5seurat\")\n",
    "starmap@meta.data$cell_type_bk <- starmap@meta.data$cell_type\n",
    "starmap@meta.data$cell_type <- as.character(starmap@meta.data$cell_type_bk)\n",
    "starmap@meta.data$cell_type[starmap@meta.data$top_level == \"Astro\"] <- as.character(starmap@meta.data$cell_type_test[starmap@meta.data$top_level == \"Astro\"])\n",
    "starmap@meta.data$cell_type <- as.factor(starmap@meta.data$cell_type)\n",
    "\n",
    "starmap_64@meta.data[[\"region\"]] <- \"NA\"\n",
    "starmap_64@meta.data[[\"cell_type\"]] <- starmap_64@meta.data[[\"predicted_cell_type\"]]  %>% as.character()\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Oligo\"] <- \"Oligo1\"\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Oligo_Klk6\"] <- \"Oligo2\"\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Oligo_Cldn11\"] <- \"Oligo3\"\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Astro\"] <- \"Astro1\"\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Astro_Cst3\"] <- \"Astro2\"\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Astro_Gfap/Vim\"] <- \"Astro3\"\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Micro\"] <- \"Micro1\"\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Micro_Gpr34\"] <- \"Micro2\"\n",
    "starmap_64@meta.data[[\"cell_type\"]][starmap_64@meta.data[[\"predicted_cell_type\"]] == \"Micro_Cst7/Ctsb\"] <- \"Micro3\"\n",
    "\n",
    "#starmap_64\n",
    "for(i in c(\"AD_mouse9781\", \"AD_mouse9721\", \"AD_mouse9930\", \"AD_mouse9919\")){\n",
    "  message(i)\n",
    "  tmp_img <- readImage(paste0(\"/Users/zhouhaowen/Desktop/At_Broad/AD/data/2021-06-22-mAD-molecular-labels/64 genes/\",i,\"-molecular-labels.tif\"),\n",
    "                       type = \"tiff\")\n",
    "  tmp_img <- Image(round(tmp_img[,,1] * 10))\n",
    "\n",
    "  starmap_64@meta.data[[\"region\"]] <- apply(starmap_64@meta.data,1,\n",
    "                                            FUN = function(x){\n",
    "                                              pos_x <- as.numeric(x[[\"x\"]])\n",
    "                                              pos_y <- as.numeric(x[[\"y\"]])\n",
    "                                              if(x[[\"sample\"]] == i){\n",
    "                                                if(as.logical(tmp_img[pos_x,pos_y] == 9)) \"Alveus\"\n",
    "                                                else \"NA\"\n",
    "                                              } \n",
    "                                              else x[[\"region\"]]\n",
    "                                            })\n",
    "  \n",
    "}\n",
    "sample_list <- c(\"exp_Ctrl_8mo\" =\"ADmouse_9735,ADmouse_9707\", \"exp_AD_8mo\" =\"ADmouse_9723,ADmouse_9723_2\",\n",
    "                 \"exp_Ctrl_13mo\"=\"ADmouse_9498,ADmouse_11351\",\"exp_AD_13mo\"=\"ADmouse_9494,ADmouse_11346\",\n",
    "                 \"val_Ctrl_8mo\" =\"AD_mouse9781\",\"val_AD_8mo\" =\"AD_mouse9721\",\n",
    "                 \"val_Ctrl_13mo\"=\"AD_mouse9930\",\"val_AD_13mo\"=\"AD_mouse9919\")\n",
    "tmp_df <- data.frame(group = rep(c(\"exp\",\"val\"),each = 4), \n",
    "                     condition = rep(c(\"Ctrl_8mo\", \"AD_8mo\",\"Ctrl_13mo\",\"AD_13mo\",    \n",
    "                                       \"Ctrl_8mo\",\"AD_8mo\",\"Ctrl_13mo\",\"AD_13mo\"), 1),\n",
    "                     sample = sample_list)\n",
    "\n",
    "\n",
    "sample_list <- c(\"exp_Ctrl_8mo1\" =\"ADmouse_9735\", \"exp_AD_8mo1\" =\"ADmouse_9723\",\n",
    "                 \"exp_Ctrl_8mo2\" =\"ADmouse_9707\", \"exp_AD_8mo2\" =\"ADmouse_9723_2\",\n",
    "                 \"exp_Ctrl_13mo1\"=\"ADmouse_9498\",\"exp_AD_13mo1\"=\"ADmouse_9494\",\n",
    "                 \"exp_Ctrl_13mo2\"=\"ADmouse_11351\",\"exp_AD_13mo2\"=\"ADmouse_11346\",\n",
    "                 \"val_Ctrl_8mo\" =\"AD_mouse9781\",\"val_AD_8mo\" =\"AD_mouse9721\",\n",
    "                 \"val_Ctrl_13mo\"=\"AD_mouse9930\",\"val_AD_13mo\"=\"AD_mouse9919\")\n",
    "#Stat\n",
    "tmp_df <- data.frame(group = rep(c(\"exp\",\"exp\",\"val\"),each = 4), \n",
    "                     condition = rep(c(\"Ctrl_8mo\", \"AD_8mo\",\"Ctrl_8mo\",\"AD_8mo\",\"Ctrl_13mo\",\"AD_13mo\",    \n",
    "                                       \"Ctrl_13mo\",\"AD_13mo\",\"Ctrl_8mo\",\"AD_8mo\",\"Ctrl_13mo\",\"AD_13mo\"), 1),\n",
    "                     sample = sample_list)\n",
    "\n",
    "tmp_df[[\"alveus_area\"]] <- 0\n",
    "for(i in seq_len(nrow(tmp_df))){\n",
    "  message(tmp_df$sample[i])\n",
    "  tmp_vec <- str_split(tmp_df$sample[i], pattern = \",\") %>% unlist\n",
    "  for(j in tmp_vec){\n",
    "    if(tmp_df$group[i] == \"exp\"){\n",
    "      tmp_df$alveus_area[i] <- tmp_df$alveus_area[i] + sum(starmap@misc[[paste0(j,\"_morph\")]]$region == 3)\n",
    "    }else if(tmp_df$group[i] == \"val\"){\n",
    "      tmp_img <- readImage(paste0(\"/Users/zhouhaowen/Desktop/At_Broad/AD/data/2021-06-22-mAD-molecular-labels/64 genes/\",j,\"-molecular-labels.tif\"),\n",
    "                       type = \"tiff\")\n",
    "      tmp_img <- Image(round(tmp_img[,,1] * 10))\n",
    "      tmp_df$alveus_area[i] <- tmp_df$alveus_area[i] + sum(tmp_img == 9)\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "for(i in c(\"Oligo1\",\"Oligo2\",\"Oligo3\",\"OPC\",\"Astro1\",\"Astro2\",\"Astro3\",\"Micro1\",\"Micro2\",\"Micro3\")){#\"Astro\",\"Micro\",\"Oligo\",\"OPC\"\n",
    "  tmp_df[[i]] <- apply(tmp_df,1,\n",
    "                       FUN = function(x){\n",
    "                         tmp_vec <- str_split(x[[\"sample\"]], pattern = \",\") %>% unlist\n",
    "                         if(x[[\"group\"]] == \"exp\"){\n",
    "                           count_num <- sum(starmap@meta.data$region == \"Alveus\" & starmap@meta.data$cell_type == i & starmap@meta.data$sample %in% tmp_vec)\n",
    "                         }else if(x[[\"group\"]] == \"val\"){\n",
    "                           count_num <- sum(starmap_64@meta.data$region == \"Alveus\" & starmap_64@meta.data$cell_type == i & starmap_64@meta.data$sample %in% tmp_vec)\n",
    "                         }\n",
    "                         count_num/as.numeric(x[[\"alveus_area\"]])*3.3^2*1000000\n",
    "                       })\n",
    "}\n",
    "\n",
    "\n",
    "tmp_df[[\"label\"]] <- paste(tmp_df$group,tmp_df$condition,sep = \"_\")\n",
    "tmp_df[[\"label\"]] <- factor(tmp_df[[\"label\"]], levels = c(\"exp_Ctrl_8mo\",\"exp_AD_8mo\",\"exp_Ctrl_13mo\",\"exp_AD_13mo\",\n",
    "                                                          \"val_Ctrl_8mo\",\"val_AD_8mo\",\"val_Ctrl_13mo\",\"val_AD_13mo\"))\n",
    "ggplot(tmp_df[,c(5:8,15)] %>% melt(id.vars = \"label\"),aes(x = label, y = value, fill = variable)) + \n",
    "  geom_bar(stat = \"identity\") + \n",
    "  theme_classic() + \n",
    "  theme(axis.text.x = element_text(angle = 90)) +\n",
    "  scale_fill_manual(values = top_level_palette_list)\n",
    "# stat test\n",
    "ad_alveus_t_test <- function(condA = \"AD_8mo\", condB = \"Ctrl_8mo\", cell_type = \"Astro\", test_group = \"exp\"){\n",
    "  t.test(x = tmp_df[[cell_type]][tmp_df$condition == condA & tmp_df$group == test_group], \n",
    "         y = tmp_df[[cell_type]][tmp_df$condition == condB & tmp_df$group == test_group], alternative = \"greater\")%>%\n",
    "    .$p.value\n",
    "}\n",
    "ad_alveus_t_test(condA = \"AD_8mo\",  condB = \"Ctrl_8mo\",  cell_type = \"Oligo1\")\n",
    "ad_alveus_t_test(condA = \"AD_13mo\", condB = \"Ctrl_13mo\", cell_type = \"Oligo1\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e10c5dd",
   "metadata": {},
   "source": [
    "## Gene Module\n",
    "### Func Def"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e754fa8a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ad_get_spatial_avg_expr <- function(starmap_obj, sample_ls, gene_ls, cell_type_ls){\n",
    "  expr_df <- data.frame(gene = rep(gene_ls , times = length(sample_ls)),\n",
    "                        sample = \"-\",\n",
    "                        R10 = 0,R20 = 0,R30 = 0,R40 = 0,\"R40+\" = 0,\n",
    "                        RA10 = 0,RA20 = 0,RA30 = 0,RA40 = 0,\"RA40+\" = 0,\n",
    "                        RB10 = 0,RB20 = 0,RB30 = 0,RB40 = 0,\"RB40+\" = 0)\n",
    "  expr_df[[\"sample\"]] <- rep(sample_ls, each = length(gene_ls))\n",
    "  expr_df[[\"group\"]] <- rep(names(sample_ls), each = length(gene_ls))\n",
    "  \n",
    "  colnames(expr_df)[7] <- \"R40+\"\n",
    "  colnames(expr_df)[12] <- \"RA40+\"\n",
    "  colnames(expr_df)[17] <- \"RB40+\"\n",
    "  \n",
    "  sample_vec <- unlist(sample_ls)\n",
    "  \n",
    "  for(i in 1:length(sample_vec)){\n",
    "    if(i == 1){\n",
    "      tmp_mat <- starmap_obj@assays$raw@data[,starmap_obj@meta.data$sample == sample_vec[1]]\n",
    "      cell_meta <- data.frame(sample = sample_vec[1],\n",
    "                              starmap_obj@misc[[paste0(sample_vec[1],\"_plaque_all\")]][[paste0(sample_vec[1],\"_cell_meta\")]]) \n",
    "    }else{\n",
    "      tmp_mat <- cbind(tmp_mat,\n",
    "                       starmap_obj@assays$raw@data[,starmap_obj@meta.data$sample == sample_vec[i]])\n",
    "      cell_meta <- rbind(cell_meta,\n",
    "                         data.frame(sample = sample_vec[i],\n",
    "                         starmap_obj@misc[[paste0(sample_vec[i],\"_plaque_all\")]][[paste0(sample_vec[i],\"_cell_meta\")]]  ) )\n",
    "    }\n",
    "  }\n",
    "  tmp_mat <- tmp_mat[,cell_meta$top_level %in% cell_type_ls]\n",
    "  cell_meta <- cell_meta[cell_meta$top_level %in% cell_type_ls, ]\n",
    "  \n",
    "  #c(\"R10\",\"R20\",\"R30\",\"R40\",\"R40+\")\n",
    "  for(i in 1:5){\n",
    "    expr_df[[i+2]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(i == 5){\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                            }else{\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                            }\n",
    "                          })\n",
    "    expr_df[[i+7]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(i == 5){\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]])[1] ])\n",
    "                            }else{\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                             cell_meta$sample == unlist(x[[\"sample\"]])[1] ])\n",
    "                            }\n",
    "                          })\n",
    "    expr_df[[i+12]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(i == 5){\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$interval %in% c(\"50um\",\"other\") & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]])[2] ])\n",
    "                            }else{\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$interval == paste0(i,\"0um\") & \n",
    "                                             cell_meta$sample == unlist(x[[\"sample\"]])[2] ])\n",
    "                            }\n",
    "                          })\n",
    "  }\n",
    "\n",
    "  expr_df\n",
    "}\n",
    "\n",
    "ad_get_spatial_tau_avg_expr <- function(starmap_obj, sample_ls, gene_ls, top_level, tau_res_ls){\n",
    "  expr_df <- data.frame(gene = rep(gene_ls , times = length(sample_ls)),\n",
    "                        sample = \"-\",\n",
    "                        T00 = 0,T50 = 0,T100 = 0,T100p = 0,T100np = 0,\n",
    "                        TA00 = 0,TA50 = 0,TA100 = 0,TA100p = 0,TA100np = 0,\n",
    "                        TB00 = 0,TB50 = 0,TB100 = 0,TB100p = 0,TB100np = 0)\n",
    "  expr_df[[\"sample\"]] <- rep(sample_ls, each = length(gene_ls))\n",
    "  expr_df[[\"group\"]] <- rep(names(sample_ls), each = length(gene_ls))\n",
    "\n",
    "  sample_vec <- unlist(sample_ls)\n",
    "  \n",
    "  for(i in 1:length(sample_vec)){\n",
    "    if(i == 1){\n",
    "      tmp_mat <- starmap_obj@assays$raw@data[,starmap_obj@meta.data$sample == sample_vec[1]]\n",
    "      cell_meta <- tau_res_ls[[paste(sample_vec[i],top_level,\"res\",sep=\"_\")]]$cell_meta\n",
    "    }else{\n",
    "      tmp_mat <- cbind(tmp_mat,\n",
    "                       starmap_obj@assays$raw@data[,starmap_obj@meta.data$sample == sample_vec[i]])\n",
    "      cell_meta <- rbind(cell_meta,\n",
    "                         tau_res_ls[[paste(sample_vec[i],top_level,\"res\",sep=\"_\")]]$cell_meta)\n",
    "    }\n",
    "  }\n",
    "  tmp_mat <- tmp_mat[,cell_meta$top_level == top_level]\n",
    "  cell_meta <- cell_meta[cell_meta$top_level == top_level, ]\n",
    "  \n",
    "  factor_ls <- c(\"Zero\",\"Low\", \"High\",\"High w ABeta\", \"High w/o ABeta\")\n",
    "  #c(\"R10\",\"R20\",\"R30\",\"R40\",\"R40+\")\n",
    "  for(i in 1:5){\n",
    "    expr_df[[i+2]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(i == 3){\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$tau_block %in% factor_ls[4:5] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                            }else{\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$tau_block == factor_ls[i] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                            }\n",
    "                          })\n",
    "    expr_df[[i+7]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(i == 3){\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$tau_block %in% factor_ls[4:5] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                            }else{\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$tau_block == factor_ls[i] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                            }\n",
    "                          })\n",
    "    expr_df[[i+12]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            if(i == 3){\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$tau_block %in% factor_ls[4:5] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                            }else{\n",
    "                              mean(tmp_mat[x[[\"gene\"]], cell_meta$tau_block == factor_ls[i] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                            }\n",
    "                          })\n",
    "  }\n",
    "\n",
    "  expr_df\n",
    "}\n",
    "\n",
    "ad_get_spatial_col_avg_expr <- function(starmap_obj, sample_ls, gene_ls, top_level, \n",
    "                                        cell_meta_from = \"plaque\",#obj or plaque\n",
    "                                        col_sel, # which column used for extract\n",
    "                                        col_vec ){ # selected levels\n",
    "  expr_df <- data.frame(gene = rep(gene_ls , times = length(sample_ls)),\n",
    "                        sample = \"-\")\n",
    "  for(j in c(\"combined\",\"A\",\"B\")){\n",
    "    for(i in col_vec){\n",
    "      expr_df[[paste0(i,\"_\",j)]] <- 0\n",
    "    }\n",
    "  }\n",
    "  expr_df[[\"sample\"]] <- rep(sample_ls, each = length(gene_ls))\n",
    "  expr_df[[\"group\"]] <- rep(names(sample_ls), each = length(gene_ls))\n",
    "\n",
    "  sample_vec <- unlist(sample_ls)\n",
    "  \n",
    "  for(i in 1:length(sample_vec)){\n",
    "    if(i == 1){\n",
    "      tmp_mat <- starmap_obj@assays$raw@data[,starmap_obj@meta.data$sample == sample_vec[1]]\n",
    "      if(cell_meta_from == \"plaque\"){\n",
    "        cell_meta <- data.frame(sample = sample_vec[1],\n",
    "                              starmap_obj@misc[[paste0(sample_vec[1],\"_plaque_all\")]][[paste0(sample_vec[1],\"_cell_meta\")]]) \n",
    "      }else if(cell_meta_from == \"obj\"){\n",
    "        cell_meta <- data.frame(sample = sample_vec[1],\n",
    "                                subset(starmap_obj@meta.data,sample == sample_vec[1])) \n",
    "      }\n",
    "      \n",
    "    }else{\n",
    "      tmp_mat <- cbind(tmp_mat,\n",
    "                       starmap_obj@assays$raw@data[,starmap_obj@meta.data$sample == sample_vec[i]])\n",
    "      if(cell_meta_from == \"plaque\"){\n",
    "        cell_meta <- rbind(cell_meta,\n",
    "                         data.frame(sample = sample_vec[i],\n",
    "                         starmap_obj@misc[[paste0(sample_vec[i],\"_plaque_all\")]][[paste0(sample_vec[i],\"_cell_meta\")]]  ) )\n",
    "      }else if(cell_meta_from == \"obj\"){\n",
    "        cell_meta <- rbind(cell_meta,\n",
    "                         data.frame(sample = sample_vec[i],\n",
    "                         subset(starmap_obj@meta.data,sample == sample_vec[i])) )\n",
    "      }\n",
    "      \n",
    "    }\n",
    "  }\n",
    "  tmp_mat <- tmp_mat[,cell_meta$top_level %in% top_level]\n",
    "  cell_meta <- cell_meta[cell_meta$top_level %in% top_level, ]\n",
    "  \n",
    "  \n",
    "  for(i in seq_len(length(col_vec))){\n",
    "    expr_df[[i+2]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta[[col_sel]] == col_vec[i] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]]) ])\n",
    "                          })\n",
    "    expr_df[[i+2+length(col_vec)]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta[[col_sel]] == col_vec[i] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]])[1] ])\n",
    "                          })\n",
    "    expr_df[[i+2+length(col_vec)*2]] <- apply(expr_df,1,\n",
    "                          FUN = function(x){\n",
    "                            mean(tmp_mat[x[[\"gene\"]], cell_meta[[col_sel]] == col_vec[i] & \n",
    "                                             cell_meta$sample %in% unlist(x[[\"sample\"]])[2] ])\n",
    "                          })\n",
    "  }\n",
    "\n",
    "  expr_df\n",
    "}\n",
    "\n",
    "ad_get_spatial_sig <- function(starmap_obj, expr_df, cell_type_ls){\n",
    "  sample_vec <-  unique(unlist(expr_df[[\"sample\"]]))\n",
    "  for(i in 1:length(sample_vec)){\n",
    "    if(i == 1){\n",
    "      tmp_mat <- starmap_obj@assays$raw@data[,starmap_obj@meta.data$sample == sample_vec[1]]\n",
    "      cell_meta <- data.frame(sample = sample_vec[1],\n",
    "                              starmap_obj@misc[[paste0(sample_vec[1],\"_plaque_all\")]][[paste0(sample_vec[1],\"_cell_meta\")]]) \n",
    "    }else{\n",
    "      tmp_mat <- cbind(tmp_mat,\n",
    "                       starmap_obj@assays$raw@data[,starmap_obj@meta.data$sample == sample_vec[i]])\n",
    "      cell_meta <- rbind(cell_meta,\n",
    "                         data.frame(sample = sample_vec[i],\n",
    "                         starmap_obj@misc[[paste0(sample_vec[i],\"_plaque_all\")]][[paste0(sample_vec[i],\"_cell_meta\")]]  ) )\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  tmp_mat <- tmp_mat[,cell_meta$top_level %in% cell_type_ls]\n",
    "  cell_meta <- cell_meta[cell_meta$top_level %in% cell_type_ls, ]\n",
    "  \n",
    "  expr_df$max_int <- apply(expr_df,1,FUN = function(x){which(as.numeric(x[3:6]) == max(as.numeric(x[3:6])))})\n",
    "  expr_df$p_val <- apply(expr_df,1,\n",
    "                             FUN = function(x){\n",
    "                               i = x[[\"max_int\"]]\n",
    "                               tmp_vec1 <- tmp_mat[x[[\"gene\"]], cell_meta$interval == paste(i,\"0um\",sep = \"\")]\n",
    "                               tmp_vec2 <- tmp_mat[x[[\"gene\"]], cell_meta$interval != paste(i,\"0um\",sep = \"\")]\n",
    "                               a <- wilcox.test(tmp_vec1,tmp_vec2, alternative = \"greater\")$p.value\n",
    "                               return(a)\n",
    "                             })\n",
    "  \n",
    "  \n",
    "  expr_df$p_val_adj <- p.adjust(expr_df$p_val)\n",
    "  expr_df\n",
    "\n",
    "}\n",
    "\n",
    "my_run.scWGCNA <- function (p.cells, s.cells, idents, features, is.pseudocell = T, \n",
    "          min.cells = 10, less = F, g.names) \n",
    "{\n",
    "  backup.options = options()\n",
    "  options(stringsAsFactors = FALSE)\n",
    "  my.trees = list()\n",
    "  if (!missing(idents)) {\n",
    "    s.Wdata = subset(s.cells, idents = idents)\n",
    "  }\n",
    "  else {\n",
    "    s.Wdata = s.cells\n",
    "  }\n",
    "  if (missing(g.names)) {\n",
    "    gnames = data.frame(x = rownames(p.cells), y = rownames(p.cells), \n",
    "                        row.names = rownames(p.cells))\n",
    "  }\n",
    "  else {\n",
    "    gnames = g.names\n",
    "    rownames(gnames) = gnames[1, ]\n",
    "  }\n",
    "  nonex = which(apply(s.Wdata@assays$RNA@counts, 1, function(x) length(which(x > \n",
    "                                                                               0))) < min.cells)\n",
    "  if (missing(features)) {\n",
    "    Expr = calc.vargenes(s.Wdata, min.cells)\n",
    "  }\n",
    "  else {\n",
    "    Expr = features\n",
    "  }\n",
    "  if (is.pseudocell == T) {\n",
    "    datExpr = p.cells@assays[[Seurat::DefaultAssay(p.cells)]]@counts[Expr, \n",
    "    ]\n",
    "    if (length(which(apply(datExpr, 1, var) > 0)) > 0) {\n",
    "      print(paste0(\"The following variable genes were not found expressed in the pseudocell object:  \", \n",
    "                   names(which(Matrix::rowSums(datExpr) == 0))))\n",
    "    }\n",
    "    datExpr = datExpr[which(apply(datExpr, 1, var) > 0), \n",
    "    ]\n",
    "    Expr = rownames(datExpr)\n",
    "  }\n",
    "  else {\n",
    "    datExpr = s.Wdata@assays$RNA@data[Expr, ]\n",
    "  }\n",
    "  print(paste0(\"We have \", length(Expr), \" genes in the variable genes object\"))\n",
    "  datExpr = t(as.matrix(datExpr))\n",
    "  my.power = choose.sfttp(datExpr)\n",
    "  my.Clnumber = 20\n",
    "  change = 0\n",
    "  genesets = list()\n",
    "  nonsig = 1\n",
    "  while (nonsig != 0) {\n",
    "    my.adjacency = WGCNA::adjacency(datExpr, type = \"signed\", \n",
    "                                    power = my.power, corFnc = \"bicor\")\n",
    "    TOM = WGCNA::TOMsimilarityFromExpr(datExpr, networkType = \"signed\", \n",
    "                                       TOMType = \"signed\", power = my.power, corType = \"bicor\")\n",
    "    colnames(TOM) <- gnames[colnames(datExpr), 2]\n",
    "    rownames(TOM) <- gnames[colnames(datExpr), 2]\n",
    "    dissTOM = 1 - TOM\n",
    "    geneTree = hclust(as.dist(dissTOM), method = \"average\")\n",
    "    nMerge = length(geneTree$height)\n",
    "    refQuantile = 0.05\n",
    "    refMerge = round(nMerge * refQuantile)\n",
    "    refHeight = geneTree$height[refMerge]\n",
    "    cutheight = signif(0.99 * (max(geneTree$height) - refHeight) + \n",
    "                         refHeight, 4)\n",
    "    x = seq(7, 30, 1)\n",
    "    y = seq(cutheight - 5e-04, cutheight + 5e-04, 1e-04)\n",
    "    w = data.frame()\n",
    "    for (i in x) {\n",
    "      for (j in y) {\n",
    "        sink(\"aux\")\n",
    "        z = table(dynamicTreeCut::cutreeDynamic(dendro = geneTree, \n",
    "                                                method = \"tree\", minClusterSize = i, deepSplit = T, \n",
    "                                                cutHeight = j, verbose = 0))\n",
    "        sink(NULL)\n",
    "        if(dim(z) > 0 ){\n",
    "          v = data.frame(i, j, dim(z), names(z[1]), unname(z[1]))\n",
    "          w = rbind(w, v)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    my.height = w$j[which(w$unname.z.1.. == min(w$unname.z.1..))]\n",
    "    if (cutheight %in% my.height) {\n",
    "      my.Clsize = w[which(w$j == cutheight), ]\n",
    "    }\n",
    "    else {\n",
    "      my.Clsize = w[which(w$j == max(my.height)), ]\n",
    "    }\n",
    "    if (change == 2) {\n",
    "      my.Clnumber = length(table(dynamicColors)) + 1\n",
    "    }\n",
    "    change = change + 1\n",
    "    if (any(w$names.z.1.. == 1)) {\n",
    "      my.Clsize = w[which(w$names.z.1.. == 1), ]\n",
    "      if (any(my.Clsize$dim.z. >= (my.Clnumber - 1))) {\n",
    "        my.Clsize = my.Clsize[which(my.Clsize$dim.z. >= \n",
    "                                      (my.Clnumber - 1)), , drop = F]\n",
    "      }\n",
    "      else {\n",
    "        my.Clsize = my.Clsize[which(my.Clsize$dim.z. == \n",
    "                                      max(my.Clsize$dim.z.)), , drop = F]\n",
    "      }\n",
    "      my.Clsize = my.Clsize[which(my.Clsize$dim.z. == min(my.Clsize$dim.z.)), \n",
    "                            , drop = F]\n",
    "      my.Clsize = my.Clsize[which(my.Clsize$i == max(my.Clsize$i)), \n",
    "                            , drop = F]\n",
    "      if (cutheight %in% my.Clsize$j) {\n",
    "        my.height = cutheight\n",
    "      }\n",
    "      else {\n",
    "        my.height = max(my.Clsize$j)\n",
    "      }\n",
    "      my.Clsize = max(my.Clsize$i)\n",
    "    }\n",
    "    if (!any(w$names.z.1.. == 1)) {\n",
    "      if (any(my.Clsize$dim.z. >= my.Clnumber)) {\n",
    "        my.Clsize = my.Clsize[which(my.Clsize$dim.z. >= \n",
    "                                      my.Clnumber), , drop = F]\n",
    "      }\n",
    "      else {\n",
    "        my.Clsize = my.Clsize[which(my.Clsize$dim.z. == \n",
    "                                      max(my.Clsize$dim.z.)), , drop = F]\n",
    "      }\n",
    "      my.Clsize = my.Clsize[which(my.Clsize$dim.z. == min(my.Clsize$dim.z.)), \n",
    "                            , drop = F]\n",
    "      my.Clsize = my.Clsize[which(my.Clsize$i == max(my.Clsize$i)), \n",
    "                            , drop = F]\n",
    "      if (cutheight %in% my.Clsize$j) {\n",
    "        my.height = cutheight\n",
    "      }\n",
    "      else {\n",
    "        my.height = max(my.Clsize$j)\n",
    "      }\n",
    "      my.Clsize = max(my.Clsize$i)\n",
    "    }\n",
    "    if (change < 3) {\n",
    "      my.Clsize = 15\n",
    "      my.height = cutheight\n",
    "    }\n",
    "    print(paste(\"my.height: \", my.height, \" .... my.Clsize: \", \n",
    "                my.Clsize))\n",
    "    dynamicMods = dynamicTreeCut::cutreeDynamic(dendro = geneTree, \n",
    "                                                method = \"tree\", minClusterSize = my.Clsize, deepSplit = T, \n",
    "                                                cutHeight = my.height)\n",
    "    table(dynamicMods)\n",
    "    dynamicColors = WGCNA::labels2colors(dynamicMods)\n",
    "    table(dynamicColors)\n",
    "    my.trees[[length(my.trees) + 1]] = list(geneTree, dynamicColors)\n",
    "    MEList = WGCNA::moduleEigengenes(as.matrix(datExpr), \n",
    "                                     colors = dynamicColors)\n",
    "    MEs = MEList$eigengenes\n",
    "    geneModuleMembership = as.data.frame(WGCNA::signedKME(datExpr, \n",
    "                                                          MEs))\n",
    "    MMPvalue = as.data.frame(WGCNA::corPvalueStudent(as.matrix(geneModuleMembership), \n",
    "                                                     nrow(datExpr)))\n",
    "    x = c()\n",
    "    xy = list()\n",
    "    dcols = 1:length(levels(as.factor(dynamicColors)))\n",
    "    grey.genes = length(which(dynamicColors == \"grey\"))\n",
    "    if (any(levels(as.factor(dynamicColors)) == \"grey\")) {\n",
    "      dcols = dcols[-which(levels(as.factor(dynamicColors)) == \n",
    "                             \"grey\")]\n",
    "    }\n",
    "    for (i in dcols) {\n",
    "      modGenes = rownames(MMPvalue)[which(dynamicColors == \n",
    "                                            levels(as.factor(dynamicColors))[i] & MMPvalue[, \n",
    "                                                                                           i] < 0.01)]\n",
    "      x = c(x, modGenes)\n",
    "      xy[[i]] = modGenes\n",
    "    }\n",
    "    genesets[[length(genesets) + 1]] = colnames(datExpr)\n",
    "    cat(paste0(grey.genes, \" genes not assigned to any module.\", \n",
    "               \"\\n\", length(which(!(colnames(datExpr) %in% x))) - \n",
    "                 grey.genes, \" genes excluded due to significance.\"))\n",
    "    nonsig = length(which(!(colnames(datExpr) %in% x)))\n",
    "    if (length(which(!(colnames(datExpr) %in% x))) != 0) {\n",
    "      dynamicColors = dynamicColors[-which(!(colnames(datExpr) %in% \n",
    "                                               x))]\n",
    "      datExpr = datExpr[, -(which(!(colnames(datExpr) %in% \n",
    "                                      x)))]\n",
    "    }\n",
    "    if (change == 2 & less == T) {\n",
    "      cat(\"\\n\\nIMPORTANT NOTE!!!\\nYou have run this analysis witht the option less=TRUE. This means that the analysis will try to reduce the number of modules detected, based on their expression pattern. If modules have very similar expression profile (distance < 0.25), they will be merged. Moreover, if a module seems to be highly expressed in only 1-3 cells ( cells expressing >2*(max(expression)/3) ), it will be removed.\\n\")\n",
    "      my.filtered = scWGCNA:::FilterMods_int(dynamicColors = dynamicColors, \n",
    "                                   s.Wdata = s.Wdata, datExpr = datExpr, geneTree = geneTree, \n",
    "                                   my.power = my.power)\n",
    "      par(mfrow = c(1, 1))\n",
    "      datExpr = my.filtered[[1]]\n",
    "      dynamicColors = my.filtered[[2]]\n",
    "    }\n",
    "  }\n",
    "  my.memberships = list()\n",
    "  my.cols = as.factor(dynamicColors)\n",
    "  for (m in 1:length(levels(my.cols))) {\n",
    "    my.mem = cbind(geneModuleMembership[my.cols == levels(my.cols)[m], \n",
    "                                        m, drop = F], MMPvalue[my.cols == levels(my.cols)[m], \n",
    "                                                               m, drop = F])\n",
    "    colnames(my.mem) = c(\"Membership\", \"p.val\")\n",
    "    my.memberships[[paste0(m, \"_\", levels(my.cols)[m])]] = my.mem\n",
    "  }\n",
    "  s.MEList = MEList\n",
    "  raw.datExpr = s.Wdata@assays$RNA@data[colnames(datExpr), \n",
    "  ]\n",
    "  raw.datExpr = t(as.matrix(raw.datExpr))\n",
    "  raw.MEList = WGCNA::moduleEigengenes(raw.datExpr, colors = dynamicColors)\n",
    "  s.MEList = raw.MEList\n",
    "  names(dynamicColors) = colnames(datExpr)\n",
    "  WGCNA_data = list()\n",
    "  WGCNA_data[[\"modules\"]] = my.memberships\n",
    "  WGCNA_data[[\"expression\"]] = datExpr\n",
    "  WGCNA_data[[\"dynamicMods\"]] = dynamicMods\n",
    "  WGCNA_data[[\"dynamicCols\"]] = dynamicColors\n",
    "  WGCNA_data[[\"MEList\"]] = MEList\n",
    "  WGCNA_data[[\"MEs\"]] = MEs\n",
    "  WGCNA_data[[\"module.genes\"]] = xy\n",
    "  WGCNA_data[[\"sc.MEList\"]] = s.MEList\n",
    "  WGCNA_data[[\"genesets.history\"]] = genesets\n",
    "  WGCNA_data[[\"moduletrees.history\"]] = my.trees\n",
    "  WGCNA_data[[\"TOM\"]] = TOM\n",
    "  WGCNA_data[[\"adjacency\"]] = my.adjacency\n",
    "  return(WGCNA_data)\n",
    "  options(backup.options)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "880b110d",
   "metadata": {},
   "source": [
    "### scWGCNA\n",
    "\n",
    "**Require R 4.0.0+**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2d6a023",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(scWGCNA)\n",
    "## scWGCNA\n",
    "library(WGCNA)\n",
    "\n",
    "options(stringsAsFactors = FALSE);\n",
    "sample_list <- list(\"8mo\" = c(\"ADmouse_9723\",\"ADmouse_9723_2\"), \"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\"))\n",
    "#Use Counts for WGCNA\n",
    "sub_list <- list(\"all\" = levels(starmap_raw@meta.data$top_level),\n",
    "              \"Neuron\" = c(\"CTX-Ex\", \"Inh\", \"CA1\", \"CA2\", \"CA3\", \"DG\"),\n",
    "              \"Glia\" = c(\"Micro\", \"Astro\", \"Oligo\"),\n",
    "              \"Micro\" = \"Micro\", \"Astro\" = \"Astro\",\"Oligo\" = \"Oligo\")\n",
    "\n",
    "starmap_raw <- readRDS(\"Aug04_starmap_w_all_spatial.RDS\")\n",
    "scWGCNA_res <- list()\n",
    "scWGCNA_less_res <- list()\n",
    "\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  message(paste(\"Prepare to run scWGCNA for\",i))\n",
    "  starmap_sub <- subset(starmap_raw,top_level %in% sub_list[[i]] & time == \"13months\" & group == \"disease\")\n",
    "  #Using SDEG\n",
    "  #starmap_sub <- starmap_sub[SDEG_ls[[\"13mo_cell_type\"]] %>% subset(group == i & p_val < 0.05) %>% .$gene,]\n",
    "  #Using custom gene list\n",
    "  tmp_vec <- c()\n",
    "  for(j in sub_list[[i]]){\n",
    "    tmp_vec <- read.csv(paste0(\"/Users/zhouhaowen/Desktop/At_Broad/AD/Aug17_Jiahao_gene_list/\",j,\".csv\")) %>% .$Gene %>% c(tmp_vec)\n",
    "  }\n",
    "  tmp_vec <- unique(tmp_vec)\n",
    "  starmap_sub <- starmap_sub[tmp_vec,]\n",
    "  starmap_sub@misc <- list()\n",
    "  starmap_sub <- ScaleData(starmap_sub, features = row.names(starmap_sub))\n",
    "  starmap_sub <- RunPCA(starmap_sub, features = row.names(starmap_sub), verbose = F)\n",
    "  \n",
    "  # scWGCNA Calculate the pseudocells\n",
    "  starmap_sub.pc <-   calculate.pseudocells(s.cells = starmap_sub, # Single cells in Seurat object\n",
    "                                            seeds=0.2, # Fraction of cells to use as seeds to aggregate pseudocells\n",
    "                                            nn = 10, # Number of neighbors to aggregate\n",
    "                                            reduction = \"pca\", # Reduction to use\n",
    "                                            dims = 1:10) # The dimensions to use\n",
    "  bicor = WGCNA::bicor\n",
    "  message(paste(i,\"run scWGCNA\"))\n",
    "  starmap_sub.scWGCNA = my_run.scWGCNA(p.cells = starmap_sub.pc, # Pseudocells (recommended), or Seurat single cells\n",
    "                                    s.cells = starmap_sub, # single cells in Seurat format\n",
    "                                    is.pseudocell = T, # We are using single cells twice this time\n",
    "                                    less = F,\n",
    "                                    features = rownames(starmap_sub)) # Recommended: variable genes\n",
    "  scWGCNA_res[[i]] <-  starmap_sub.scWGCNA\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b322034e",
   "metadata": {},
   "source": [
    "### Spatial Expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd9ab374",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Spatial interval\n",
    "for(cell_type in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  tmp_WGCNA <- scWGCNA_res[[cell_type]]\n",
    "  #scW.p.dendro(scWGCNA.data = tmp_WGCNA)\n",
    "  sample_list <- list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\"))\n",
    "  tmp_df <- ad_get_spatial_avg_expr(starmap_raw, list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\")),names(tmp_WGCNA[[\"dynamicCols\"]]),sub_list[[cell_type]])\n",
    "  #Heatmap\n",
    "  tmp_df <- data.frame(tmp_df, module = tmp_WGCNA$dynamicMods) %>% arrange(module)\n",
    "  # is DEG\n",
    "  tmp_df$is_DEG <- sapply(tmp_df$gene,function(x){ifelse(as.character(x) %in% DE_13mo[[paste0(cell_type,\"combined\")]]$rowname, \"yes\", \"no\")})\n",
    "  require(ComplexHeatmap)\n",
    "  tmp_df_bk <- tmp_df\n",
    "  tmp_df[,3:7] <- as.matrix(tmp_df[,3:7]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df[,8:12] <- as.matrix(tmp_df[,8:12]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df[,13:17] <- as.matrix(tmp_df[,13:17]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df_bk <- tmp_df_bk[!is.na(tmp_df[,3:17] %>% rowSums()),]\n",
    "  tmp_df <- tmp_df[!is.na(tmp_df[,3:17] %>% rowSums()),]\n",
    "  tmp_df <- ad_get_spatial_sig(starmap_raw,tmp_df,sub_list[[cell_type]])\n",
    "  #is spa sig\n",
    "  tmp_df$is_spatial_sig <- sapply(tmp_df$p_val,function(x){ifelse(as.numeric(x) < 0.05, \"yes\", \"no\")})\n",
    "  scWGCNA_res[[cell_type]][[\"expr\"]] <- tmp_df\n",
    "  scWGCNA_res[[cell_type]][[\"expr_hm\"]] <-\n",
    "  Heatmap(tmp_df[,3:17] %>% as.matrix(),\n",
    "          cluster_columns = F,cluster_rows  = T, show_row_names = F, show_row_dend = T,\n",
    "          row_split = as.factor(paste0(tmp_df$module)), column_title = paste(\"13mo\", cell_type),\n",
    "          row_title_gp = gpar())\n",
    "  tmp_df2 <- tmp_df %>% group_by(module) %>% \n",
    "    summarise(MR10 = mean(R10),MR20 = mean(R20),MR30 = mean(R30),MR40 = mean(R40),MR40. = mean(R40.),\n",
    "              MRA10 = mean(RA10),MRA20 = mean(RA20),MRA30 = mean(RA30),MRA40 = mean(RA40),MRA40. = mean(RA40.),\n",
    "              MRB10 = mean(RB10),MRB20 = mean(RB20),MRB30 = mean(RB30),MBR40 = mean(RB40),MRB40. = mean(RB40.))\n",
    "  row.names(tmp_df2) <- paste0(\"mod_\",tmp_df2$module)\n",
    "  scWGCNA_res[[cell_type]][[\"expr_avg\"]] <- tmp_df2\n",
    "  scWGCNA_res[[cell_type]][[\"expr_avg_hm\"]] <-\n",
    "  Heatmap(tmp_df2[,2:16] %>% as.matrix() %>% `row.names<-`(paste0(\"mod_\",tmp_df2$module)),\n",
    "          cluster_columns = F,cluster_rows = F, show_row_names = T, show_row_dend = F,\n",
    "          row_title = paste(\"13mo\", cell_type), column_split = rep(c(\"Combined\",\"sample1\",\"sample2\"),each = 5),\n",
    "          border = TRUE, rect_gp = gpar(col = \"black\", lwd = 1)) \n",
    "  \n",
    "  tmp_df_bk <- subset(tmp_df_bk,module == switch(cell_type,\"Micro\" = 1, \"Astro\" = 3, \"Oligo\" = 1))\n",
    "  tmp_df <- subset(tmp_df,module == switch(cell_type,\"Micro\" = 1, \"Astro\" = 3, \"Oligo\" = 1)) %>%\n",
    "    subset(is_DEG == \"yes\" | is_spatial_sig == \"yes\") %>% arrange(p_val)\n",
    "  row.names(tmp_df) <- tmp_df$gene\n",
    "\n",
    "  row_anno <- rowAnnotation(DEG = tmp_df$is_DEG, \n",
    "                             Spatial_sig = tmp_df$is_spatial_sig,\n",
    "                          col = list(DEG = c(\"yes\" = \"darkred\", \"no\" = \"white\"),\n",
    "                                      Spatial_sig = c(\"yes\" = \"darkblue\", \"no\" = \"white\")))\n",
    "\n",
    "  scWGCNA_res[[cell_type]][[\"expr_hm_sel\"]] <-\n",
    "  Heatmap(tmp_df[,8:17] %>% as.matrix(),column_split = rep(c(\"sample1\",\"sample2\"),each = 5),\n",
    "        cluster_columns = F,cluster_rows  = F, show_row_names = T, show_row_dend = T,\n",
    "        column_title = paste(\"Scaled Anno Subset 13mo\", cell_type),row_title_gp = gpar(), border = TRUE, \n",
    "        rect_gp = gpar(col = \"black\", lwd = 1), right_annotation = row_anno, \n",
    "        top_annotation = HeatmapAnnotation(expr_sum = anno_barplot(colSums(tmp_df[,8:17]))))\n",
    "  #scWGCNA_res[[cell_type]][[\"expr_hm_sel_raw\"]] <-\n",
    "  #Heatmap(tmp_df[,8:17] %>% as.matrix(),column_split = rep(c(\"sample1\",\"sample2\"),each = 5),\n",
    "  #      cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = T,\n",
    "  #      column_title = paste(\"Raw Anno Subset 13mo\", cell_type),row_title_gp = gpar(), border = TRUE, \n",
    "  #      rect_gp = gpar(col = \"black\", lwd = 1), right_annotation = row_anno, \n",
    "  #      top_annotation = HeatmapAnnotation(expr_sum = anno_barplot(colSums(tmp_df_bk[,8:17]))))\n",
    "}\n",
    "\n",
    "#Export pdf\n",
    "pdf(\"Aug30_WGCNA_hm.pdf\",width = 6, height = 10)\n",
    "for(cell_type in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  draw(scWGCNA_res[[cell_type]][[\"expr_hm\"]])\n",
    "  draw(scWGCNA_res[[cell_type]][[\"expr_avg_hm\"]])\n",
    "  draw(scWGCNA_res[[cell_type]][[\"expr_hm_sel\"]])\n",
    "  #draw(scWGCNA_res[[cell_type]][[\"expr_hm_sel_raw\"]])\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "# Cell type sep\n",
    "for(cell_type in c(\"Micro\",\"Astro\",\"Oligo\",\"Oligo+OPC\")){\n",
    "  #scW.p.dendro(scWGCNA.data = tmp_WGCNA)\n",
    "  sample_list <- list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\"))\n",
    "  if(cell_type == \"Oligo+OPC\"){\n",
    "    tmp_WGCNA <- scWGCNA_res[[\"Oligo\"]]\n",
    "    scWGCNA_res[[\"Oligo+OPC\"]] <- list()\n",
    "    tmp_df <- ad_get_spatial_col_avg_expr(starmap_raw, \n",
    "                                          list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\")), \n",
    "                                          names(tmp_WGCNA[[\"dynamicCols\"]]),\n",
    "                                          c(\"Oligo\",\"OPC\"),cell_type_list$Oligo)\n",
    "    tmp_vec <- cell_type_list$Oligo\n",
    "  }else if(cell_type == \"Oligo\"){\n",
    "    tmp_WGCNA <- scWGCNA_res[[cell_type]]\n",
    "    tmp_df <- ad_get_spatial_col_avg_expr(starmap_raw, \n",
    "                                          list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\")),\n",
    "                                          names(tmp_WGCNA[[\"dynamicCols\"]]),\n",
    "                                          cell_type,\"cell_type\",\n",
    "                                          c(\"Oligo1\", \"Oligo2\", \"Oligo3\"))\n",
    "    tmp_vec <- c(\"Oligo1\", \"Oligo2\", \"Oligo3\")\n",
    "  }else{\n",
    "    tmp_WGCNA <- scWGCNA_res[[cell_type]]\n",
    "    tmp_df <- ad_get_spatial_col_avg_expr(starmap_raw, \n",
    "                                          list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\")),\n",
    "                                          names(tmp_WGCNA[[\"dynamicCols\"]]),\n",
    "                                          cell_type,\"cell_type\",cell_type_list[[cell_type]])\n",
    "    tmp_vec <- cell_type_list[[cell_type]]\n",
    "  }\n",
    "  \n",
    "  #Heatmap\n",
    "  tmp_df <- data.frame(tmp_df, module = tmp_WGCNA$dynamicMods) %>% arrange(module)\n",
    "  # is DEG\n",
    "  if(cell_type == \"Oligo+OPC\"){\n",
    "    tmp_df$is_DEG <- sapply(tmp_df$gene,function(x){ifelse(as.character(x) %in% DE_13mo[[paste0(\"Oligo\",\"combined\")]]$rowname, \"yes\", \"no\")})\n",
    "  }else{\n",
    "    tmp_df$is_DEG <- sapply(tmp_df$gene,function(x){ifelse(as.character(x) %in% DE_13mo[[paste0(cell_type,\"combined\")]]$rowname, \"yes\", \"no\")})\n",
    "  }\n",
    "  \n",
    "  require(ComplexHeatmap)\n",
    "  tmp_df_bk <- tmp_df\n",
    "  tmp_df[,3:(2+length(tmp_vec))] <- as.matrix(tmp_df[,3:(2+length(tmp_vec))]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df[,(3+length(tmp_vec)):(2+length(tmp_vec)*2)] <-\n",
    "    as.matrix(tmp_df[,(3+length(tmp_vec)):(2+length(tmp_vec)*2)]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df[,(3+length(tmp_vec)*2):(2+length(tmp_vec)*3)] <-\n",
    "    as.matrix(tmp_df[,(3+length(tmp_vec)*2):(2+length(tmp_vec)*3)]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df_bk <- tmp_df_bk[!is.na(tmp_df[,c(3:(2+length(tmp_vec)*3))] %>% rowSums()),]\n",
    "  tmp_df <- tmp_df[!is.na(tmp_df[,c(3:(2+length(tmp_vec)*3))] %>% rowSums()),]\n",
    "  \n",
    "  tmp_df2 <- tmp_df %>% group_by(module) %>% summarise()\n",
    "  for(i in colnames(tmp_df)[3:(2+length(tmp_vec)*3)]){\n",
    "    tmp_df2[[i]] <- data.frame(module = tmp_df$module, score = tmp_df[[i]]) %>% group_by(module) %>%\n",
    "      summarise(avg = mean(score)) %>% .$avg\n",
    "  }\n",
    "\n",
    "  scWGCNA_res[[cell_type]][[\"cell_type_expr\"]] <- tmp_df\n",
    "  scWGCNA_res[[cell_type]][[\"cell_type_expr_hm\"]] <-\n",
    "  Heatmap(tmp_df[,c(3:(2+length(tmp_vec)*3))] %>% as.matrix(),\n",
    "          cluster_columns = F,cluster_rows  = T, show_row_names = F, show_row_dend = T,\n",
    "          row_split = as.factor(paste0(tmp_df$module)), column_title = paste(\"13mo\", cell_type),\n",
    "          row_title_gp = gpar())\n",
    "  \n",
    "  row.names(tmp_df2) <- paste0(\"mod_\",tmp_df2$module)\n",
    "  scWGCNA_res[[cell_type]][[\"cell_type_expr_avg\"]] <- tmp_df2\n",
    "  scWGCNA_res[[cell_type]][[\"cell_type_expr_avg_hm\"]] <-\n",
    "  Heatmap(tmp_df2[,c(2:(1+length(tmp_vec)*3))] %>% as.matrix() %>% `row.names<-`(paste0(\"mod_\",tmp_df2$module)),\n",
    "          cluster_columns = F,cluster_rows = F, show_row_names = T, show_row_dend = F,\n",
    "          row_title = paste(\"13mo\", cell_type), column_split = rep(c(\"Combined\",\"sample1\",\"sample2\"),each = length(tmp_vec)),\n",
    "          border = TRUE, rect_gp = gpar(col = \"black\", lwd = 1)) \n",
    "  \n",
    "  tmp_df <- subset(tmp_df,module == switch(cell_type,\"Micro\" = 1, \"Astro\" = 3, \"Oligo\" = 1, \"Oligo+OPC\" = 1)) %>%\n",
    "  subset(is_DEG == \"yes\")\n",
    "  row.names(tmp_df) <- tmp_df$gene\n",
    "\n",
    "  row_anno <- rowAnnotation(DEG = tmp_df$is_DEG,\n",
    "                          col = list(DEG = c(\"yes\" = \"darkred\", \"no\" = \"white\")))\n",
    "\n",
    "  scWGCNA_res[[cell_type]][[\"cell_type_expr_hm_sel\"]] <-\n",
    "  Heatmap(tmp_df[,c((3+length(tmp_vec)):(2+length(tmp_vec)*3))] %>% as.matrix(),column_split = rep(c(\"sample1\",\"sample2\"),each = length(tmp_vec)),\n",
    "        cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = T,\n",
    "        column_title = paste(\"Scaled Anno Subset 13mo\", cell_type),row_title_gp = gpar(), border = TRUE, \n",
    "        rect_gp = gpar(col = \"black\", lwd = 1), right_annotation = row_anno, \n",
    "        top_annotation = HeatmapAnnotation(expr_sum = anno_barplot(colSums(tmp_df[,(3+length(tmp_vec)):(2+length(tmp_vec)*3)]))))\n",
    "}\n",
    "\n",
    "#Export pdf\n",
    "pdf(\"Sep07_WGCNA_cell_type_avg_hm.pdf\",width = 6, height = 6)\n",
    "for(cell_type in c(\"Micro\",\"Astro\",\"Oligo\",\"Oligo+OPC\")){\n",
    "  draw(scWGCNA_res[[cell_type]][[\"cell_type_expr_avg_hm\"]])\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "# Condition sep\n",
    "for(cell_type in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  #scW.p.dendro(scWGCNA.data = tmp_WGCNA)\n",
    "  sample_list <- list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\",\"ADmouse_9498\",\"ADmouse_11351\"),\n",
    "                      \"8mo\" = c(\"ADmouse_9723\",\"ADmouse_9723_2\",\"ADmouse_9735\",\"ADmouse_9707\"))\n",
    "  tmp_WGCNA <- scWGCNA_res[[cell_type]]\n",
    "  tmp_df <- ad_get_spatial_col_avg_expr(starmap_raw, \n",
    "                                        sample_list,\n",
    "                                        names(tmp_WGCNA[[\"dynamicCols\"]]),\n",
    "                                        cell_type,\"obj\",\"group\",c(\"disease\",\"control\"))\n",
    "  tmp_vec <- cell_type_list[[cell_type]]\n",
    "  \n",
    "  #Heatmap\n",
    "  tmp_df <- left_join(tmp_df, data.frame(gene = names(tmp_WGCNA$dynamicCols), \n",
    "                                         module = tmp_WGCNA$dynamicMods), by = \"gene\") %>% arrange(group)\n",
    "  tmp_df <- data.frame(gene = tmp_df$gene[tmp_df$group == \"8mo\"],\n",
    "                       disease_8mo = tmp_df$disease_combined[tmp_df$group == \"8mo\"],\n",
    "                       control_8mo = tmp_df$control_combined[tmp_df$group == \"8mo\"],\n",
    "                       disease_13mo = tmp_df$disease_combined[tmp_df$group == \"13mo\"],\n",
    "                       control_13mo = tmp_df$control_combined[tmp_df$group == \"13mo\"],\n",
    "                       module = tmp_df$module[tmp_df$group == \"8mo\"])\n",
    "  require(ComplexHeatmap)\n",
    "  tmp_df[,2:5] <- as.matrix(tmp_df[,2:5]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df <- tmp_df[!is.na(tmp_df[,2:5] %>% rowSums()),]\n",
    "  \n",
    "  tmp_df2 <- tmp_df %>% group_by(module) %>% summarise()\n",
    "  for(i in colnames(tmp_df)[2:5]){\n",
    "    tmp_df2[[i]] <- data.frame(module = tmp_df$module, score = tmp_df[[i]]) %>% group_by(module) %>%\n",
    "      summarise(avg = mean(score)) %>% .$avg\n",
    "  }\n",
    "\n",
    "  scWGCNA_res[[cell_type]][[\"group_expr_hm\"]] <-\n",
    "  Heatmap(tmp_df[,2:5] %>% as.matrix(),\n",
    "          cluster_columns = F,cluster_rows  = T, show_row_names = F, show_row_dend = T,\n",
    "          row_split = as.factor(paste0(tmp_df$module)), column_title = paste(\"13mo\", cell_type),\n",
    "          row_title_gp = gpar())\n",
    "  \n",
    "  row.names(tmp_df2) <- paste0(\"mod_\",tmp_df2$module)\n",
    "  #scWGCNA_res[[cell_type]][[\"group_expr_avg\"]] <- tmp_df2\n",
    "  scWGCNA_res[[cell_type]][[\"group_expr_avg_hm\"]] <-\n",
    "  Heatmap(tmp_df2[,2:5] %>% as.matrix() %>% `row.names<-`(paste0(\"mod_\",tmp_df2$module)),\n",
    "          cluster_columns = F,cluster_rows = F, show_row_names = T, show_row_dend = F,\n",
    "          row_title = paste(\"13mo\", cell_type),\n",
    "          border = TRUE, rect_gp = gpar(col = \"black\", lwd = 1)) \n",
    "  \n",
    "  tmp_df <- subset(tmp_df,module == switch(cell_type,\"Micro\" = 1, \"Astro\" = 3, \"Oligo\" = 1, \"Oligo+OPC\" = 1))\n",
    "  row.names(tmp_df) <- tmp_df$gene\n",
    "\n",
    "  scWGCNA_res[[cell_type]][[\"group_expr_hm_sel\"]] <-\n",
    "  Heatmap(tmp_df[,2:5] %>% as.matrix(),\n",
    "        cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = T,\n",
    "        column_title = paste(\"Scaled Anno Subset 13mo\", cell_type),row_title_gp = gpar(), border = TRUE, \n",
    "        rect_gp = gpar(col = \"black\", lwd = 1),\n",
    "        top_annotation = HeatmapAnnotation(expr_sum = anno_barplot(colSums(tmp_df[,2:5]))))\n",
    "}\n",
    "\n",
    "#Export pdf\n",
    "pdf(\"Sep11_WGCNA_group_expr_hm.pdf\",width = 6, height = 6)\n",
    "for(cell_type in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  draw(scWGCNA_res[[cell_type]][[\"group_expr_hm\"]])\n",
    "  draw(scWGCNA_res[[cell_type]][[\"group_expr_avg_hm\"]])\n",
    "  draw(scWGCNA_res[[cell_type]][[\"group_expr_hm_sel\"]])\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "# Chech module significance with AD\n",
    "plot_list <- list()\n",
    "for(cell_type in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  tmp_WGCNA <- scWGCNA_res[[cell_type]]\n",
    "\n",
    "  tmp_df <- data.frame(gene = names(tmp_WGCNA[[\"dynamicCols\"]]), \n",
    "                       module = tmp_WGCNA$dynamicMods) %>% arrange(module)\n",
    "  tmp_df2 <- data.frame(module = unique(tmp_df$module), gene_num = 0, deg_num = 0, p_val = 1)\n",
    "  tmp_df2$gene_num <- sapply(tmp_df2$module,function(x){sum(tmp_df$module == as.numeric(x))})\n",
    "  \n",
    "  tmp_df2$deg_num <- sapply(tmp_df2$module,function(x){sum(tmp_df$gene[tmp_df$module == as.numeric(x)] %in% DE_13mo[[paste0(cell_type,\"combined\")]]$rowname)})\n",
    "  tmp_vec <- c()\n",
    "  for(j in sub_list[[cell_type]]){\n",
    "    tmp_vec <- read.csv(paste0(\"/Users/zhouhaowen/Desktop/At_Broad/AD/Aug17_Jiahao_gene_list/\",j,\".csv\")) %>% .$Gene %>% as.character() %>% c(tmp_vec)\n",
    "  }\n",
    "  tmp_vec <- unique(tmp_vec)\n",
    "  #chisq test\n",
    "\n",
    "  for(j in seq_len(nrow(tmp_df2))){\n",
    "    tmp_obj <- suppressWarnings(chisq.test(x = matrix(c(tmp_df2$deg_num[j],\n",
    "                                                        tmp_df2$gene_num[j] - tmp_df2$deg_num[j],\n",
    "                                                        sum(tmp_vec%in%DE_13mo[[paste0(cell_type,\"combined\")]]$rowname),\n",
    "                                                        sum(!tmp_vec%in%DE_13mo[[paste0(cell_type,\"combined\")]]$rowname)), nrow = 2)))\n",
    "    tmp_df2$p_val[j] <- tmp_obj$p.value\n",
    "  }\n",
    "  tmp_df2[[\"neg_log10p\"]] <- -log10(tmp_df2$p_val)\n",
    "  tmp_df2 <- arrange(tmp_df2,neg_log10p)\n",
    "  tmp_df2$module <- factor(tmp_df2$module,levels = tmp_df2$module)\n",
    "  tmp_df2 <- subset(tmp_df2,gene_num >= 15)\n",
    "  scWGCNA_res[[cell_type]][[\"mod_test\"]] <- tmp_df2\n",
    "  scWGCNA_res[[cell_type]][[\"gene_mod\"]] <- subset(tmp_df, module %in% tmp_df2$module)\n",
    "  plot_list[[cell_type]] <- ggplot(tmp_df2, \n",
    "                           aes(x = module, y = neg_log10p)) + \n",
    "    geom_bar(stat = \"identity\", fill = \"darkblue\") + \n",
    "    coord_flip() + \n",
    "    theme_classic() +\n",
    "    labs(title = cell_type) + \n",
    "    scale_fill_gradientn(colors = rev(brewer.pal(7,\"RdYlGn\")))\n",
    "}\n",
    "\n",
    "#Export gene module res\n",
    "wb <- createWorkbook()\n",
    "sheet <- createSheet(wb, sheetName = \"gene module\")\n",
    "for(i in 1:3){\n",
    "  addDataFrame(scWGCNA_res[[c(\"Micro\",\"Astro\",\"Oligo\")[i]]][[\"gene_mod\"]], sheet,startRow = 1, startColumn = i*3-1)\n",
    "}\n",
    "sheet <- createSheet(wb, sheetName = \"module test\")\n",
    "for(i in 1:3){\n",
    "  addDataFrame(scWGCNA_res[[c(\"Micro\",\"Astro\",\"Oligo\")[i]]][[\"mod_test\"]][,1:4], \n",
    "               sheet,startRow = 1, startColumn = i*6-5)\n",
    "}\n",
    "\n",
    "saveWorkbook(wb,file = \"Sep09_gene_module.xlsx\")\n",
    "\n",
    "#pathway barplot\n",
    "plot_list <- list()\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\")){\n",
    "  tmp_df <- read.csv(paste0(\"/Users/zhouhaowen/Desktop/At_Broad/AD/Pathway_analysis/Aug22_gprofiler_Hu_selected/\",i,\".csv\"))\n",
    "  tmp_df <- tmp_df %>% arrange(adjusted_p_value) %>% .[seq_len(min(15,nrow(.))),]\n",
    "  tmp_df$term_name <- factor(tmp_df$term_name, levels = rev(tmp_df$term_name))\n",
    "  plot_list[[i]] <- ggplot(tmp_df, \n",
    "                           aes(x = term_name, y = negative_log10_of_adjusted_p_value)) + \n",
    "    geom_bar(stat = \"identity\", fill = \"darkblue\") + \n",
    "    coord_flip() + \n",
    "    theme_classic() +\n",
    "    labs(title = i) + \n",
    "    scale_fill_gradientn(colors = rev(brewer.pal(7,\"RdYlGn\")))\n",
    "}\n",
    "\n",
    "pdf(\"Aug29_gene_module_pathway.pdf\", width = 8,height = 6)\n",
    "ggarrange(plotlist = plot_list, ncol = 1,nrow = 1)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5583fda3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## archived tau heatmap\n",
    "for(cell_type in c(\"Oligo\")){\n",
    "  tmp_WGCNA <- scWGCNA_res[[cell_type]]\n",
    "  #scW.p.dendro(scWGCNA.data = tmp_WGCNA)\n",
    "  sample_list <- list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\"))\n",
    "  tmp_df <- ad_get_spatial_tau_avg_expr(starmap_raw, list(\"13mo\" = c(\"ADmouse_9494\",\"ADmouse_11346\")),names(tmp_WGCNA[[\"dynamicCols\"]]),\"Oligo\",tau_res)\n",
    "  #Heatmap\n",
    "  tmp_df <- data.frame(tmp_df, module = tmp_WGCNA$dynamicMods) %>% arrange(module)\n",
    "  # is DEG\n",
    "  #tmp_df$is_DEG <- sapply(tmp_df$gene,function(x){ifelse(as.character(x) %in% DE_13mo[[paste0(cell_type,\"combined\")]]$rowname, \"yes\", \"no\")})\n",
    "  require(ComplexHeatmap)\n",
    "  tmp_df_bk <- tmp_df\n",
    "  tmp_df[,3:5] <- as.matrix(tmp_df[,3:5]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df[,8:10] <- as.matrix(tmp_df[,8:10]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df[,13:15] <- as.matrix(tmp_df[,13:15]) %>% t() %>% scale() %>% t()\n",
    "  tmp_df_bk <- tmp_df_bk[!is.na(tmp_df[,c(3:5,8:10,13:15)] %>% rowSums()),]\n",
    "  tmp_df <- tmp_df[!is.na(tmp_df[,c(3:5,8:10,13:15)] %>% rowSums()),]\n",
    "\n",
    "  scWGCNA_res[[cell_type]][[\"tau_expr\"]] <- tmp_df\n",
    "  scWGCNA_res[[cell_type]][[\"tau_expr_hm\"]] <-\n",
    "  Heatmap(tmp_df[,c(3:5,8:10,13:15)] %>% as.matrix(),\n",
    "          cluster_columns = F,cluster_rows  = T, show_row_names = F, show_row_dend = T,\n",
    "          row_split = as.factor(paste0(tmp_df$module)), column_title = paste(\"13mo\", cell_type),\n",
    "          row_title_gp = gpar())\n",
    "  tmp_df2 <- tmp_df %>% group_by(module) %>% \n",
    "    summarise(MT00 = mean(T00),MT50 = mean(T50),MT100 = mean(T100),MT100p = mean(T100p),MT100np= mean(T100np),\n",
    "              MTA00 = mean(TA00),MTA50 = mean(TA50),MTA100 = mean(TA100),MTA100p = mean(TA100p),MTA100np = mean(TA100np),\n",
    "              MTB00 = mean(TB00),MTB50 = mean(TB50),MTB100 = mean(TB100),MTB100p = mean(TB100p),MTB100np = mean(TB100np))\n",
    "  row.names(tmp_df2) <- paste0(\"mod_\",tmp_df2$module)\n",
    "  scWGCNA_res[[cell_type]][[\"tau_expr_avg\"]] <- tmp_df2\n",
    "  scWGCNA_res[[cell_type]][[\"tau_expr_avg_hm\"]] <-\n",
    "  Heatmap(tmp_df2[,c(2:4,7:9,12:14)] %>% as.matrix() %>% `row.names<-`(paste0(\"mod_\",tmp_df2$module)),\n",
    "          cluster_columns = F,cluster_rows = F, show_row_names = T, show_row_dend = F,\n",
    "          row_title = paste(\"13mo\", cell_type), column_split = rep(c(\"Combined\",\"sample1\",\"sample2\"),each = 3),\n",
    "          border = TRUE, rect_gp = gpar(col = \"black\", lwd = 1)) \n",
    "  \n",
    "  tmp_df_bk <- subset(tmp_df_bk,module == switch(cell_type,\"Micro\" = 1, \"Astro\" = 3, \"Oligo\" = 1))\n",
    "  tmp_df <- subset(tmp_df,module == switch(cell_type,\"Micro\" = 1, \"Astro\" = 3, \"Oligo\" = 1))\n",
    "  row.names(tmp_df) <- tmp_df$gene\n",
    "\n",
    "  scWGCNA_res[[cell_type]][[\"tau_expr_hm_sel\"]] <-\n",
    "  Heatmap(tmp_df[,c(8:10,13:15)] %>% as.matrix(),column_split = rep(c(\"sample1\",\"sample2\"),each = 3),\n",
    "        cluster_columns = F,cluster_rows  = T, show_row_names = T, show_row_dend = T,\n",
    "        column_title = paste(\"Scaled Anno Subset 13mo\", cell_type),row_title_gp = gpar(), border = TRUE, \n",
    "        rect_gp = gpar(col = \"black\", lwd = 1), \n",
    "        top_annotation = HeatmapAnnotation(expr_sum = anno_barplot(colSums(tmp_df[,c(8:10,13:15)]))))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61e57767",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e5acf50",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "gpres_ls <- list()\n",
    "require(gprofiler2)\n",
    "for(i in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\")){\n",
    "  gpres_ls[[i]] <- list() \n",
    "  for(j in scWGCNA_less_res[[i]][[\"dynamicMods\"]] %>% as.factor() %>% levels()){\n",
    "    message(paste(i,\"Mod\",j))\n",
    "    tmp_vec <- scWGCNA_less_res[[\"Neuron\"]][[\"dynamicCols\"]] %>% names() %>% .[scWGCNA_less_res[[i]][[\"dynamicMods\"]] == j]\n",
    "    gpres_ls[[i]][[paste0(\"mod_\",j)]] <- gost(query = tmp_vec,organism = \"mmusculus\")\n",
    "    gpres_ls[[i]][[paste0(\"mod_\",j)]][[\"result\"]][[\"query\"]] <- paste0(\"mod_\",j)\n",
    "    gpres_ls[[i]][[paste0(\"mod_\",j)]][[\"result\"]][[\"mod\"]] <- j\n",
    "  }\n",
    "  tmp_df <- lapply(gpres_ls[[i]],function(x){x[[\"result\"]]}) %>% do.call(rbind.data.frame,.)  %>%\n",
    "    subset(source == \"GO:BP\" & p_value < 0.01)\n",
    "  # Top 5\n",
    "  tmp_df <- tmp_df[sapply(levels(as.factor(tmp_df$mod)),function(x){which(tmp_df$mod== x) %>% .[seq_len(min(length(.), 5))]}) %>% unlist,] %>%\n",
    "    mutate(mod = as.numeric(mod)) %>% arrange(mod, p_value)\n",
    "  tmp_df$query <- factor(tmp_df$query, levels = paste0(\"mod_\", tmp_df$mod %>% as.factor()%>% levels))\n",
    "  gpres_ls[[i]][[\"summary\"]] <- ggplot(tmp_df, aes(x = term_name, y = -log10(p_value), fill = query)) + \n",
    "    geom_bar(stat = \"identity\", position = \"dodge\") + \n",
    "    theme_classic() + \n",
    "    coord_flip() + \n",
    "    labs(title = paste(i,\"p < 0.01 top5 per module\"))\n",
    "}\n",
    "\n",
    "#Export xlsx\n",
    "library(xlsx)\n",
    "wb = createWorkbook()\n",
    "#subtype\n",
    "for(j in c(\"Micro\",\"Astro\",\"Oligo\",\"Neuron\")){#\"Ex\",\"Micro\", \"Oligo\", \"Astro\", \"Inhi\", \"CADG\"\n",
    "  sheet <- createSheet(wb, paste(j,\"Enriched Pathway\"))\n",
    "  tmp_df <- lapply(gpres_ls[[i]],function(x){x[[\"result\"]]}) %>% do.call(rbind.data.frame,.)  %>%\n",
    "    subset(source == \"GO:BP\" & p_value < 0.01)\n",
    "  # All\n",
    "  tmp_df <- tmp_df[sapply(levels(as.factor(tmp_df$mod)),function(x){which(tmp_df$mod== x)}) %>% unlist,] %>%\n",
    "    mutate(mod = as.numeric(mod)) %>% arrange(mod, p_value)\n",
    "  addDataFrame(tmp_df, sheet=sheet, startColumn=1, startRow=1, row.names=FALSE)\n",
    "}\n",
    "\n",
    "saveWorkbook(wb, \"Aug09_gene_module_enriched_pathway.xlsx\")\n",
    "getwd()"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "name,tags,-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "name": "R"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
